{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 6: finalize the population assignment and save it in master SQC file, GWAS covar file, and keep files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages({\n",
    "    library(tidyverse)\n",
    "    library(data.table)\n",
    "}))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1-0) Read the file names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "source('0_file_names.R')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_names <- get_file_names()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1-1) Read the input files\n",
    "\n",
    "- `sqc` (sample quality control) file from UK Biobank has the same order as in fam file for the array fam file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "source('../sample_qc_functions.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_sqc_df <- read_master_sqc(file_names)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's rename PC1-PC40 to Global_PC1-Global_PC40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_sqc_df %>% colnames() %>%\n",
    "lapply(function(x){\n",
    "    ifelse(startsWith(x, 'PC'), paste0('Global_', x), x)\n",
    "}) -> master_sqc_df_colnames\n",
    "\n",
    "colnames(master_sqc_df) <- master_sqc_df_colnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6-2) Read the population-specific PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "bind_rows(\n",
    "    # read from file_names$pop_refinement_pca\n",
    "    \n",
    "    c('white_british', 'non_british_white', 'african', 'e_asian') %>%\n",
    "    lapply(function(pop){\n",
    "        file.path(file_names$pop_refinement_pca, sprintf('ukb24983_%s.eigenvec', pop)) %>%\n",
    "        fread(select=c('#FID', 'IID', paste0('PC', 1:40)), colClasses=c('#FID'='character', 'IID'='character')) %>%\n",
    "        rename('FID' = '#FID') %>% \n",
    "        mutate(population = pop)\n",
    "    }) %>% bind_rows(),\n",
    "    \n",
    "    # read from file_names$pop_specific_pca\n",
    "\n",
    "    c('s_asian') %>%\n",
    "    lapply(function(pop){\n",
    "        file.path(file_names$pop_specific_pca, sprintf('ukb24983_%s.eigenvec', pop)) %>%\n",
    "        fread(select=c('#FID', 'IID', paste0('PC', 1:40)), colClasses=c('#FID'='character', 'IID'='character')) %>%\n",
    "        rename('FID' = '#FID') %>% \n",
    "        mutate(population = pop)\n",
    "    }) %>% bind_rows()\n",
    "\n",
    ") -> pop_specific_PCs_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>population</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>white_british    </td><td>337129</td></tr>\n",
       "\t<tr><td>non_british_white</td><td> 24905</td></tr>\n",
       "\t<tr><td>s_asian          </td><td>  7831</td></tr>\n",
       "\t<tr><td>african          </td><td>  6497</td></tr>\n",
       "\t<tr><td>e_asian          </td><td>  1704</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " population & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t white\\_british     & 337129\\\\\n",
       "\t non\\_british\\_white &  24905\\\\\n",
       "\t s\\_asian           &   7831\\\\\n",
       "\t african           &   6497\\\\\n",
       "\t e\\_asian           &   1704\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 5 × 2\n",
       "\n",
       "| population &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| white_british     | 337129 |\n",
       "| non_british_white |  24905 |\n",
       "| s_asian           |   7831 |\n",
       "| african           |   6497 |\n",
       "| e_asian           |   1704 |\n",
       "\n"
      ],
      "text/plain": [
       "  population        n     \n",
       "1 white_british     337129\n",
       "2 non_british_white  24905\n",
       "3 s_asian             7831\n",
       "4 african             6497\n",
       "5 e_asian             1704"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pop_specific_PCs_df %>% count(population) %>% arrange(-n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6-3) Read \"semi-related\" population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_names$semi_related_file %>%\n",
    "fread(col.names=c('FID', 'IID'), colClasses = 'character') %>%\n",
    "mutate(population = 'related') -> related_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>44632</li><li>3</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 44632\n",
       "\\item 3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 44632\n",
       "2. 3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 44632     3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "related_df %>% dim()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6-4) Join the data frames and finalize the population assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'FID'</li><li>'IID'</li><li>'father'</li><li>'mother'</li><li>'sex_code'</li><li>'batch'</li><li>'sex'</li><li>'affymetrix_field_1'</li><li>'affymetrix_field_2'</li><li>'genotyping_array'</li><li>'Batch'</li><li>'Plate_Name'</li><li>'Well'</li><li>'Cluster_CR'</li><li>'dQC'</li><li>'Internal_Pico_ng_uL'</li><li>'Submitted_Gender'</li><li>'Inferred_Gender'</li><li>'X_intensity'</li><li>'Y_intensity'</li><li>'Submitted_Plate_Name'</li><li>'Submitted_Well'</li><li>'sample_qc_missing_rate'</li><li>'heterozygosity'</li><li>'heterozygosity_pc_corrected'</li><li>'het_missing_outliers'</li><li>'putative_sex_chromosome_aneuploidy'</li><li>'in_kinship_table'</li><li>'excluded_from_kinship_inference'</li><li>'excess_relatives'</li><li>'in_white_British_ancestry_subset'</li><li>'used_in_pca_calculation'</li><li>'Global_PC1'</li><li>'Global_PC2'</li><li>'Global_PC3'</li><li>'Global_PC4'</li><li>'Global_PC5'</li><li>'Global_PC6'</li><li>'Global_PC7'</li><li>'Global_PC8'</li><li>'Global_PC9'</li><li>'Global_PC10'</li><li>'Global_PC11'</li><li>'Global_PC12'</li><li>'Global_PC13'</li><li>'Global_PC14'</li><li>'Global_PC15'</li><li>'Global_PC16'</li><li>'Global_PC17'</li><li>'Global_PC18'</li><li>'Global_PC19'</li><li>'Global_PC20'</li><li>'Global_PC21'</li><li>'Global_PC22'</li><li>'Global_PC23'</li><li>'Global_PC24'</li><li>'Global_PC25'</li><li>'Global_PC26'</li><li>'Global_PC27'</li><li>'Global_PC28'</li><li>'Global_PC29'</li><li>'Global_PC30'</li><li>'Global_PC31'</li><li>'Global_PC32'</li><li>'Global_PC33'</li><li>'Global_PC34'</li><li>'Global_PC35'</li><li>'Global_PC36'</li><li>'Global_PC37'</li><li>'Global_PC38'</li><li>'Global_PC39'</li><li>'Global_PC40'</li><li>'in_Phasing_Input_chr1_22'</li><li>'in_Phasing_Input_chrX'</li><li>'in_Phasing_Input_chrXY'</li><li>'fam_order'</li><li>'has_exome'</li><li>'in_remove_file'</li><li>'f21000'</li><li>'f21000_sub_label'</li><li>'f21000_top_label'</li><li>'f21000_0'</li><li>'f21000_1'</li><li>'f21000_2'</li><li>'pass_QC_filter'</li><li>'pass_filter'</li><li>'self_reported_NBW'</li><li>'self_reported_White'</li><li>'self_reported_Asian'</li><li>'self_reported_Black'</li><li>'self_reported_Mixed_or_Other'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'FID'\n",
       "\\item 'IID'\n",
       "\\item 'father'\n",
       "\\item 'mother'\n",
       "\\item 'sex\\_code'\n",
       "\\item 'batch'\n",
       "\\item 'sex'\n",
       "\\item 'affymetrix\\_field\\_1'\n",
       "\\item 'affymetrix\\_field\\_2'\n",
       "\\item 'genotyping\\_array'\n",
       "\\item 'Batch'\n",
       "\\item 'Plate\\_Name'\n",
       "\\item 'Well'\n",
       "\\item 'Cluster\\_CR'\n",
       "\\item 'dQC'\n",
       "\\item 'Internal\\_Pico\\_ng\\_uL'\n",
       "\\item 'Submitted\\_Gender'\n",
       "\\item 'Inferred\\_Gender'\n",
       "\\item 'X\\_intensity'\n",
       "\\item 'Y\\_intensity'\n",
       "\\item 'Submitted\\_Plate\\_Name'\n",
       "\\item 'Submitted\\_Well'\n",
       "\\item 'sample\\_qc\\_missing\\_rate'\n",
       "\\item 'heterozygosity'\n",
       "\\item 'heterozygosity\\_pc\\_corrected'\n",
       "\\item 'het\\_missing\\_outliers'\n",
       "\\item 'putative\\_sex\\_chromosome\\_aneuploidy'\n",
       "\\item 'in\\_kinship\\_table'\n",
       "\\item 'excluded\\_from\\_kinship\\_inference'\n",
       "\\item 'excess\\_relatives'\n",
       "\\item 'in\\_white\\_British\\_ancestry\\_subset'\n",
       "\\item 'used\\_in\\_pca\\_calculation'\n",
       "\\item 'Global\\_PC1'\n",
       "\\item 'Global\\_PC2'\n",
       "\\item 'Global\\_PC3'\n",
       "\\item 'Global\\_PC4'\n",
       "\\item 'Global\\_PC5'\n",
       "\\item 'Global\\_PC6'\n",
       "\\item 'Global\\_PC7'\n",
       "\\item 'Global\\_PC8'\n",
       "\\item 'Global\\_PC9'\n",
       "\\item 'Global\\_PC10'\n",
       "\\item 'Global\\_PC11'\n",
       "\\item 'Global\\_PC12'\n",
       "\\item 'Global\\_PC13'\n",
       "\\item 'Global\\_PC14'\n",
       "\\item 'Global\\_PC15'\n",
       "\\item 'Global\\_PC16'\n",
       "\\item 'Global\\_PC17'\n",
       "\\item 'Global\\_PC18'\n",
       "\\item 'Global\\_PC19'\n",
       "\\item 'Global\\_PC20'\n",
       "\\item 'Global\\_PC21'\n",
       "\\item 'Global\\_PC22'\n",
       "\\item 'Global\\_PC23'\n",
       "\\item 'Global\\_PC24'\n",
       "\\item 'Global\\_PC25'\n",
       "\\item 'Global\\_PC26'\n",
       "\\item 'Global\\_PC27'\n",
       "\\item 'Global\\_PC28'\n",
       "\\item 'Global\\_PC29'\n",
       "\\item 'Global\\_PC30'\n",
       "\\item 'Global\\_PC31'\n",
       "\\item 'Global\\_PC32'\n",
       "\\item 'Global\\_PC33'\n",
       "\\item 'Global\\_PC34'\n",
       "\\item 'Global\\_PC35'\n",
       "\\item 'Global\\_PC36'\n",
       "\\item 'Global\\_PC37'\n",
       "\\item 'Global\\_PC38'\n",
       "\\item 'Global\\_PC39'\n",
       "\\item 'Global\\_PC40'\n",
       "\\item 'in\\_Phasing\\_Input\\_chr1\\_22'\n",
       "\\item 'in\\_Phasing\\_Input\\_chrX'\n",
       "\\item 'in\\_Phasing\\_Input\\_chrXY'\n",
       "\\item 'fam\\_order'\n",
       "\\item 'has\\_exome'\n",
       "\\item 'in\\_remove\\_file'\n",
       "\\item 'f21000'\n",
       "\\item 'f21000\\_sub\\_label'\n",
       "\\item 'f21000\\_top\\_label'\n",
       "\\item 'f21000\\_0'\n",
       "\\item 'f21000\\_1'\n",
       "\\item 'f21000\\_2'\n",
       "\\item 'pass\\_QC\\_filter'\n",
       "\\item 'pass\\_filter'\n",
       "\\item 'self\\_reported\\_NBW'\n",
       "\\item 'self\\_reported\\_White'\n",
       "\\item 'self\\_reported\\_Asian'\n",
       "\\item 'self\\_reported\\_Black'\n",
       "\\item 'self\\_reported\\_Mixed\\_or\\_Other'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'FID'\n",
       "2. 'IID'\n",
       "3. 'father'\n",
       "4. 'mother'\n",
       "5. 'sex_code'\n",
       "6. 'batch'\n",
       "7. 'sex'\n",
       "8. 'affymetrix_field_1'\n",
       "9. 'affymetrix_field_2'\n",
       "10. 'genotyping_array'\n",
       "11. 'Batch'\n",
       "12. 'Plate_Name'\n",
       "13. 'Well'\n",
       "14. 'Cluster_CR'\n",
       "15. 'dQC'\n",
       "16. 'Internal_Pico_ng_uL'\n",
       "17. 'Submitted_Gender'\n",
       "18. 'Inferred_Gender'\n",
       "19. 'X_intensity'\n",
       "20. 'Y_intensity'\n",
       "21. 'Submitted_Plate_Name'\n",
       "22. 'Submitted_Well'\n",
       "23. 'sample_qc_missing_rate'\n",
       "24. 'heterozygosity'\n",
       "25. 'heterozygosity_pc_corrected'\n",
       "26. 'het_missing_outliers'\n",
       "27. 'putative_sex_chromosome_aneuploidy'\n",
       "28. 'in_kinship_table'\n",
       "29. 'excluded_from_kinship_inference'\n",
       "30. 'excess_relatives'\n",
       "31. 'in_white_British_ancestry_subset'\n",
       "32. 'used_in_pca_calculation'\n",
       "33. 'Global_PC1'\n",
       "34. 'Global_PC2'\n",
       "35. 'Global_PC3'\n",
       "36. 'Global_PC4'\n",
       "37. 'Global_PC5'\n",
       "38. 'Global_PC6'\n",
       "39. 'Global_PC7'\n",
       "40. 'Global_PC8'\n",
       "41. 'Global_PC9'\n",
       "42. 'Global_PC10'\n",
       "43. 'Global_PC11'\n",
       "44. 'Global_PC12'\n",
       "45. 'Global_PC13'\n",
       "46. 'Global_PC14'\n",
       "47. 'Global_PC15'\n",
       "48. 'Global_PC16'\n",
       "49. 'Global_PC17'\n",
       "50. 'Global_PC18'\n",
       "51. 'Global_PC19'\n",
       "52. 'Global_PC20'\n",
       "53. 'Global_PC21'\n",
       "54. 'Global_PC22'\n",
       "55. 'Global_PC23'\n",
       "56. 'Global_PC24'\n",
       "57. 'Global_PC25'\n",
       "58. 'Global_PC26'\n",
       "59. 'Global_PC27'\n",
       "60. 'Global_PC28'\n",
       "61. 'Global_PC29'\n",
       "62. 'Global_PC30'\n",
       "63. 'Global_PC31'\n",
       "64. 'Global_PC32'\n",
       "65. 'Global_PC33'\n",
       "66. 'Global_PC34'\n",
       "67. 'Global_PC35'\n",
       "68. 'Global_PC36'\n",
       "69. 'Global_PC37'\n",
       "70. 'Global_PC38'\n",
       "71. 'Global_PC39'\n",
       "72. 'Global_PC40'\n",
       "73. 'in_Phasing_Input_chr1_22'\n",
       "74. 'in_Phasing_Input_chrX'\n",
       "75. 'in_Phasing_Input_chrXY'\n",
       "76. 'fam_order'\n",
       "77. 'has_exome'\n",
       "78. 'in_remove_file'\n",
       "79. 'f21000'\n",
       "80. 'f21000_sub_label'\n",
       "81. 'f21000_top_label'\n",
       "82. 'f21000_0'\n",
       "83. 'f21000_1'\n",
       "84. 'f21000_2'\n",
       "85. 'pass_QC_filter'\n",
       "86. 'pass_filter'\n",
       "87. 'self_reported_NBW'\n",
       "88. 'self_reported_White'\n",
       "89. 'self_reported_Asian'\n",
       "90. 'self_reported_Black'\n",
       "91. 'self_reported_Mixed_or_Other'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"FID\"                                \"IID\"                               \n",
       " [3] \"father\"                             \"mother\"                            \n",
       " [5] \"sex_code\"                           \"batch\"                             \n",
       " [7] \"sex\"                                \"affymetrix_field_1\"                \n",
       " [9] \"affymetrix_field_2\"                 \"genotyping_array\"                  \n",
       "[11] \"Batch\"                              \"Plate_Name\"                        \n",
       "[13] \"Well\"                               \"Cluster_CR\"                        \n",
       "[15] \"dQC\"                                \"Internal_Pico_ng_uL\"               \n",
       "[17] \"Submitted_Gender\"                   \"Inferred_Gender\"                   \n",
       "[19] \"X_intensity\"                        \"Y_intensity\"                       \n",
       "[21] \"Submitted_Plate_Name\"               \"Submitted_Well\"                    \n",
       "[23] \"sample_qc_missing_rate\"             \"heterozygosity\"                    \n",
       "[25] \"heterozygosity_pc_corrected\"        \"het_missing_outliers\"              \n",
       "[27] \"putative_sex_chromosome_aneuploidy\" \"in_kinship_table\"                  \n",
       "[29] \"excluded_from_kinship_inference\"    \"excess_relatives\"                  \n",
       "[31] \"in_white_British_ancestry_subset\"   \"used_in_pca_calculation\"           \n",
       "[33] \"Global_PC1\"                         \"Global_PC2\"                        \n",
       "[35] \"Global_PC3\"                         \"Global_PC4\"                        \n",
       "[37] \"Global_PC5\"                         \"Global_PC6\"                        \n",
       "[39] \"Global_PC7\"                         \"Global_PC8\"                        \n",
       "[41] \"Global_PC9\"                         \"Global_PC10\"                       \n",
       "[43] \"Global_PC11\"                        \"Global_PC12\"                       \n",
       "[45] \"Global_PC13\"                        \"Global_PC14\"                       \n",
       "[47] \"Global_PC15\"                        \"Global_PC16\"                       \n",
       "[49] \"Global_PC17\"                        \"Global_PC18\"                       \n",
       "[51] \"Global_PC19\"                        \"Global_PC20\"                       \n",
       "[53] \"Global_PC21\"                        \"Global_PC22\"                       \n",
       "[55] \"Global_PC23\"                        \"Global_PC24\"                       \n",
       "[57] \"Global_PC25\"                        \"Global_PC26\"                       \n",
       "[59] \"Global_PC27\"                        \"Global_PC28\"                       \n",
       "[61] \"Global_PC29\"                        \"Global_PC30\"                       \n",
       "[63] \"Global_PC31\"                        \"Global_PC32\"                       \n",
       "[65] \"Global_PC33\"                        \"Global_PC34\"                       \n",
       "[67] \"Global_PC35\"                        \"Global_PC36\"                       \n",
       "[69] \"Global_PC37\"                        \"Global_PC38\"                       \n",
       "[71] \"Global_PC39\"                        \"Global_PC40\"                       \n",
       "[73] \"in_Phasing_Input_chr1_22\"           \"in_Phasing_Input_chrX\"             \n",
       "[75] \"in_Phasing_Input_chrXY\"             \"fam_order\"                         \n",
       "[77] \"has_exome\"                          \"in_remove_file\"                    \n",
       "[79] \"f21000\"                             \"f21000_sub_label\"                  \n",
       "[81] \"f21000_top_label\"                   \"f21000_0\"                          \n",
       "[83] \"f21000_1\"                           \"f21000_2\"                          \n",
       "[85] \"pass_QC_filter\"                     \"pass_filter\"                       \n",
       "[87] \"self_reported_NBW\"                  \"self_reported_White\"               \n",
       "[89] \"self_reported_Asian\"                \"self_reported_Black\"               \n",
       "[91] \"self_reported_Mixed_or_Other\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_df %>% colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>father</th><th scope=col>mother</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>0</td><td>488377</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 3\n",
       "\\begin{tabular}{lll}\n",
       " father & mother & n\\\\\n",
       " <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 & 0 & 488377\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 3\n",
       "\n",
       "| father &lt;int&gt; | mother &lt;int&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 0 | 0 | 488377 |\n",
       "\n"
      ],
      "text/plain": [
       "  father mother n     \n",
       "1 0      0      488377"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_df %>% count(father, mother)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_sqc_df %>%\n",
    "select(-father, -mother) %>%\n",
    "mutate(FID = as.character(FID), IID = as.character(IID)) %>%\n",
    "left_join(\n",
    "    pop_specific_PCs_df, by=c('FID', 'IID')\n",
    ") %>%\n",
    "mutate(\n",
    "    population = if_else(\n",
    "        (FID %in% (related_df %>% pull(FID))) & (IID %in% (related_df %>% pull(IID))),\n",
    "        'related',\n",
    "        population\n",
    "    ),\n",
    "    population = if_else(\n",
    "        pass_QC_filter,\n",
    "        if_else((is.na(population) & pass_filter), 'others', population),\n",
    "        'DO_NOT_PASS_SQC'\n",
    "    )\n",
    ") -> master_sqc_refined_pop_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The number of individuals in each population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 9 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>population</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>white_british    </td><td>337129</td></tr>\n",
       "\t<tr><td>related          </td><td> 44632</td></tr>\n",
       "\t<tr><td>NA               </td><td> 35092</td></tr>\n",
       "\t<tr><td>others           </td><td> 28656</td></tr>\n",
       "\t<tr><td>non_british_white</td><td> 24905</td></tr>\n",
       "\t<tr><td>s_asian          </td><td>  7831</td></tr>\n",
       "\t<tr><td>african          </td><td>  6497</td></tr>\n",
       "\t<tr><td>DO_NOT_PASS_SQC  </td><td>  1931</td></tr>\n",
       "\t<tr><td>e_asian          </td><td>  1704</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 9 × 2\n",
       "\\begin{tabular}{ll}\n",
       " population & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t white\\_british     & 337129\\\\\n",
       "\t related           &  44632\\\\\n",
       "\t NA                &  35092\\\\\n",
       "\t others            &  28656\\\\\n",
       "\t non\\_british\\_white &  24905\\\\\n",
       "\t s\\_asian           &   7831\\\\\n",
       "\t african           &   6497\\\\\n",
       "\t DO\\_NOT\\_PASS\\_SQC   &   1931\\\\\n",
       "\t e\\_asian           &   1704\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 9 × 2\n",
       "\n",
       "| population &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| white_british     | 337129 |\n",
       "| related           |  44632 |\n",
       "| NA                |  35092 |\n",
       "| others            |  28656 |\n",
       "| non_british_white |  24905 |\n",
       "| s_asian           |   7831 |\n",
       "| african           |   6497 |\n",
       "| DO_NOT_PASS_SQC   |   1931 |\n",
       "| e_asian           |   1704 |\n",
       "\n"
      ],
      "text/plain": [
       "  population        n     \n",
       "1 white_british     337129\n",
       "2 related            44632\n",
       "3 NA                 35092\n",
       "4 others             28656\n",
       "5 non_british_white  24905\n",
       "6 s_asian             7831\n",
       "7 african             6497\n",
       "8 DO_NOT_PASS_SQC     1931\n",
       "9 e_asian             1704"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_refined_pop_df %>%\n",
    "count(population) %>%\n",
    "arrange(-n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here, \"NA\" has the sample-QC passed related individuals who are not included in our semi-related population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 8 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>population</th><th scope=col>UKBB</th><th scope=col>UKBL</th><th scope=col>w_exome</th><th scope=col>wo_exome</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>white_british    </td><td>300094</td><td>37035</td><td>137920</td><td>199209</td><td>337129</td></tr>\n",
       "\t<tr><td>related          </td><td> 39873</td><td> 4759</td><td> 18100</td><td> 26532</td><td> 44632</td></tr>\n",
       "\t<tr><td>others           </td><td> 26594</td><td> 2062</td><td> 12049</td><td> 16607</td><td> 28656</td></tr>\n",
       "\t<tr><td>non_british_white</td><td> 22406</td><td> 2499</td><td> 10432</td><td> 14473</td><td> 24905</td></tr>\n",
       "\t<tr><td>s_asian          </td><td>  7831</td><td>    0</td><td>  3541</td><td>  4290</td><td>  7831</td></tr>\n",
       "\t<tr><td>african          </td><td>  6497</td><td>    0</td><td>  2716</td><td>  3781</td><td>  6497</td></tr>\n",
       "\t<tr><td>DO_NOT_PASS_SQC  </td><td>  1805</td><td>  126</td><td>   707</td><td>  1224</td><td>  1931</td></tr>\n",
       "\t<tr><td>e_asian          </td><td>  1704</td><td>    0</td><td>   725</td><td>   979</td><td>  1704</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 8 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " population & UKBB & UKBL & w\\_exome & wo\\_exome & n\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t white\\_british     & 300094 & 37035 & 137920 & 199209 & 337129\\\\\n",
       "\t related           &  39873 &  4759 &  18100 &  26532 &  44632\\\\\n",
       "\t others            &  26594 &  2062 &  12049 &  16607 &  28656\\\\\n",
       "\t non\\_british\\_white &  22406 &  2499 &  10432 &  14473 &  24905\\\\\n",
       "\t s\\_asian           &   7831 &     0 &   3541 &   4290 &   7831\\\\\n",
       "\t african           &   6497 &     0 &   2716 &   3781 &   6497\\\\\n",
       "\t DO\\_NOT\\_PASS\\_SQC   &   1805 &   126 &    707 &   1224 &   1931\\\\\n",
       "\t e\\_asian           &   1704 &     0 &    725 &    979 &   1704\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 8 × 6\n",
       "\n",
       "| population &lt;chr&gt; | UKBB &lt;dbl&gt; | UKBL &lt;dbl&gt; | w_exome &lt;dbl&gt; | wo_exome &lt;dbl&gt; | n &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| white_british     | 300094 | 37035 | 137920 | 199209 | 337129 |\n",
       "| related           |  39873 |  4759 |  18100 |  26532 |  44632 |\n",
       "| others            |  26594 |  2062 |  12049 |  16607 |  28656 |\n",
       "| non_british_white |  22406 |  2499 |  10432 |  14473 |  24905 |\n",
       "| s_asian           |   7831 |     0 |   3541 |   4290 |   7831 |\n",
       "| african           |   6497 |     0 |   2716 |   3781 |   6497 |\n",
       "| DO_NOT_PASS_SQC   |   1805 |   126 |    707 |   1224 |   1931 |\n",
       "| e_asian           |   1704 |     0 |    725 |    979 |   1704 |\n",
       "\n"
      ],
      "text/plain": [
       "  population        UKBB   UKBL  w_exome wo_exome n     \n",
       "1 white_british     300094 37035 137920  199209   337129\n",
       "2 related            39873  4759  18100   26532    44632\n",
       "3 others             26594  2062  12049   16607    28656\n",
       "4 non_british_white  22406  2499  10432   14473    24905\n",
       "5 s_asian             7831     0   3541    4290     7831\n",
       "6 african             6497     0   2716    3781     6497\n",
       "7 DO_NOT_PASS_SQC     1805   126    707    1224     1931\n",
       "8 e_asian             1704     0    725     979     1704"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_refined_pop_df %>% show_population_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The joint frequency of our population definition and the self-reported ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 8 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>population</th><th scope=col>Asian or Asian British</th><th scope=col>Black or Black British</th><th scope=col>Chinese</th><th scope=col>Do not know</th><th scope=col>Mixed</th><th scope=col>Other ethnic group</th><th scope=col>Prefer not to answer</th><th scope=col>White</th><th scope=col>&lt;NA&gt;</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>african          </td><td>   0</td><td>6305</td><td>   0</td><td>  8</td><td>   0</td><td>   0</td><td> 120</td><td>     0</td><td> 64</td></tr>\n",
       "\t<tr><td>DO_NOT_PASS_SQC  </td><td>  68</td><td>  12</td><td>   1</td><td>  0</td><td>   5</td><td>   7</td><td>   5</td><td>  1710</td><td>123</td></tr>\n",
       "\t<tr><td>e_asian          </td><td> 279</td><td>   0</td><td>1367</td><td>  4</td><td>   0</td><td>   0</td><td>  20</td><td>     0</td><td> 34</td></tr>\n",
       "\t<tr><td>non_british_white</td><td>   0</td><td>   0</td><td>   0</td><td>  0</td><td>   0</td><td>   0</td><td>   0</td><td> 24895</td><td> 10</td></tr>\n",
       "\t<tr><td>others           </td><td> 676</td><td> 680</td><td>  52</td><td>152</td><td>2539</td><td>4007</td><td>1084</td><td> 18661</td><td>805</td></tr>\n",
       "\t<tr><td>related          </td><td> 386</td><td> 281</td><td>  44</td><td>  8</td><td> 123</td><td> 162</td><td> 134</td><td> 43387</td><td>107</td></tr>\n",
       "\t<tr><td>s_asian          </td><td>7617</td><td>   0</td><td>   0</td><td> 17</td><td>   0</td><td>   0</td><td>  83</td><td>     0</td><td>114</td></tr>\n",
       "\t<tr><td>white_british    </td><td>   0</td><td>   0</td><td>   0</td><td>  0</td><td>   0</td><td>   0</td><td>   0</td><td>336987</td><td>142</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 8 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " population & Asian or Asian British & Black or Black British & Chinese & Do not know & Mixed & Other ethnic group & Prefer not to answer & White & <NA>\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t african           &    0 & 6305 &    0 &   8 &    0 &    0 &  120 &      0 &  64\\\\\n",
       "\t DO\\_NOT\\_PASS\\_SQC   &   68 &   12 &    1 &   0 &    5 &    7 &    5 &   1710 & 123\\\\\n",
       "\t e\\_asian           &  279 &    0 & 1367 &   4 &    0 &    0 &   20 &      0 &  34\\\\\n",
       "\t non\\_british\\_white &    0 &    0 &    0 &   0 &    0 &    0 &    0 &  24895 &  10\\\\\n",
       "\t others            &  676 &  680 &   52 & 152 & 2539 & 4007 & 1084 &  18661 & 805\\\\\n",
       "\t related           &  386 &  281 &   44 &   8 &  123 &  162 &  134 &  43387 & 107\\\\\n",
       "\t s\\_asian           & 7617 &    0 &    0 &  17 &    0 &    0 &   83 &      0 & 114\\\\\n",
       "\t white\\_british     &    0 &    0 &    0 &   0 &    0 &    0 &    0 & 336987 & 142\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 8 × 10\n",
       "\n",
       "| population &lt;chr&gt; | Asian or Asian British &lt;dbl&gt; | Black or Black British &lt;dbl&gt; | Chinese &lt;dbl&gt; | Do not know &lt;dbl&gt; | Mixed &lt;dbl&gt; | Other ethnic group &lt;dbl&gt; | Prefer not to answer &lt;dbl&gt; | White &lt;dbl&gt; | &lt;NA&gt; &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| african           |    0 | 6305 |    0 |   8 |    0 |    0 |  120 |      0 |  64 |\n",
       "| DO_NOT_PASS_SQC   |   68 |   12 |    1 |   0 |    5 |    7 |    5 |   1710 | 123 |\n",
       "| e_asian           |  279 |    0 | 1367 |   4 |    0 |    0 |   20 |      0 |  34 |\n",
       "| non_british_white |    0 |    0 |    0 |   0 |    0 |    0 |    0 |  24895 |  10 |\n",
       "| others            |  676 |  680 |   52 | 152 | 2539 | 4007 | 1084 |  18661 | 805 |\n",
       "| related           |  386 |  281 |   44 |   8 |  123 |  162 |  134 |  43387 | 107 |\n",
       "| s_asian           | 7617 |    0 |    0 |  17 |    0 |    0 |   83 |      0 | 114 |\n",
       "| white_british     |    0 |    0 |    0 |   0 |    0 |    0 |    0 | 336987 | 142 |\n",
       "\n"
      ],
      "text/plain": [
       "  population        Asian or Asian British Black or Black British Chinese\n",
       "1 african              0                   6305                      0   \n",
       "2 DO_NOT_PASS_SQC     68                     12                      1   \n",
       "3 e_asian            279                      0                   1367   \n",
       "4 non_british_white    0                      0                      0   \n",
       "5 others             676                    680                     52   \n",
       "6 related            386                    281                     44   \n",
       "7 s_asian           7617                      0                      0   \n",
       "8 white_british        0                      0                      0   \n",
       "  Do not know Mixed Other ethnic group Prefer not to answer White  <NA>\n",
       "1   8            0     0                120                      0  64 \n",
       "2   0            5     7                  5                   1710 123 \n",
       "3   4            0     0                 20                      0  34 \n",
       "4   0            0     0                  0                  24895  10 \n",
       "5 152         2539  4007               1084                  18661 805 \n",
       "6   8          123   162                134                  43387 107 \n",
       "7  17            0     0                 83                      0 114 \n",
       "8   0            0     0                  0                 336987 142 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_refined_pop_df  %>% \n",
    "drop_na(population) %>%\n",
    "count(population, f21000_top_label) %>% \n",
    "spread(f21000_top_label, n, fill= 0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6-5) Prepare and join GWAS covariates and training/val/test split info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "yob_df <- fread(\n",
    "    cmd=paste(\n",
    "        'cat', \n",
    "        file_names$covar_yob_tab,\n",
    "        '| cut -f1,11'\n",
    "    ),\n",
    "    colClasses=c('character', 'numeric')\n",
    ") %>%\n",
    "rename(\n",
    "    'IID' = 'f.eid',\n",
    "    'BirthYear' = 'f.34.0.0'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "CNV_df <- fread(\n",
    "    file_names$covar_CNV_f, select=c('FID', 'IID', 'N_CNV', 'LEN_CNV'),\n",
    "    colClasses=c('FID'='character', 'IID'='character')\n",
    ")\n",
    "\n",
    "split_df <- lapply(\n",
    "    c('train', 'valid', 'test'),\n",
    "    function(s){\n",
    "        fread(\n",
    "            file.path(\n",
    "                file_names$covar_split_dir, \n",
    "                sprintf('ukb24983_white_british_%s.phe', s)\n",
    "            ), \n",
    "            col.names=c('FID', 'IID'),\n",
    "            colClasses='character'\n",
    "        ) %>%\n",
    "        mutate(split = s)\n",
    "    }\n",
    ") %>%\n",
    "bind_rows() %>%\n",
    "mutate(\n",
    "    split = if_else(split == 'valid', 'val', split)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_sqc_all <- master_sqc_refined_pop_df %>%\n",
    "left_join(yob_df, by='IID') %>%\n",
    "left_join(\n",
    "    fread(file_names$covar_bmi_f, data.table=F, sep='\\t', col.names=c('FID', 'IID', 'BMI'), colClasses = c('character', 'character', 'numeric')),\n",
    "    by=c('FID', 'IID')\n",
    ") %>%\n",
    "left_join(\n",
    "    fread(file_names$covar_age_f, data.table=F, sep='\\t', colClasses=c('IID'='character')), by='IID'\n",
    ") %>%\n",
    "left_join(CNV_df, by=c('FID', 'IID')) %>% \n",
    "rename('sex_str' = 'sex') %>%\n",
    "mutate(\n",
    "    age = if_else(\n",
    "        is.na(BirthYear),\n",
    "        2017 - median(yob_df$BirthYear, na.rm = T),\n",
    "        2017 - BirthYear        \n",
    "    ),\n",
    "    sex   = if_else(sex_str == 'male', 1, if_else(sex_str == 'female', 0, NULL)),\n",
    "    Array = if_else(genotyping_array == 'UKBB', 1, 0)\n",
    ") %>%\n",
    "left_join(split_df, by=c('FID', 'IID')) %>%\n",
    "mutate(\n",
    "    split = if_else(population == 'white_british', split, population)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 11 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>split</th><th scope=col>population</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>train            </td><td>white_british    </td><td>235991</td></tr>\n",
       "\t<tr><td>test             </td><td>white_british    </td><td> 67425</td></tr>\n",
       "\t<tr><td>related          </td><td>related          </td><td> 44632</td></tr>\n",
       "\t<tr><td>NA               </td><td>NA               </td><td> 35092</td></tr>\n",
       "\t<tr><td>val              </td><td>white_british    </td><td> 33713</td></tr>\n",
       "\t<tr><td>others           </td><td>others           </td><td> 28656</td></tr>\n",
       "\t<tr><td>non_british_white</td><td>non_british_white</td><td> 24905</td></tr>\n",
       "\t<tr><td>s_asian          </td><td>s_asian          </td><td>  7831</td></tr>\n",
       "\t<tr><td>african          </td><td>african          </td><td>  6497</td></tr>\n",
       "\t<tr><td>DO_NOT_PASS_SQC  </td><td>DO_NOT_PASS_SQC  </td><td>  1931</td></tr>\n",
       "\t<tr><td>e_asian          </td><td>e_asian          </td><td>  1704</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 11 × 3\n",
       "\\begin{tabular}{lll}\n",
       " split & population & n\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t train             & white\\_british     & 235991\\\\\n",
       "\t test              & white\\_british     &  67425\\\\\n",
       "\t related           & related           &  44632\\\\\n",
       "\t NA                & NA                &  35092\\\\\n",
       "\t val               & white\\_british     &  33713\\\\\n",
       "\t others            & others            &  28656\\\\\n",
       "\t non\\_british\\_white & non\\_british\\_white &  24905\\\\\n",
       "\t s\\_asian           & s\\_asian           &   7831\\\\\n",
       "\t african           & african           &   6497\\\\\n",
       "\t DO\\_NOT\\_PASS\\_SQC   & DO\\_NOT\\_PASS\\_SQC   &   1931\\\\\n",
       "\t e\\_asian           & e\\_asian           &   1704\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 11 × 3\n",
       "\n",
       "| split &lt;chr&gt; | population &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| train             | white_british     | 235991 |\n",
       "| test              | white_british     |  67425 |\n",
       "| related           | related           |  44632 |\n",
       "| NA                | NA                |  35092 |\n",
       "| val               | white_british     |  33713 |\n",
       "| others            | others            |  28656 |\n",
       "| non_british_white | non_british_white |  24905 |\n",
       "| s_asian           | s_asian           |   7831 |\n",
       "| african           | african           |   6497 |\n",
       "| DO_NOT_PASS_SQC   | DO_NOT_PASS_SQC   |   1931 |\n",
       "| e_asian           | e_asian           |   1704 |\n",
       "\n"
      ],
      "text/plain": [
       "   split             population        n     \n",
       "1  train             white_british     235991\n",
       "2  test              white_british      67425\n",
       "3  related           related            44632\n",
       "4  NA                NA                 35092\n",
       "5  val               white_british      33713\n",
       "6  others            others             28656\n",
       "7  non_british_white non_british_white  24905\n",
       "8  s_asian           s_asian             7831\n",
       "9  african           african             6497\n",
       "10 DO_NOT_PASS_SQC   DO_NOT_PASS_SQC     1931\n",
       "11 e_asian           e_asian             1704"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_all %>%\n",
    "count(split, population) %>%\n",
    "arrange(-n)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'FID'</li><li>'IID'</li><li>'sex_code'</li><li>'batch'</li><li>'sex_str'</li><li>'affymetrix_field_1'</li><li>'affymetrix_field_2'</li><li>'genotyping_array'</li><li>'Batch'</li><li>'Plate_Name'</li><li>'Well'</li><li>'Cluster_CR'</li><li>'dQC'</li><li>'Internal_Pico_ng_uL'</li><li>'Submitted_Gender'</li><li>'Inferred_Gender'</li><li>'X_intensity'</li><li>'Y_intensity'</li><li>'Submitted_Plate_Name'</li><li>'Submitted_Well'</li><li>'sample_qc_missing_rate'</li><li>'heterozygosity'</li><li>'heterozygosity_pc_corrected'</li><li>'het_missing_outliers'</li><li>'putative_sex_chromosome_aneuploidy'</li><li>'in_kinship_table'</li><li>'excluded_from_kinship_inference'</li><li>'excess_relatives'</li><li>'in_white_British_ancestry_subset'</li><li>'used_in_pca_calculation'</li><li>'Global_PC1'</li><li>'Global_PC2'</li><li>'Global_PC3'</li><li>'Global_PC4'</li><li>'Global_PC5'</li><li>'Global_PC6'</li><li>'Global_PC7'</li><li>'Global_PC8'</li><li>'Global_PC9'</li><li>'Global_PC10'</li><li>'Global_PC11'</li><li>'Global_PC12'</li><li>'Global_PC13'</li><li>'Global_PC14'</li><li>'Global_PC15'</li><li>'Global_PC16'</li><li>'Global_PC17'</li><li>'Global_PC18'</li><li>'Global_PC19'</li><li>'Global_PC20'</li><li>'Global_PC21'</li><li>'Global_PC22'</li><li>'Global_PC23'</li><li>'Global_PC24'</li><li>'Global_PC25'</li><li>'Global_PC26'</li><li>'Global_PC27'</li><li>'Global_PC28'</li><li>'Global_PC29'</li><li>'Global_PC30'</li><li>'Global_PC31'</li><li>'Global_PC32'</li><li>'Global_PC33'</li><li>'Global_PC34'</li><li>'Global_PC35'</li><li>'Global_PC36'</li><li>'Global_PC37'</li><li>'Global_PC38'</li><li>'Global_PC39'</li><li>'Global_PC40'</li><li>'in_Phasing_Input_chr1_22'</li><li>'in_Phasing_Input_chrX'</li><li>'in_Phasing_Input_chrXY'</li><li>'fam_order'</li><li>'has_exome'</li><li>'in_remove_file'</li><li>'f21000'</li><li>'f21000_sub_label'</li><li>'f21000_top_label'</li><li>'f21000_0'</li><li>'f21000_1'</li><li>'f21000_2'</li><li>'pass_QC_filter'</li><li>'pass_filter'</li><li>'self_reported_NBW'</li><li>'self_reported_White'</li><li>'self_reported_Asian'</li><li>'self_reported_Black'</li><li>'self_reported_Mixed_or_Other'</li><li>'PC1'</li><li>'PC2'</li><li>'PC3'</li><li>'PC4'</li><li>'PC5'</li><li>'PC6'</li><li>'PC7'</li><li>'PC8'</li><li>'PC9'</li><li>'PC10'</li><li>'PC11'</li><li>'PC12'</li><li>'PC13'</li><li>'PC14'</li><li>'PC15'</li><li>'PC16'</li><li>'PC17'</li><li>'PC18'</li><li>'PC19'</li><li>'PC20'</li><li>'PC21'</li><li>'PC22'</li><li>'PC23'</li><li>'PC24'</li><li>'PC25'</li><li>'PC26'</li><li>'PC27'</li><li>'PC28'</li><li>'PC29'</li><li>'PC30'</li><li>'PC31'</li><li>'PC32'</li><li>'PC33'</li><li>'PC34'</li><li>'PC35'</li><li>'PC36'</li><li>'PC37'</li><li>'PC38'</li><li>'PC39'</li><li>'PC40'</li><li>'population'</li><li>'BirthYear'</li><li>'BMI'</li><li>'dob'</li><li>'age0'</li><li>'age1'</li><li>'age2'</li><li>'age3'</li><li>'N_CNV'</li><li>'LEN_CNV'</li><li>'age'</li><li>'sex'</li><li>'Array'</li><li>'split'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'FID'\n",
       "\\item 'IID'\n",
       "\\item 'sex\\_code'\n",
       "\\item 'batch'\n",
       "\\item 'sex\\_str'\n",
       "\\item 'affymetrix\\_field\\_1'\n",
       "\\item 'affymetrix\\_field\\_2'\n",
       "\\item 'genotyping\\_array'\n",
       "\\item 'Batch'\n",
       "\\item 'Plate\\_Name'\n",
       "\\item 'Well'\n",
       "\\item 'Cluster\\_CR'\n",
       "\\item 'dQC'\n",
       "\\item 'Internal\\_Pico\\_ng\\_uL'\n",
       "\\item 'Submitted\\_Gender'\n",
       "\\item 'Inferred\\_Gender'\n",
       "\\item 'X\\_intensity'\n",
       "\\item 'Y\\_intensity'\n",
       "\\item 'Submitted\\_Plate\\_Name'\n",
       "\\item 'Submitted\\_Well'\n",
       "\\item 'sample\\_qc\\_missing\\_rate'\n",
       "\\item 'heterozygosity'\n",
       "\\item 'heterozygosity\\_pc\\_corrected'\n",
       "\\item 'het\\_missing\\_outliers'\n",
       "\\item 'putative\\_sex\\_chromosome\\_aneuploidy'\n",
       "\\item 'in\\_kinship\\_table'\n",
       "\\item 'excluded\\_from\\_kinship\\_inference'\n",
       "\\item 'excess\\_relatives'\n",
       "\\item 'in\\_white\\_British\\_ancestry\\_subset'\n",
       "\\item 'used\\_in\\_pca\\_calculation'\n",
       "\\item 'Global\\_PC1'\n",
       "\\item 'Global\\_PC2'\n",
       "\\item 'Global\\_PC3'\n",
       "\\item 'Global\\_PC4'\n",
       "\\item 'Global\\_PC5'\n",
       "\\item 'Global\\_PC6'\n",
       "\\item 'Global\\_PC7'\n",
       "\\item 'Global\\_PC8'\n",
       "\\item 'Global\\_PC9'\n",
       "\\item 'Global\\_PC10'\n",
       "\\item 'Global\\_PC11'\n",
       "\\item 'Global\\_PC12'\n",
       "\\item 'Global\\_PC13'\n",
       "\\item 'Global\\_PC14'\n",
       "\\item 'Global\\_PC15'\n",
       "\\item 'Global\\_PC16'\n",
       "\\item 'Global\\_PC17'\n",
       "\\item 'Global\\_PC18'\n",
       "\\item 'Global\\_PC19'\n",
       "\\item 'Global\\_PC20'\n",
       "\\item 'Global\\_PC21'\n",
       "\\item 'Global\\_PC22'\n",
       "\\item 'Global\\_PC23'\n",
       "\\item 'Global\\_PC24'\n",
       "\\item 'Global\\_PC25'\n",
       "\\item 'Global\\_PC26'\n",
       "\\item 'Global\\_PC27'\n",
       "\\item 'Global\\_PC28'\n",
       "\\item 'Global\\_PC29'\n",
       "\\item 'Global\\_PC30'\n",
       "\\item 'Global\\_PC31'\n",
       "\\item 'Global\\_PC32'\n",
       "\\item 'Global\\_PC33'\n",
       "\\item 'Global\\_PC34'\n",
       "\\item 'Global\\_PC35'\n",
       "\\item 'Global\\_PC36'\n",
       "\\item 'Global\\_PC37'\n",
       "\\item 'Global\\_PC38'\n",
       "\\item 'Global\\_PC39'\n",
       "\\item 'Global\\_PC40'\n",
       "\\item 'in\\_Phasing\\_Input\\_chr1\\_22'\n",
       "\\item 'in\\_Phasing\\_Input\\_chrX'\n",
       "\\item 'in\\_Phasing\\_Input\\_chrXY'\n",
       "\\item 'fam\\_order'\n",
       "\\item 'has\\_exome'\n",
       "\\item 'in\\_remove\\_file'\n",
       "\\item 'f21000'\n",
       "\\item 'f21000\\_sub\\_label'\n",
       "\\item 'f21000\\_top\\_label'\n",
       "\\item 'f21000\\_0'\n",
       "\\item 'f21000\\_1'\n",
       "\\item 'f21000\\_2'\n",
       "\\item 'pass\\_QC\\_filter'\n",
       "\\item 'pass\\_filter'\n",
       "\\item 'self\\_reported\\_NBW'\n",
       "\\item 'self\\_reported\\_White'\n",
       "\\item 'self\\_reported\\_Asian'\n",
       "\\item 'self\\_reported\\_Black'\n",
       "\\item 'self\\_reported\\_Mixed\\_or\\_Other'\n",
       "\\item 'PC1'\n",
       "\\item 'PC2'\n",
       "\\item 'PC3'\n",
       "\\item 'PC4'\n",
       "\\item 'PC5'\n",
       "\\item 'PC6'\n",
       "\\item 'PC7'\n",
       "\\item 'PC8'\n",
       "\\item 'PC9'\n",
       "\\item 'PC10'\n",
       "\\item 'PC11'\n",
       "\\item 'PC12'\n",
       "\\item 'PC13'\n",
       "\\item 'PC14'\n",
       "\\item 'PC15'\n",
       "\\item 'PC16'\n",
       "\\item 'PC17'\n",
       "\\item 'PC18'\n",
       "\\item 'PC19'\n",
       "\\item 'PC20'\n",
       "\\item 'PC21'\n",
       "\\item 'PC22'\n",
       "\\item 'PC23'\n",
       "\\item 'PC24'\n",
       "\\item 'PC25'\n",
       "\\item 'PC26'\n",
       "\\item 'PC27'\n",
       "\\item 'PC28'\n",
       "\\item 'PC29'\n",
       "\\item 'PC30'\n",
       "\\item 'PC31'\n",
       "\\item 'PC32'\n",
       "\\item 'PC33'\n",
       "\\item 'PC34'\n",
       "\\item 'PC35'\n",
       "\\item 'PC36'\n",
       "\\item 'PC37'\n",
       "\\item 'PC38'\n",
       "\\item 'PC39'\n",
       "\\item 'PC40'\n",
       "\\item 'population'\n",
       "\\item 'BirthYear'\n",
       "\\item 'BMI'\n",
       "\\item 'dob'\n",
       "\\item 'age0'\n",
       "\\item 'age1'\n",
       "\\item 'age2'\n",
       "\\item 'age3'\n",
       "\\item 'N\\_CNV'\n",
       "\\item 'LEN\\_CNV'\n",
       "\\item 'age'\n",
       "\\item 'sex'\n",
       "\\item 'Array'\n",
       "\\item 'split'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'FID'\n",
       "2. 'IID'\n",
       "3. 'sex_code'\n",
       "4. 'batch'\n",
       "5. 'sex_str'\n",
       "6. 'affymetrix_field_1'\n",
       "7. 'affymetrix_field_2'\n",
       "8. 'genotyping_array'\n",
       "9. 'Batch'\n",
       "10. 'Plate_Name'\n",
       "11. 'Well'\n",
       "12. 'Cluster_CR'\n",
       "13. 'dQC'\n",
       "14. 'Internal_Pico_ng_uL'\n",
       "15. 'Submitted_Gender'\n",
       "16. 'Inferred_Gender'\n",
       "17. 'X_intensity'\n",
       "18. 'Y_intensity'\n",
       "19. 'Submitted_Plate_Name'\n",
       "20. 'Submitted_Well'\n",
       "21. 'sample_qc_missing_rate'\n",
       "22. 'heterozygosity'\n",
       "23. 'heterozygosity_pc_corrected'\n",
       "24. 'het_missing_outliers'\n",
       "25. 'putative_sex_chromosome_aneuploidy'\n",
       "26. 'in_kinship_table'\n",
       "27. 'excluded_from_kinship_inference'\n",
       "28. 'excess_relatives'\n",
       "29. 'in_white_British_ancestry_subset'\n",
       "30. 'used_in_pca_calculation'\n",
       "31. 'Global_PC1'\n",
       "32. 'Global_PC2'\n",
       "33. 'Global_PC3'\n",
       "34. 'Global_PC4'\n",
       "35. 'Global_PC5'\n",
       "36. 'Global_PC6'\n",
       "37. 'Global_PC7'\n",
       "38. 'Global_PC8'\n",
       "39. 'Global_PC9'\n",
       "40. 'Global_PC10'\n",
       "41. 'Global_PC11'\n",
       "42. 'Global_PC12'\n",
       "43. 'Global_PC13'\n",
       "44. 'Global_PC14'\n",
       "45. 'Global_PC15'\n",
       "46. 'Global_PC16'\n",
       "47. 'Global_PC17'\n",
       "48. 'Global_PC18'\n",
       "49. 'Global_PC19'\n",
       "50. 'Global_PC20'\n",
       "51. 'Global_PC21'\n",
       "52. 'Global_PC22'\n",
       "53. 'Global_PC23'\n",
       "54. 'Global_PC24'\n",
       "55. 'Global_PC25'\n",
       "56. 'Global_PC26'\n",
       "57. 'Global_PC27'\n",
       "58. 'Global_PC28'\n",
       "59. 'Global_PC29'\n",
       "60. 'Global_PC30'\n",
       "61. 'Global_PC31'\n",
       "62. 'Global_PC32'\n",
       "63. 'Global_PC33'\n",
       "64. 'Global_PC34'\n",
       "65. 'Global_PC35'\n",
       "66. 'Global_PC36'\n",
       "67. 'Global_PC37'\n",
       "68. 'Global_PC38'\n",
       "69. 'Global_PC39'\n",
       "70. 'Global_PC40'\n",
       "71. 'in_Phasing_Input_chr1_22'\n",
       "72. 'in_Phasing_Input_chrX'\n",
       "73. 'in_Phasing_Input_chrXY'\n",
       "74. 'fam_order'\n",
       "75. 'has_exome'\n",
       "76. 'in_remove_file'\n",
       "77. 'f21000'\n",
       "78. 'f21000_sub_label'\n",
       "79. 'f21000_top_label'\n",
       "80. 'f21000_0'\n",
       "81. 'f21000_1'\n",
       "82. 'f21000_2'\n",
       "83. 'pass_QC_filter'\n",
       "84. 'pass_filter'\n",
       "85. 'self_reported_NBW'\n",
       "86. 'self_reported_White'\n",
       "87. 'self_reported_Asian'\n",
       "88. 'self_reported_Black'\n",
       "89. 'self_reported_Mixed_or_Other'\n",
       "90. 'PC1'\n",
       "91. 'PC2'\n",
       "92. 'PC3'\n",
       "93. 'PC4'\n",
       "94. 'PC5'\n",
       "95. 'PC6'\n",
       "96. 'PC7'\n",
       "97. 'PC8'\n",
       "98. 'PC9'\n",
       "99. 'PC10'\n",
       "100. 'PC11'\n",
       "101. 'PC12'\n",
       "102. 'PC13'\n",
       "103. 'PC14'\n",
       "104. 'PC15'\n",
       "105. 'PC16'\n",
       "106. 'PC17'\n",
       "107. 'PC18'\n",
       "108. 'PC19'\n",
       "109. 'PC20'\n",
       "110. 'PC21'\n",
       "111. 'PC22'\n",
       "112. 'PC23'\n",
       "113. 'PC24'\n",
       "114. 'PC25'\n",
       "115. 'PC26'\n",
       "116. 'PC27'\n",
       "117. 'PC28'\n",
       "118. 'PC29'\n",
       "119. 'PC30'\n",
       "120. 'PC31'\n",
       "121. 'PC32'\n",
       "122. 'PC33'\n",
       "123. 'PC34'\n",
       "124. 'PC35'\n",
       "125. 'PC36'\n",
       "126. 'PC37'\n",
       "127. 'PC38'\n",
       "128. 'PC39'\n",
       "129. 'PC40'\n",
       "130. 'population'\n",
       "131. 'BirthYear'\n",
       "132. 'BMI'\n",
       "133. 'dob'\n",
       "134. 'age0'\n",
       "135. 'age1'\n",
       "136. 'age2'\n",
       "137. 'age3'\n",
       "138. 'N_CNV'\n",
       "139. 'LEN_CNV'\n",
       "140. 'age'\n",
       "141. 'sex'\n",
       "142. 'Array'\n",
       "143. 'split'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"FID\"                                \"IID\"                               \n",
       "  [3] \"sex_code\"                           \"batch\"                             \n",
       "  [5] \"sex_str\"                            \"affymetrix_field_1\"                \n",
       "  [7] \"affymetrix_field_2\"                 \"genotyping_array\"                  \n",
       "  [9] \"Batch\"                              \"Plate_Name\"                        \n",
       " [11] \"Well\"                               \"Cluster_CR\"                        \n",
       " [13] \"dQC\"                                \"Internal_Pico_ng_uL\"               \n",
       " [15] \"Submitted_Gender\"                   \"Inferred_Gender\"                   \n",
       " [17] \"X_intensity\"                        \"Y_intensity\"                       \n",
       " [19] \"Submitted_Plate_Name\"               \"Submitted_Well\"                    \n",
       " [21] \"sample_qc_missing_rate\"             \"heterozygosity\"                    \n",
       " [23] \"heterozygosity_pc_corrected\"        \"het_missing_outliers\"              \n",
       " [25] \"putative_sex_chromosome_aneuploidy\" \"in_kinship_table\"                  \n",
       " [27] \"excluded_from_kinship_inference\"    \"excess_relatives\"                  \n",
       " [29] \"in_white_British_ancestry_subset\"   \"used_in_pca_calculation\"           \n",
       " [31] \"Global_PC1\"                         \"Global_PC2\"                        \n",
       " [33] \"Global_PC3\"                         \"Global_PC4\"                        \n",
       " [35] \"Global_PC5\"                         \"Global_PC6\"                        \n",
       " [37] \"Global_PC7\"                         \"Global_PC8\"                        \n",
       " [39] \"Global_PC9\"                         \"Global_PC10\"                       \n",
       " [41] \"Global_PC11\"                        \"Global_PC12\"                       \n",
       " [43] \"Global_PC13\"                        \"Global_PC14\"                       \n",
       " [45] \"Global_PC15\"                        \"Global_PC16\"                       \n",
       " [47] \"Global_PC17\"                        \"Global_PC18\"                       \n",
       " [49] \"Global_PC19\"                        \"Global_PC20\"                       \n",
       " [51] \"Global_PC21\"                        \"Global_PC22\"                       \n",
       " [53] \"Global_PC23\"                        \"Global_PC24\"                       \n",
       " [55] \"Global_PC25\"                        \"Global_PC26\"                       \n",
       " [57] \"Global_PC27\"                        \"Global_PC28\"                       \n",
       " [59] \"Global_PC29\"                        \"Global_PC30\"                       \n",
       " [61] \"Global_PC31\"                        \"Global_PC32\"                       \n",
       " [63] \"Global_PC33\"                        \"Global_PC34\"                       \n",
       " [65] \"Global_PC35\"                        \"Global_PC36\"                       \n",
       " [67] \"Global_PC37\"                        \"Global_PC38\"                       \n",
       " [69] \"Global_PC39\"                        \"Global_PC40\"                       \n",
       " [71] \"in_Phasing_Input_chr1_22\"           \"in_Phasing_Input_chrX\"             \n",
       " [73] \"in_Phasing_Input_chrXY\"             \"fam_order\"                         \n",
       " [75] \"has_exome\"                          \"in_remove_file\"                    \n",
       " [77] \"f21000\"                             \"f21000_sub_label\"                  \n",
       " [79] \"f21000_top_label\"                   \"f21000_0\"                          \n",
       " [81] \"f21000_1\"                           \"f21000_2\"                          \n",
       " [83] \"pass_QC_filter\"                     \"pass_filter\"                       \n",
       " [85] \"self_reported_NBW\"                  \"self_reported_White\"               \n",
       " [87] \"self_reported_Asian\"                \"self_reported_Black\"               \n",
       " [89] \"self_reported_Mixed_or_Other\"       \"PC1\"                               \n",
       " [91] \"PC2\"                                \"PC3\"                               \n",
       " [93] \"PC4\"                                \"PC5\"                               \n",
       " [95] \"PC6\"                                \"PC7\"                               \n",
       " [97] \"PC8\"                                \"PC9\"                               \n",
       " [99] \"PC10\"                               \"PC11\"                              \n",
       "[101] \"PC12\"                               \"PC13\"                              \n",
       "[103] \"PC14\"                               \"PC15\"                              \n",
       "[105] \"PC16\"                               \"PC17\"                              \n",
       "[107] \"PC18\"                               \"PC19\"                              \n",
       "[109] \"PC20\"                               \"PC21\"                              \n",
       "[111] \"PC22\"                               \"PC23\"                              \n",
       "[113] \"PC24\"                               \"PC25\"                              \n",
       "[115] \"PC26\"                               \"PC27\"                              \n",
       "[117] \"PC28\"                               \"PC29\"                              \n",
       "[119] \"PC30\"                               \"PC31\"                              \n",
       "[121] \"PC32\"                               \"PC33\"                              \n",
       "[123] \"PC34\"                               \"PC35\"                              \n",
       "[125] \"PC36\"                               \"PC37\"                              \n",
       "[127] \"PC38\"                               \"PC39\"                              \n",
       "[129] \"PC40\"                               \"population\"                        \n",
       "[131] \"BirthYear\"                          \"BMI\"                               \n",
       "[133] \"dob\"                                \"age0\"                              \n",
       "[135] \"age1\"                               \"age2\"                              \n",
       "[137] \"age3\"                               \"N_CNV\"                             \n",
       "[139] \"LEN_CNV\"                            \"age\"                               \n",
       "[141] \"sex\"                                \"Array\"                             \n",
       "[143] \"split\"                             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "master_sqc_all %>% colnames()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6-6) write the results into files\n",
    "\n",
    "### master SQC file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_sqc_all %>%\n",
    "rename('#FID' = 'FID') %>%\n",
    "fwrite(file_names$master_sqc_file, sep='\\t', na = \"NA\", quote=F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### keep file for each population\n",
    "\n",
    "In this version, we don't support the exclude phe file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "for (pop in c('white_british', 'non_british_white', 'african', 's_asian', 'e_asian', 'related', 'others')){\n",
    "    master_sqc_all %>% filter(population == pop) %>% select(FID, IID) %>%\n",
    "    fwrite(file.path(dirname(file_names$master_sqc_file), paste0('ukb24983_', pop, '.phe')), sep='\\t', col.names = F)\n",
    "\n",
    "#     master_sqc_all %>% filter(population != pop) %>% select(FID, IID) %>%\n",
    "#     fwrite(file.path(dirname(file_names$master_sqc_file), paste0('ukb24983_', pop, '.exclude.phe')), sep='\\t', col.names = F)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GWAS covar file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_covar_df <- master_sqc_all %>% \n",
    "select(\n",
    "    FID, IID, \n",
    "    population, split, \n",
    "    age, age0, age1, age2, age3, \n",
    "    sex, BMI, N_CNV, LEN_CNV, \n",
    "    Array, paste0('PC', 1:40), paste0('Global_PC', 1:40)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_covar_df %>%\n",
    "rename('#FID' = 'FID') %>%\n",
    "fwrite(file_names$GWAS_covar_file, sep='\\t', na = \"NA\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
