{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages({\n",
    "    library(tidyverse)\n",
    "    library(data.table)\n",
    "}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input\n",
    "pvar_f <- '/oak/stanford/groups/mrivas/ukbb24983/exome/pgen/oqfe_2020/ukb24983_exomeOQFE.pvar.zst'\n",
    "data_d <- '/scratch/groups/mrivas/ukbb24983/exome/annotation/20201025_exome_oqfe_2020'\n",
    "af_hwe_f   <- file.path(data_d, 'ukb24983_exomeOQFE.afreq_hwe.20201025.pvar.zst')\n",
    "vep_f      <- file.path(data_d, 'UKBexomeOQFE.vep101.tsv.gz')\n",
    "liftOver_f <- file.path(data_d, 'UKBexomeOQFE.hg19.tsv.gz')\n",
    "vep_csq_f  <- file.path('..', 'VEP_consequence_group.tsv')\n",
    "\n",
    "# output\n",
    "annot_f         <- file.path(data_d, 'ukb24983_exomeOQFE.annotation.tsv')\n",
    "annot_compact_f <- file.path(data_d, 'ukb24983_exomeOQFE.annotation.compact.tsv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "compact_fields <- c(\n",
    "    'CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER',\n",
    "    'Allele', 'Csq', 'Consequence', 'SYMBOL', 'Gene',\n",
    "    'f_miss', 'UKB_white_british_hwe_p', 'UKB_white_british_AF', 'UKB_AF',\n",
    "    'CHROM_hg19', 'POS_hg19', 'REF_hg19', 'ALT_hg19', 'liftOver_unmapped_reason'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "\n",
    "cat_or_zcat <- function(f){\n",
    "    ifelse(endsWith(f, '.zst'), 'zstdcat', ifelse(endsWith(f, '.gz'), 'zcat', 'cat'))\n",
    "}\n",
    "\n",
    "fread_CHROM <- function(f, select=NULL){\n",
    "    fread(cmd=paste(cat_or_zcat(f), f, \"| sed -e 's/^chr//g'\",  ' | head -n10000'), colClasses = c('#CHROM'='character'), select=select) %>% rename('CHROM'='#CHROM')\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvar_f     %>% fread_CHROM() -> pvar_df\n",
    "liftOver_f %>% fread_CHROM() -> liftOver_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "af_hwe_f %>% fread_CHROM() -> af_hwe_df\n",
    "vep_f    %>% fread_CHROM() -> vep_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "vep_csq_f %>% fread(select=c('#Consequence', 'Csq')) %>% rename('Consequence'='#Consequence') -> vep_csq_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvar_df %>%\n",
    "left_join(vep_df, by=c('CHROM', 'POS', 'REF', 'ALT')) %>%\n",
    "left_join(vep_csq_df, by='Consequence') %>%\n",
    "left_join(af_hwe_df, by=c('CHROM', 'POS', 'ID', 'REF', 'ALT')) %>%\n",
    "left_join(liftOver_df, by=c('CHROM', 'POS', 'ID', 'REF', 'ALT')) -> full_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_df %>%\n",
    "select(all_of(c(compact_fields, setdiff(colnames(full_df), compact_fields)))) %>%\n",
    "rename('#CHROM' = 'CHROM') %>%\n",
    "fwrite(annot_f, sep='\\t', na = \"NA\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_df %>%\n",
    "select(all_of(compact_fields)) %>%\n",
    "rename('#CHROM' = 'CHROM') %>%\n",
    "fwrite(annot_compact_f, sep='\\t', na = \"NA\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>9</li><li>154</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 9\n",
       "\\item 154\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 9\n",
       "2. 154\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]   9 154"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(full_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'ID'</li><li>'REF'</li><li>'ALT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'ID'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'ID'\n",
       "4. 'REF'\n",
       "5. 'ALT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"CHROM\" \"POS\"   \"ID\"    \"REF\"   \"ALT\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(pvar_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'ID'</li><li>'REF'</li><li>'ALT'</li><li>'CHROM_hg19'</li><li>'POS_hg19'</li><li>'REF_hg19'</li><li>'ALT_hg19'</li><li>'liftOver_unmapped_reason'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'ID'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\item 'CHROM\\_hg19'\n",
       "\\item 'POS\\_hg19'\n",
       "\\item 'REF\\_hg19'\n",
       "\\item 'ALT\\_hg19'\n",
       "\\item 'liftOver\\_unmapped\\_reason'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'ID'\n",
       "4. 'REF'\n",
       "5. 'ALT'\n",
       "6. 'CHROM_hg19'\n",
       "7. 'POS_hg19'\n",
       "8. 'REF_hg19'\n",
       "9. 'ALT_hg19'\n",
       "10. 'liftOver_unmapped_reason'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CHROM\"                    \"POS\"                     \n",
       " [3] \"ID\"                       \"REF\"                     \n",
       " [5] \"ALT\"                      \"CHROM_hg19\"              \n",
       " [7] \"POS_hg19\"                 \"REF_hg19\"                \n",
       " [9] \"ALT_hg19\"                 \"liftOver_unmapped_reason\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(liftOver_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'ID'</li><li>'REF'</li><li>'ALT'</li><li>'UKB_AF'</li><li>'UKB_white_british_AF'</li><li>'UKB_non_british_white_AF'</li><li>'UKB_african_AF'</li><li>'UKB_s_asian_AF'</li><li>'UKB_e_asian_AF'</li><li>'UKB_related_AF'</li><li>'UKB_others_AF'</li><li>'UKB_hwe_p'</li><li>'UKB_white_british_hwe_p'</li><li>'UKB_non_british_white_hwe_p'</li><li>'UKB_african_hwe_p'</li><li>'UKB_s_asian_hwe_p'</li><li>'UKB_e_asian_hwe_p'</li><li>'UKB_related_hwe_p'</li><li>'UKB_others_hwe_p'</li><li>'UKB_OBS_CT'</li><li>'UKB_white_british_OBS_CT'</li><li>'UKB_non_british_white_OBS_CT'</li><li>'UKB_african_OBS_CT'</li><li>'UKB_s_asian_OBS_CT'</li><li>'UKB_e_asian_OBS_CT'</li><li>'UKB_related_OBS_CT'</li><li>'UKB_others_OBS_CT'</li><li>'UKB_MISSING_CT'</li><li>'UKB_white_british_MISSING_CT'</li><li>'UKB_non_british_white_MISSING_CT'</li><li>'UKB_african_MISSING_CT'</li><li>'UKB_s_asian_MISSING_CT'</li><li>'UKB_e_asian_MISSING_CT'</li><li>'UKB_related_MISSING_CT'</li><li>'UKB_others_MISSING_CT'</li><li>'UKB_HOM_REF_CT'</li><li>'UKB_white_british_HOM_REF_CT'</li><li>'UKB_non_british_white_HOM_REF_CT'</li><li>'UKB_african_HOM_REF_CT'</li><li>'UKB_s_asian_HOM_REF_CT'</li><li>'UKB_e_asian_HOM_REF_CT'</li><li>'UKB_related_HOM_REF_CT'</li><li>'UKB_others_HOM_REF_CT'</li><li>'UKB_HET_REF_ALT_CTS'</li><li>'UKB_white_british_HET_REF_ALT_CTS'</li><li>'UKB_non_british_white_HET_REF_ALT_CTS'</li><li>'UKB_african_HET_REF_ALT_CTS'</li><li>'UKB_s_asian_HET_REF_ALT_CTS'</li><li>'UKB_e_asian_HET_REF_ALT_CTS'</li><li>'UKB_related_HET_REF_ALT_CTS'</li><li>'UKB_others_HET_REF_ALT_CTS'</li><li>'UKB_TWO_ALT_GENO_CTS'</li><li>'UKB_white_british_TWO_ALT_GENO_CTS'</li><li>'UKB_non_british_white_TWO_ALT_GENO_CTS'</li><li>'UKB_african_TWO_ALT_GENO_CTS'</li><li>'UKB_s_asian_TWO_ALT_GENO_CTS'</li><li>'UKB_e_asian_TWO_ALT_GENO_CTS'</li><li>'UKB_related_TWO_ALT_GENO_CTS'</li><li>'UKB_others_TWO_ALT_GENO_CTS'</li><li>'UKB_HAP_REF_CT'</li><li>'UKB_white_british_HAP_REF_CT'</li><li>'UKB_non_british_white_HAP_REF_CT'</li><li>'UKB_african_HAP_REF_CT'</li><li>'UKB_s_asian_HAP_REF_CT'</li><li>'UKB_e_asian_HAP_REF_CT'</li><li>'UKB_related_HAP_REF_CT'</li><li>'UKB_others_HAP_REF_CT'</li><li>'UKB_HAP_ALT_CTS'</li><li>'UKB_white_british_HAP_ALT_CTS'</li><li>'UKB_non_british_white_HAP_ALT_CTS'</li><li>'UKB_african_HAP_ALT_CTS'</li><li>'UKB_s_asian_HAP_ALT_CTS'</li><li>'UKB_e_asian_HAP_ALT_CTS'</li><li>'UKB_related_HAP_ALT_CTS'</li><li>'UKB_others_HAP_ALT_CTS'</li><li>'f_miss'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'ID'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\item 'UKB\\_AF'\n",
       "\\item 'UKB\\_white\\_british\\_AF'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_AF'\n",
       "\\item 'UKB\\_african\\_AF'\n",
       "\\item 'UKB\\_s\\_asian\\_AF'\n",
       "\\item 'UKB\\_e\\_asian\\_AF'\n",
       "\\item 'UKB\\_related\\_AF'\n",
       "\\item 'UKB\\_others\\_AF'\n",
       "\\item 'UKB\\_hwe\\_p'\n",
       "\\item 'UKB\\_white\\_british\\_hwe\\_p'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_hwe\\_p'\n",
       "\\item 'UKB\\_african\\_hwe\\_p'\n",
       "\\item 'UKB\\_s\\_asian\\_hwe\\_p'\n",
       "\\item 'UKB\\_e\\_asian\\_hwe\\_p'\n",
       "\\item 'UKB\\_related\\_hwe\\_p'\n",
       "\\item 'UKB\\_others\\_hwe\\_p'\n",
       "\\item 'UKB\\_OBS\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_OBS\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_OBS\\_CT'\n",
       "\\item 'UKB\\_african\\_OBS\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_OBS\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_OBS\\_CT'\n",
       "\\item 'UKB\\_related\\_OBS\\_CT'\n",
       "\\item 'UKB\\_others\\_OBS\\_CT'\n",
       "\\item 'UKB\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_african\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_related\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_others\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_african\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_related\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_others\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_african\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_related\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_others\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_african\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_related\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_others\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_african\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_related\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_others\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_african\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_related\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_others\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'f\\_miss'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'ID'\n",
       "4. 'REF'\n",
       "5. 'ALT'\n",
       "6. 'UKB_AF'\n",
       "7. 'UKB_white_british_AF'\n",
       "8. 'UKB_non_british_white_AF'\n",
       "9. 'UKB_african_AF'\n",
       "10. 'UKB_s_asian_AF'\n",
       "11. 'UKB_e_asian_AF'\n",
       "12. 'UKB_related_AF'\n",
       "13. 'UKB_others_AF'\n",
       "14. 'UKB_hwe_p'\n",
       "15. 'UKB_white_british_hwe_p'\n",
       "16. 'UKB_non_british_white_hwe_p'\n",
       "17. 'UKB_african_hwe_p'\n",
       "18. 'UKB_s_asian_hwe_p'\n",
       "19. 'UKB_e_asian_hwe_p'\n",
       "20. 'UKB_related_hwe_p'\n",
       "21. 'UKB_others_hwe_p'\n",
       "22. 'UKB_OBS_CT'\n",
       "23. 'UKB_white_british_OBS_CT'\n",
       "24. 'UKB_non_british_white_OBS_CT'\n",
       "25. 'UKB_african_OBS_CT'\n",
       "26. 'UKB_s_asian_OBS_CT'\n",
       "27. 'UKB_e_asian_OBS_CT'\n",
       "28. 'UKB_related_OBS_CT'\n",
       "29. 'UKB_others_OBS_CT'\n",
       "30. 'UKB_MISSING_CT'\n",
       "31. 'UKB_white_british_MISSING_CT'\n",
       "32. 'UKB_non_british_white_MISSING_CT'\n",
       "33. 'UKB_african_MISSING_CT'\n",
       "34. 'UKB_s_asian_MISSING_CT'\n",
       "35. 'UKB_e_asian_MISSING_CT'\n",
       "36. 'UKB_related_MISSING_CT'\n",
       "37. 'UKB_others_MISSING_CT'\n",
       "38. 'UKB_HOM_REF_CT'\n",
       "39. 'UKB_white_british_HOM_REF_CT'\n",
       "40. 'UKB_non_british_white_HOM_REF_CT'\n",
       "41. 'UKB_african_HOM_REF_CT'\n",
       "42. 'UKB_s_asian_HOM_REF_CT'\n",
       "43. 'UKB_e_asian_HOM_REF_CT'\n",
       "44. 'UKB_related_HOM_REF_CT'\n",
       "45. 'UKB_others_HOM_REF_CT'\n",
       "46. 'UKB_HET_REF_ALT_CTS'\n",
       "47. 'UKB_white_british_HET_REF_ALT_CTS'\n",
       "48. 'UKB_non_british_white_HET_REF_ALT_CTS'\n",
       "49. 'UKB_african_HET_REF_ALT_CTS'\n",
       "50. 'UKB_s_asian_HET_REF_ALT_CTS'\n",
       "51. 'UKB_e_asian_HET_REF_ALT_CTS'\n",
       "52. 'UKB_related_HET_REF_ALT_CTS'\n",
       "53. 'UKB_others_HET_REF_ALT_CTS'\n",
       "54. 'UKB_TWO_ALT_GENO_CTS'\n",
       "55. 'UKB_white_british_TWO_ALT_GENO_CTS'\n",
       "56. 'UKB_non_british_white_TWO_ALT_GENO_CTS'\n",
       "57. 'UKB_african_TWO_ALT_GENO_CTS'\n",
       "58. 'UKB_s_asian_TWO_ALT_GENO_CTS'\n",
       "59. 'UKB_e_asian_TWO_ALT_GENO_CTS'\n",
       "60. 'UKB_related_TWO_ALT_GENO_CTS'\n",
       "61. 'UKB_others_TWO_ALT_GENO_CTS'\n",
       "62. 'UKB_HAP_REF_CT'\n",
       "63. 'UKB_white_british_HAP_REF_CT'\n",
       "64. 'UKB_non_british_white_HAP_REF_CT'\n",
       "65. 'UKB_african_HAP_REF_CT'\n",
       "66. 'UKB_s_asian_HAP_REF_CT'\n",
       "67. 'UKB_e_asian_HAP_REF_CT'\n",
       "68. 'UKB_related_HAP_REF_CT'\n",
       "69. 'UKB_others_HAP_REF_CT'\n",
       "70. 'UKB_HAP_ALT_CTS'\n",
       "71. 'UKB_white_british_HAP_ALT_CTS'\n",
       "72. 'UKB_non_british_white_HAP_ALT_CTS'\n",
       "73. 'UKB_african_HAP_ALT_CTS'\n",
       "74. 'UKB_s_asian_HAP_ALT_CTS'\n",
       "75. 'UKB_e_asian_HAP_ALT_CTS'\n",
       "76. 'UKB_related_HAP_ALT_CTS'\n",
       "77. 'UKB_others_HAP_ALT_CTS'\n",
       "78. 'f_miss'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CHROM\"                                 \n",
       " [2] \"POS\"                                   \n",
       " [3] \"ID\"                                    \n",
       " [4] \"REF\"                                   \n",
       " [5] \"ALT\"                                   \n",
       " [6] \"UKB_AF\"                                \n",
       " [7] \"UKB_white_british_AF\"                  \n",
       " [8] \"UKB_non_british_white_AF\"              \n",
       " [9] \"UKB_african_AF\"                        \n",
       "[10] \"UKB_s_asian_AF\"                        \n",
       "[11] \"UKB_e_asian_AF\"                        \n",
       "[12] \"UKB_related_AF\"                        \n",
       "[13] \"UKB_others_AF\"                         \n",
       "[14] \"UKB_hwe_p\"                             \n",
       "[15] \"UKB_white_british_hwe_p\"               \n",
       "[16] \"UKB_non_british_white_hwe_p\"           \n",
       "[17] \"UKB_african_hwe_p\"                     \n",
       "[18] \"UKB_s_asian_hwe_p\"                     \n",
       "[19] \"UKB_e_asian_hwe_p\"                     \n",
       "[20] \"UKB_related_hwe_p\"                     \n",
       "[21] \"UKB_others_hwe_p\"                      \n",
       "[22] \"UKB_OBS_CT\"                            \n",
       "[23] \"UKB_white_british_OBS_CT\"              \n",
       "[24] \"UKB_non_british_white_OBS_CT\"          \n",
       "[25] \"UKB_african_OBS_CT\"                    \n",
       "[26] \"UKB_s_asian_OBS_CT\"                    \n",
       "[27] \"UKB_e_asian_OBS_CT\"                    \n",
       "[28] \"UKB_related_OBS_CT\"                    \n",
       "[29] \"UKB_others_OBS_CT\"                     \n",
       "[30] \"UKB_MISSING_CT\"                        \n",
       "[31] \"UKB_white_british_MISSING_CT\"          \n",
       "[32] \"UKB_non_british_white_MISSING_CT\"      \n",
       "[33] \"UKB_african_MISSING_CT\"                \n",
       "[34] \"UKB_s_asian_MISSING_CT\"                \n",
       "[35] \"UKB_e_asian_MISSING_CT\"                \n",
       "[36] \"UKB_related_MISSING_CT\"                \n",
       "[37] \"UKB_others_MISSING_CT\"                 \n",
       "[38] \"UKB_HOM_REF_CT\"                        \n",
       "[39] \"UKB_white_british_HOM_REF_CT\"          \n",
       "[40] \"UKB_non_british_white_HOM_REF_CT\"      \n",
       "[41] \"UKB_african_HOM_REF_CT\"                \n",
       "[42] \"UKB_s_asian_HOM_REF_CT\"                \n",
       "[43] \"UKB_e_asian_HOM_REF_CT\"                \n",
       "[44] \"UKB_related_HOM_REF_CT\"                \n",
       "[45] \"UKB_others_HOM_REF_CT\"                 \n",
       "[46] \"UKB_HET_REF_ALT_CTS\"                   \n",
       "[47] \"UKB_white_british_HET_REF_ALT_CTS\"     \n",
       "[48] \"UKB_non_british_white_HET_REF_ALT_CTS\" \n",
       "[49] \"UKB_african_HET_REF_ALT_CTS\"           \n",
       "[50] \"UKB_s_asian_HET_REF_ALT_CTS\"           \n",
       "[51] \"UKB_e_asian_HET_REF_ALT_CTS\"           \n",
       "[52] \"UKB_related_HET_REF_ALT_CTS\"           \n",
       "[53] \"UKB_others_HET_REF_ALT_CTS\"            \n",
       "[54] \"UKB_TWO_ALT_GENO_CTS\"                  \n",
       "[55] \"UKB_white_british_TWO_ALT_GENO_CTS\"    \n",
       "[56] \"UKB_non_british_white_TWO_ALT_GENO_CTS\"\n",
       "[57] \"UKB_african_TWO_ALT_GENO_CTS\"          \n",
       "[58] \"UKB_s_asian_TWO_ALT_GENO_CTS\"          \n",
       "[59] \"UKB_e_asian_TWO_ALT_GENO_CTS\"          \n",
       "[60] \"UKB_related_TWO_ALT_GENO_CTS\"          \n",
       "[61] \"UKB_others_TWO_ALT_GENO_CTS\"           \n",
       "[62] \"UKB_HAP_REF_CT\"                        \n",
       "[63] \"UKB_white_british_HAP_REF_CT\"          \n",
       "[64] \"UKB_non_british_white_HAP_REF_CT\"      \n",
       "[65] \"UKB_african_HAP_REF_CT\"                \n",
       "[66] \"UKB_s_asian_HAP_REF_CT\"                \n",
       "[67] \"UKB_e_asian_HAP_REF_CT\"                \n",
       "[68] \"UKB_related_HAP_REF_CT\"                \n",
       "[69] \"UKB_others_HAP_REF_CT\"                 \n",
       "[70] \"UKB_HAP_ALT_CTS\"                       \n",
       "[71] \"UKB_white_british_HAP_ALT_CTS\"         \n",
       "[72] \"UKB_non_british_white_HAP_ALT_CTS\"     \n",
       "[73] \"UKB_african_HAP_ALT_CTS\"               \n",
       "[74] \"UKB_s_asian_HAP_ALT_CTS\"               \n",
       "[75] \"UKB_e_asian_HAP_ALT_CTS\"               \n",
       "[76] \"UKB_related_HAP_ALT_CTS\"               \n",
       "[77] \"UKB_others_HAP_ALT_CTS\"                \n",
       "[78] \"f_miss\"                                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(af_hwe_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'REF'</li><li>'ALT'</li><li>'FILTER'</li><li>'Allele'</li><li>'Consequence'</li><li>'IMPACT'</li><li>'SYMBOL'</li><li>'Gene'</li><li>'Feature_type'</li><li>'Feature'</li><li>'BIOTYPE'</li><li>'EXON'</li><li>'INTRON'</li><li>'HGVSc'</li><li>'HGVSp'</li><li>'cDNA_position'</li><li>'CDS_position'</li><li>'Protein_position'</li><li>'Amino_acids'</li><li>'Codons'</li><li>'Existing_variation'</li><li>'ALLELE_NUM'</li><li>'DISTANCE'</li><li>'STRAND'</li><li>'FLAGS'</li><li>'VARIANT_CLASS'</li><li>'SYMBOL_SOURCE'</li><li>'HGNC_ID'</li><li>'CANONICAL'</li><li>'MANE'</li><li>'TSL'</li><li>'APPRIS'</li><li>'CCDS'</li><li>'ENSP'</li><li>'SWISSPROT'</li><li>'TREMBL'</li><li>'UNIPARC'</li><li>'GENE_PHENO'</li><li>'SIFT'</li><li>'PolyPhen'</li><li>'DOMAINS'</li><li>'miRNA'</li><li>'HGVS_OFFSET'</li><li>'AF'</li><li>'AFR_AF'</li><li>'AMR_AF'</li><li>'EAS_AF'</li><li>'EUR_AF'</li><li>'SAS_AF'</li><li>'AA_AF'</li><li>'EA_AF'</li><li>'gnomAD_AF'</li><li>'gnomAD_AFR_AF'</li><li>'gnomAD_AMR_AF'</li><li>'gnomAD_ASJ_AF'</li><li>'gnomAD_EAS_AF'</li><li>'gnomAD_FIN_AF'</li><li>'gnomAD_NFE_AF'</li><li>'gnomAD_OTH_AF'</li><li>'gnomAD_SAS_AF'</li><li>'MAX_AF'</li><li>'MAX_AF_POPS'</li><li>'CLIN_SIG'</li><li>'SOMATIC'</li><li>'PHENO'</li><li>'PUBMED'</li><li>'VAR_SYNONYMS'</li><li>'MOTIF_NAME'</li><li>'MOTIF_POS'</li><li>'HIGH_INF_POS'</li><li>'MOTIF_SCORE_CHANGE'</li><li>'TRANSCRIPTION_FACTORS'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\item 'FILTER'\n",
       "\\item 'Allele'\n",
       "\\item 'Consequence'\n",
       "\\item 'IMPACT'\n",
       "\\item 'SYMBOL'\n",
       "\\item 'Gene'\n",
       "\\item 'Feature\\_type'\n",
       "\\item 'Feature'\n",
       "\\item 'BIOTYPE'\n",
       "\\item 'EXON'\n",
       "\\item 'INTRON'\n",
       "\\item 'HGVSc'\n",
       "\\item 'HGVSp'\n",
       "\\item 'cDNA\\_position'\n",
       "\\item 'CDS\\_position'\n",
       "\\item 'Protein\\_position'\n",
       "\\item 'Amino\\_acids'\n",
       "\\item 'Codons'\n",
       "\\item 'Existing\\_variation'\n",
       "\\item 'ALLELE\\_NUM'\n",
       "\\item 'DISTANCE'\n",
       "\\item 'STRAND'\n",
       "\\item 'FLAGS'\n",
       "\\item 'VARIANT\\_CLASS'\n",
       "\\item 'SYMBOL\\_SOURCE'\n",
       "\\item 'HGNC\\_ID'\n",
       "\\item 'CANONICAL'\n",
       "\\item 'MANE'\n",
       "\\item 'TSL'\n",
       "\\item 'APPRIS'\n",
       "\\item 'CCDS'\n",
       "\\item 'ENSP'\n",
       "\\item 'SWISSPROT'\n",
       "\\item 'TREMBL'\n",
       "\\item 'UNIPARC'\n",
       "\\item 'GENE\\_PHENO'\n",
       "\\item 'SIFT'\n",
       "\\item 'PolyPhen'\n",
       "\\item 'DOMAINS'\n",
       "\\item 'miRNA'\n",
       "\\item 'HGVS\\_OFFSET'\n",
       "\\item 'AF'\n",
       "\\item 'AFR\\_AF'\n",
       "\\item 'AMR\\_AF'\n",
       "\\item 'EAS\\_AF'\n",
       "\\item 'EUR\\_AF'\n",
       "\\item 'SAS\\_AF'\n",
       "\\item 'AA\\_AF'\n",
       "\\item 'EA\\_AF'\n",
       "\\item 'gnomAD\\_AF'\n",
       "\\item 'gnomAD\\_AFR\\_AF'\n",
       "\\item 'gnomAD\\_AMR\\_AF'\n",
       "\\item 'gnomAD\\_ASJ\\_AF'\n",
       "\\item 'gnomAD\\_EAS\\_AF'\n",
       "\\item 'gnomAD\\_FIN\\_AF'\n",
       "\\item 'gnomAD\\_NFE\\_AF'\n",
       "\\item 'gnomAD\\_OTH\\_AF'\n",
       "\\item 'gnomAD\\_SAS\\_AF'\n",
       "\\item 'MAX\\_AF'\n",
       "\\item 'MAX\\_AF\\_POPS'\n",
       "\\item 'CLIN\\_SIG'\n",
       "\\item 'SOMATIC'\n",
       "\\item 'PHENO'\n",
       "\\item 'PUBMED'\n",
       "\\item 'VAR\\_SYNONYMS'\n",
       "\\item 'MOTIF\\_NAME'\n",
       "\\item 'MOTIF\\_POS'\n",
       "\\item 'HIGH\\_INF\\_POS'\n",
       "\\item 'MOTIF\\_SCORE\\_CHANGE'\n",
       "\\item 'TRANSCRIPTION\\_FACTORS'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'REF'\n",
       "4. 'ALT'\n",
       "5. 'FILTER'\n",
       "6. 'Allele'\n",
       "7. 'Consequence'\n",
       "8. 'IMPACT'\n",
       "9. 'SYMBOL'\n",
       "10. 'Gene'\n",
       "11. 'Feature_type'\n",
       "12. 'Feature'\n",
       "13. 'BIOTYPE'\n",
       "14. 'EXON'\n",
       "15. 'INTRON'\n",
       "16. 'HGVSc'\n",
       "17. 'HGVSp'\n",
       "18. 'cDNA_position'\n",
       "19. 'CDS_position'\n",
       "20. 'Protein_position'\n",
       "21. 'Amino_acids'\n",
       "22. 'Codons'\n",
       "23. 'Existing_variation'\n",
       "24. 'ALLELE_NUM'\n",
       "25. 'DISTANCE'\n",
       "26. 'STRAND'\n",
       "27. 'FLAGS'\n",
       "28. 'VARIANT_CLASS'\n",
       "29. 'SYMBOL_SOURCE'\n",
       "30. 'HGNC_ID'\n",
       "31. 'CANONICAL'\n",
       "32. 'MANE'\n",
       "33. 'TSL'\n",
       "34. 'APPRIS'\n",
       "35. 'CCDS'\n",
       "36. 'ENSP'\n",
       "37. 'SWISSPROT'\n",
       "38. 'TREMBL'\n",
       "39. 'UNIPARC'\n",
       "40. 'GENE_PHENO'\n",
       "41. 'SIFT'\n",
       "42. 'PolyPhen'\n",
       "43. 'DOMAINS'\n",
       "44. 'miRNA'\n",
       "45. 'HGVS_OFFSET'\n",
       "46. 'AF'\n",
       "47. 'AFR_AF'\n",
       "48. 'AMR_AF'\n",
       "49. 'EAS_AF'\n",
       "50. 'EUR_AF'\n",
       "51. 'SAS_AF'\n",
       "52. 'AA_AF'\n",
       "53. 'EA_AF'\n",
       "54. 'gnomAD_AF'\n",
       "55. 'gnomAD_AFR_AF'\n",
       "56. 'gnomAD_AMR_AF'\n",
       "57. 'gnomAD_ASJ_AF'\n",
       "58. 'gnomAD_EAS_AF'\n",
       "59. 'gnomAD_FIN_AF'\n",
       "60. 'gnomAD_NFE_AF'\n",
       "61. 'gnomAD_OTH_AF'\n",
       "62. 'gnomAD_SAS_AF'\n",
       "63. 'MAX_AF'\n",
       "64. 'MAX_AF_POPS'\n",
       "65. 'CLIN_SIG'\n",
       "66. 'SOMATIC'\n",
       "67. 'PHENO'\n",
       "68. 'PUBMED'\n",
       "69. 'VAR_SYNONYMS'\n",
       "70. 'MOTIF_NAME'\n",
       "71. 'MOTIF_POS'\n",
       "72. 'HIGH_INF_POS'\n",
       "73. 'MOTIF_SCORE_CHANGE'\n",
       "74. 'TRANSCRIPTION_FACTORS'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CHROM\"                 \"POS\"                   \"REF\"                  \n",
       " [4] \"ALT\"                   \"FILTER\"                \"Allele\"               \n",
       " [7] \"Consequence\"           \"IMPACT\"                \"SYMBOL\"               \n",
       "[10] \"Gene\"                  \"Feature_type\"          \"Feature\"              \n",
       "[13] \"BIOTYPE\"               \"EXON\"                  \"INTRON\"               \n",
       "[16] \"HGVSc\"                 \"HGVSp\"                 \"cDNA_position\"        \n",
       "[19] \"CDS_position\"          \"Protein_position\"      \"Amino_acids\"          \n",
       "[22] \"Codons\"                \"Existing_variation\"    \"ALLELE_NUM\"           \n",
       "[25] \"DISTANCE\"              \"STRAND\"                \"FLAGS\"                \n",
       "[28] \"VARIANT_CLASS\"         \"SYMBOL_SOURCE\"         \"HGNC_ID\"              \n",
       "[31] \"CANONICAL\"             \"MANE\"                  \"TSL\"                  \n",
       "[34] \"APPRIS\"                \"CCDS\"                  \"ENSP\"                 \n",
       "[37] \"SWISSPROT\"             \"TREMBL\"                \"UNIPARC\"              \n",
       "[40] \"GENE_PHENO\"            \"SIFT\"                  \"PolyPhen\"             \n",
       "[43] \"DOMAINS\"               \"miRNA\"                 \"HGVS_OFFSET\"          \n",
       "[46] \"AF\"                    \"AFR_AF\"                \"AMR_AF\"               \n",
       "[49] \"EAS_AF\"                \"EUR_AF\"                \"SAS_AF\"               \n",
       "[52] \"AA_AF\"                 \"EA_AF\"                 \"gnomAD_AF\"            \n",
       "[55] \"gnomAD_AFR_AF\"         \"gnomAD_AMR_AF\"         \"gnomAD_ASJ_AF\"        \n",
       "[58] \"gnomAD_EAS_AF\"         \"gnomAD_FIN_AF\"         \"gnomAD_NFE_AF\"        \n",
       "[61] \"gnomAD_OTH_AF\"         \"gnomAD_SAS_AF\"         \"MAX_AF\"               \n",
       "[64] \"MAX_AF_POPS\"           \"CLIN_SIG\"              \"SOMATIC\"              \n",
       "[67] \"PHENO\"                 \"PUBMED\"                \"VAR_SYNONYMS\"         \n",
       "[70] \"MOTIF_NAME\"            \"MOTIF_POS\"             \"HIGH_INF_POS\"         \n",
       "[73] \"MOTIF_SCORE_CHANGE\"    \"TRANSCRIPTION_FACTORS\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(vep_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
