{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages({\n",
    "    library(tidyverse)\n",
    "    library(data.table)\n",
    "}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input files\n",
    "var_qc_f    <- '/oak/stanford/groups/mrivas/ukbb24983/array-combined/annotation/annotation_20201012/ukb24983_cal_hla_cnv.var_QC.tsv.gz'\n",
    "cal_annot_f <- '/oak/stanford/groups/mrivas/ukbb24983/cal/annotation_20201002/ukb24983_cal_cALL_v2_hg19.vep101-loftee.Csq.tsv.gz'\n",
    "cnv_annot_f <- '/oak/stanford/groups/mrivas/ukbb24983/cnv/annotation_20201003/cnv.vep101-loftee.20201009.Csq.tsv.gz'\n",
    "ld_indep_f  <- '/oak/stanford/groups/mrivas/ukbb24983/array-combined/annotation/ld_indep_20201015/ukb24983_cal_hla_cnv.white_british.bool.prune.in.tsv.gz'\n",
    "\n",
    "# output\n",
    "var_annot_f <- '/oak/stanford/groups/mrivas/ukbb24983/array-combined/annotation/annotation_20201012/ukb24983_cal_hla_cnv.annot_20201023.tsv'\n",
    "var_annot_compact_f <- '/oak/stanford/groups/mrivas/ukbb24983/array-combined/annotation/annotation_20201012/ukb24983_cal_hla_cnv.annot_compact_20201023.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat_or_zcat <- function(f){\n",
    "    ifelse(endsWith(f, '.zst'), 'zstdcat', ifelse(endsWith(f, '.gz'), 'zcat', 'cat'))\n",
    "}\n",
    "\n",
    "fread_CHROM <- function(f){\n",
    "    fread(cmd=paste(cat_or_zcat(f), f), colClasses = c('#CHROM'='character')) %>% rename('CHROM'='#CHROM')\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_POS_total <- function(df){\n",
    "    # compute the location on the linear coordinate system (chr1-22, X, Y, MT) for plotting\n",
    "    df %>% select(CHROM, POS, ID) %>%\n",
    "    mutate(CHROM_X = if_else(CHROM == 'XY', 'X', CHROM)) -> CHROM_POS_df\n",
    "\n",
    "    CHROM_POS_df %>% group_by(CHROM_X) %>%\n",
    "    summarise(chr_len = max(POS), .groups = 'drop') %>%\n",
    "    left_join(data.frame(CHROM_X = c(1:22, 'X', 'Y', 'MT'), CHROM_order=1:25, stringsAsFactors=F), by='CHROM_X') %>%\n",
    "    arrange(CHROM_order) %>% mutate(CHROM_tot= cumsum(as.numeric(chr_len)) - chr_len) %>%\n",
    "    select(CHROM_X, CHROM_tot) %>% left_join(CHROM_POS_df, by='CHROM_X') %>%\n",
    "    mutate(POS_total = POS + CHROM_tot) %>%\n",
    "    select(ID, POS_total)    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_qc_f    %>% fread_CHROM() -> var_qc_df\n",
    "cal_annot_f %>% fread_CHROM() -> cal_annot_df\n",
    "cnv_annot_f %>% fread_CHROM() -> cnv_annot_df\n",
    "ld_indep_f  %>% fread(colClasses = c('#ID'='character')) %>% rename('ID'='#ID') -> ld_indep_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_annot_df %>%\n",
    "select(-pvar_order) %>%\n",
    "rename('CNV_POS_s'='POS_s', 'CNV_POS_e'='POS_e') %>%\n",
    "bind_rows(cal_annot_df) -> cal_cnv_annot_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_cnv_annot_df %>%\n",
    "select(-CHROM, -POS, -REF, -ALT) %>%\n",
    "right_join(\n",
    "    var_qc_df %>%\n",
    "    mutate(sort_order=1:n()),\n",
    "    by='ID'\n",
    ") %>%\n",
    "mutate(ld_indep = ID %in% (ld_indep_df$ID)) %>%\n",
    "arrange(sort_order) %>% \n",
    "select(-sort_order) %>% \n",
    "left_join(get_POS_total(var_qc_df), by='ID') -> full_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1080968</li><li>160</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1080968\n",
       "\\item 160\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1080968\n",
       "2. 160\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1080968     160"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "full_df %>% dim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols <- c(\n",
    "    'CHROM', 'POS', 'ID', 'REF', 'ALT', \n",
    "    'FILTER', 'POS_total', 'Allele', 'Csq', 'Consequence', \n",
    "    'SYMBOL', 'Gene', 'ld_indep', 'geno_data_source', 'array',\n",
    "    'CNV_POS_s', 'CNV_POS_e', 'UKB_white_british_MAF', \n",
    "    'hwe_p', 'mgi_notes', 'f_miss', 'f_miss_UKBB', 'f_miss_UKBL',\n",
    "    'LoF', 'LoF_filter', 'LoF_flags', 'LoF_info'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_df %>% \n",
    "select(all_of(c(cols, 'HGVSp'))) %>%\n",
    "rename('#CHROM' = 'CHROM') %>%\n",
    "fwrite(var_annot_compact_f, sep='\\t', na = \"NA\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_df %>% \n",
    "select(all_of(c(cols, setdiff(colnames(full_df), cols)))) %>%\n",
    "rename('#CHROM' = 'CHROM') %>%\n",
    "fwrite(var_annot_f, sep='\\t', na = \"NA\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'ID'</li><li>'REF'</li><li>'ALT'</li><li>'FILTER'</li><li>'POS_total'</li><li>'Allele'</li><li>'Csq'</li><li>'Consequence'</li><li>'SYMBOL'</li><li>'Gene'</li><li>'ld_indep'</li><li>'geno_data_source'</li><li>'array'</li><li>'CNV_POS_s'</li><li>'CNV_POS_e'</li><li>'UKB_white_british_MAF'</li><li>'hwe_p'</li><li>'mgi_notes'</li><li>'f_miss'</li><li>'f_miss_UKBB'</li><li>'f_miss_UKBL'</li><li>'LoF'</li><li>'LoF_filter'</li><li>'LoF_flags'</li><li>'LoF_info'</li><li>'IMPACT'</li><li>'Feature_type'</li><li>'Feature'</li><li>'BIOTYPE'</li><li>'EXON'</li><li>'INTRON'</li><li>'HGVSc'</li><li>'HGVSp'</li><li>'cDNA_position'</li><li>'CDS_position'</li><li>'Protein_position'</li><li>'Amino_acids'</li><li>'Codons'</li><li>'Existing_variation'</li><li>'ALLELE_NUM'</li><li>'DISTANCE'</li><li>'STRAND'</li><li>'FLAGS'</li><li>'VARIANT_CLASS'</li><li>'SYMBOL_SOURCE'</li><li>'HGNC_ID'</li><li>'CANONICAL'</li><li>'MANE'</li><li>'TSL'</li><li>'APPRIS'</li><li>'CCDS'</li><li>'ENSP'</li><li>'SWISSPROT'</li><li>'TREMBL'</li><li>'UNIPARC'</li><li>'GENE_PHENO'</li><li>'SIFT'</li><li>'PolyPhen'</li><li>'DOMAINS'</li><li>'miRNA'</li><li>'HGVS_OFFSET'</li><li>'AF'</li><li>'AFR_AF'</li><li>'AMR_AF'</li><li>'EAS_AF'</li><li>'EUR_AF'</li><li>'SAS_AF'</li><li>'AA_AF'</li><li>'EA_AF'</li><li>'gnomAD_AF'</li><li>'gnomAD_AFR_AF'</li><li>'gnomAD_AMR_AF'</li><li>'gnomAD_ASJ_AF'</li><li>'gnomAD_EAS_AF'</li><li>'gnomAD_FIN_AF'</li><li>'gnomAD_NFE_AF'</li><li>'gnomAD_OTH_AF'</li><li>'gnomAD_SAS_AF'</li><li>'MAX_AF'</li><li>'MAX_AF_POPS'</li><li>'CLIN_SIG'</li><li>'SOMATIC'</li><li>'PHENO'</li><li>'PUBMED'</li><li>'VAR_SYNONYMS'</li><li>'MOTIF_NAME'</li><li>'MOTIF_POS'</li><li>'HIGH_INF_POS'</li><li>'MOTIF_SCORE_CHANGE'</li><li>'TRANSCRIPTION_FACTORS'</li><li>'UKB_AF'</li><li>'UKB_white_british_AF'</li><li>'UKB_non_british_white_AF'</li><li>'UKB_african_AF'</li><li>'UKB_s_asian_AF'</li><li>'UKB_e_asian_AF'</li><li>'UKB_related_AF'</li><li>'UKB_others_AF'</li><li>'UKB_OBS_CT'</li><li>'UKB_white_british_OBS_CT'</li><li>'UKB_non_british_white_OBS_CT'</li><li>'UKB_african_OBS_CT'</li><li>'UKB_s_asian_OBS_CT'</li><li>'UKB_e_asian_OBS_CT'</li><li>'UKB_related_OBS_CT'</li><li>'UKB_others_OBS_CT'</li><li>'UKB_UKBB_OBS_CT'</li><li>'UKB_UKBL_OBS_CT'</li><li>'UKB_MISSING_CT'</li><li>'UKB_white_british_MISSING_CT'</li><li>'UKB_non_british_white_MISSING_CT'</li><li>'UKB_african_MISSING_CT'</li><li>'UKB_s_asian_MISSING_CT'</li><li>'UKB_e_asian_MISSING_CT'</li><li>'UKB_related_MISSING_CT'</li><li>'UKB_others_MISSING_CT'</li><li>'UKB_UKBB_MISSING_CT'</li><li>'UKB_UKBL_MISSING_CT'</li><li>'UKB_HOM_REF_CT'</li><li>'UKB_white_british_HOM_REF_CT'</li><li>'UKB_non_british_white_HOM_REF_CT'</li><li>'UKB_african_HOM_REF_CT'</li><li>'UKB_s_asian_HOM_REF_CT'</li><li>'UKB_e_asian_HOM_REF_CT'</li><li>'UKB_related_HOM_REF_CT'</li><li>'UKB_others_HOM_REF_CT'</li><li>'UKB_HET_REF_ALT_CTS'</li><li>'UKB_white_british_HET_REF_ALT_CTS'</li><li>'UKB_non_british_white_HET_REF_ALT_CTS'</li><li>'UKB_african_HET_REF_ALT_CTS'</li><li>'UKB_s_asian_HET_REF_ALT_CTS'</li><li>'UKB_e_asian_HET_REF_ALT_CTS'</li><li>'UKB_related_HET_REF_ALT_CTS'</li><li>'UKB_others_HET_REF_ALT_CTS'</li><li>'UKB_TWO_ALT_GENO_CTS'</li><li>'UKB_white_british_TWO_ALT_GENO_CTS'</li><li>'UKB_non_british_white_TWO_ALT_GENO_CTS'</li><li>'UKB_african_TWO_ALT_GENO_CTS'</li><li>'UKB_s_asian_TWO_ALT_GENO_CTS'</li><li>'UKB_e_asian_TWO_ALT_GENO_CTS'</li><li>'UKB_related_TWO_ALT_GENO_CTS'</li><li>'UKB_others_TWO_ALT_GENO_CTS'</li><li>'UKB_HAP_REF_CT'</li><li>'UKB_white_british_HAP_REF_CT'</li><li>'UKB_non_british_white_HAP_REF_CT'</li><li>'UKB_african_HAP_REF_CT'</li><li>'UKB_s_asian_HAP_REF_CT'</li><li>'UKB_e_asian_HAP_REF_CT'</li><li>'UKB_related_HAP_REF_CT'</li><li>'UKB_others_HAP_REF_CT'</li><li>'UKB_HAP_ALT_CTS'</li><li>'UKB_white_british_HAP_ALT_CTS'</li><li>'UKB_non_british_white_HAP_ALT_CTS'</li><li>'UKB_african_HAP_ALT_CTS'</li><li>'UKB_s_asian_HAP_ALT_CTS'</li><li>'UKB_e_asian_HAP_ALT_CTS'</li><li>'UKB_related_HAP_ALT_CTS'</li><li>'UKB_others_HAP_ALT_CTS'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'ID'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\item 'FILTER'\n",
       "\\item 'POS\\_total'\n",
       "\\item 'Allele'\n",
       "\\item 'Csq'\n",
       "\\item 'Consequence'\n",
       "\\item 'SYMBOL'\n",
       "\\item 'Gene'\n",
       "\\item 'ld\\_indep'\n",
       "\\item 'geno\\_data\\_source'\n",
       "\\item 'array'\n",
       "\\item 'CNV\\_POS\\_s'\n",
       "\\item 'CNV\\_POS\\_e'\n",
       "\\item 'UKB\\_white\\_british\\_MAF'\n",
       "\\item 'hwe\\_p'\n",
       "\\item 'mgi\\_notes'\n",
       "\\item 'f\\_miss'\n",
       "\\item 'f\\_miss\\_UKBB'\n",
       "\\item 'f\\_miss\\_UKBL'\n",
       "\\item 'LoF'\n",
       "\\item 'LoF\\_filter'\n",
       "\\item 'LoF\\_flags'\n",
       "\\item 'LoF\\_info'\n",
       "\\item 'IMPACT'\n",
       "\\item 'Feature\\_type'\n",
       "\\item 'Feature'\n",
       "\\item 'BIOTYPE'\n",
       "\\item 'EXON'\n",
       "\\item 'INTRON'\n",
       "\\item 'HGVSc'\n",
       "\\item 'HGVSp'\n",
       "\\item 'cDNA\\_position'\n",
       "\\item 'CDS\\_position'\n",
       "\\item 'Protein\\_position'\n",
       "\\item 'Amino\\_acids'\n",
       "\\item 'Codons'\n",
       "\\item 'Existing\\_variation'\n",
       "\\item 'ALLELE\\_NUM'\n",
       "\\item 'DISTANCE'\n",
       "\\item 'STRAND'\n",
       "\\item 'FLAGS'\n",
       "\\item 'VARIANT\\_CLASS'\n",
       "\\item 'SYMBOL\\_SOURCE'\n",
       "\\item 'HGNC\\_ID'\n",
       "\\item 'CANONICAL'\n",
       "\\item 'MANE'\n",
       "\\item 'TSL'\n",
       "\\item 'APPRIS'\n",
       "\\item 'CCDS'\n",
       "\\item 'ENSP'\n",
       "\\item 'SWISSPROT'\n",
       "\\item 'TREMBL'\n",
       "\\item 'UNIPARC'\n",
       "\\item 'GENE\\_PHENO'\n",
       "\\item 'SIFT'\n",
       "\\item 'PolyPhen'\n",
       "\\item 'DOMAINS'\n",
       "\\item 'miRNA'\n",
       "\\item 'HGVS\\_OFFSET'\n",
       "\\item 'AF'\n",
       "\\item 'AFR\\_AF'\n",
       "\\item 'AMR\\_AF'\n",
       "\\item 'EAS\\_AF'\n",
       "\\item 'EUR\\_AF'\n",
       "\\item 'SAS\\_AF'\n",
       "\\item 'AA\\_AF'\n",
       "\\item 'EA\\_AF'\n",
       "\\item 'gnomAD\\_AF'\n",
       "\\item 'gnomAD\\_AFR\\_AF'\n",
       "\\item 'gnomAD\\_AMR\\_AF'\n",
       "\\item 'gnomAD\\_ASJ\\_AF'\n",
       "\\item 'gnomAD\\_EAS\\_AF'\n",
       "\\item 'gnomAD\\_FIN\\_AF'\n",
       "\\item 'gnomAD\\_NFE\\_AF'\n",
       "\\item 'gnomAD\\_OTH\\_AF'\n",
       "\\item 'gnomAD\\_SAS\\_AF'\n",
       "\\item 'MAX\\_AF'\n",
       "\\item 'MAX\\_AF\\_POPS'\n",
       "\\item 'CLIN\\_SIG'\n",
       "\\item 'SOMATIC'\n",
       "\\item 'PHENO'\n",
       "\\item 'PUBMED'\n",
       "\\item 'VAR\\_SYNONYMS'\n",
       "\\item 'MOTIF\\_NAME'\n",
       "\\item 'MOTIF\\_POS'\n",
       "\\item 'HIGH\\_INF\\_POS'\n",
       "\\item 'MOTIF\\_SCORE\\_CHANGE'\n",
       "\\item 'TRANSCRIPTION\\_FACTORS'\n",
       "\\item 'UKB\\_AF'\n",
       "\\item 'UKB\\_white\\_british\\_AF'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_AF'\n",
       "\\item 'UKB\\_african\\_AF'\n",
       "\\item 'UKB\\_s\\_asian\\_AF'\n",
       "\\item 'UKB\\_e\\_asian\\_AF'\n",
       "\\item 'UKB\\_related\\_AF'\n",
       "\\item 'UKB\\_others\\_AF'\n",
       "\\item 'UKB\\_OBS\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_OBS\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_OBS\\_CT'\n",
       "\\item 'UKB\\_african\\_OBS\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_OBS\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_OBS\\_CT'\n",
       "\\item 'UKB\\_related\\_OBS\\_CT'\n",
       "\\item 'UKB\\_others\\_OBS\\_CT'\n",
       "\\item 'UKB\\_UKBB\\_OBS\\_CT'\n",
       "\\item 'UKB\\_UKBL\\_OBS\\_CT'\n",
       "\\item 'UKB\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_african\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_related\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_others\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_UKBB\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_UKBL\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_african\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_related\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_others\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_african\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_related\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_others\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_african\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_related\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_others\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_african\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_related\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_others\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_african\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_related\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_others\\_HAP\\_ALT\\_CTS'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'ID'\n",
       "4. 'REF'\n",
       "5. 'ALT'\n",
       "6. 'FILTER'\n",
       "7. 'POS_total'\n",
       "8. 'Allele'\n",
       "9. 'Csq'\n",
       "10. 'Consequence'\n",
       "11. 'SYMBOL'\n",
       "12. 'Gene'\n",
       "13. 'ld_indep'\n",
       "14. 'geno_data_source'\n",
       "15. 'array'\n",
       "16. 'CNV_POS_s'\n",
       "17. 'CNV_POS_e'\n",
       "18. 'UKB_white_british_MAF'\n",
       "19. 'hwe_p'\n",
       "20. 'mgi_notes'\n",
       "21. 'f_miss'\n",
       "22. 'f_miss_UKBB'\n",
       "23. 'f_miss_UKBL'\n",
       "24. 'LoF'\n",
       "25. 'LoF_filter'\n",
       "26. 'LoF_flags'\n",
       "27. 'LoF_info'\n",
       "28. 'IMPACT'\n",
       "29. 'Feature_type'\n",
       "30. 'Feature'\n",
       "31. 'BIOTYPE'\n",
       "32. 'EXON'\n",
       "33. 'INTRON'\n",
       "34. 'HGVSc'\n",
       "35. 'HGVSp'\n",
       "36. 'cDNA_position'\n",
       "37. 'CDS_position'\n",
       "38. 'Protein_position'\n",
       "39. 'Amino_acids'\n",
       "40. 'Codons'\n",
       "41. 'Existing_variation'\n",
       "42. 'ALLELE_NUM'\n",
       "43. 'DISTANCE'\n",
       "44. 'STRAND'\n",
       "45. 'FLAGS'\n",
       "46. 'VARIANT_CLASS'\n",
       "47. 'SYMBOL_SOURCE'\n",
       "48. 'HGNC_ID'\n",
       "49. 'CANONICAL'\n",
       "50. 'MANE'\n",
       "51. 'TSL'\n",
       "52. 'APPRIS'\n",
       "53. 'CCDS'\n",
       "54. 'ENSP'\n",
       "55. 'SWISSPROT'\n",
       "56. 'TREMBL'\n",
       "57. 'UNIPARC'\n",
       "58. 'GENE_PHENO'\n",
       "59. 'SIFT'\n",
       "60. 'PolyPhen'\n",
       "61. 'DOMAINS'\n",
       "62. 'miRNA'\n",
       "63. 'HGVS_OFFSET'\n",
       "64. 'AF'\n",
       "65. 'AFR_AF'\n",
       "66. 'AMR_AF'\n",
       "67. 'EAS_AF'\n",
       "68. 'EUR_AF'\n",
       "69. 'SAS_AF'\n",
       "70. 'AA_AF'\n",
       "71. 'EA_AF'\n",
       "72. 'gnomAD_AF'\n",
       "73. 'gnomAD_AFR_AF'\n",
       "74. 'gnomAD_AMR_AF'\n",
       "75. 'gnomAD_ASJ_AF'\n",
       "76. 'gnomAD_EAS_AF'\n",
       "77. 'gnomAD_FIN_AF'\n",
       "78. 'gnomAD_NFE_AF'\n",
       "79. 'gnomAD_OTH_AF'\n",
       "80. 'gnomAD_SAS_AF'\n",
       "81. 'MAX_AF'\n",
       "82. 'MAX_AF_POPS'\n",
       "83. 'CLIN_SIG'\n",
       "84. 'SOMATIC'\n",
       "85. 'PHENO'\n",
       "86. 'PUBMED'\n",
       "87. 'VAR_SYNONYMS'\n",
       "88. 'MOTIF_NAME'\n",
       "89. 'MOTIF_POS'\n",
       "90. 'HIGH_INF_POS'\n",
       "91. 'MOTIF_SCORE_CHANGE'\n",
       "92. 'TRANSCRIPTION_FACTORS'\n",
       "93. 'UKB_AF'\n",
       "94. 'UKB_white_british_AF'\n",
       "95. 'UKB_non_british_white_AF'\n",
       "96. 'UKB_african_AF'\n",
       "97. 'UKB_s_asian_AF'\n",
       "98. 'UKB_e_asian_AF'\n",
       "99. 'UKB_related_AF'\n",
       "100. 'UKB_others_AF'\n",
       "101. 'UKB_OBS_CT'\n",
       "102. 'UKB_white_british_OBS_CT'\n",
       "103. 'UKB_non_british_white_OBS_CT'\n",
       "104. 'UKB_african_OBS_CT'\n",
       "105. 'UKB_s_asian_OBS_CT'\n",
       "106. 'UKB_e_asian_OBS_CT'\n",
       "107. 'UKB_related_OBS_CT'\n",
       "108. 'UKB_others_OBS_CT'\n",
       "109. 'UKB_UKBB_OBS_CT'\n",
       "110. 'UKB_UKBL_OBS_CT'\n",
       "111. 'UKB_MISSING_CT'\n",
       "112. 'UKB_white_british_MISSING_CT'\n",
       "113. 'UKB_non_british_white_MISSING_CT'\n",
       "114. 'UKB_african_MISSING_CT'\n",
       "115. 'UKB_s_asian_MISSING_CT'\n",
       "116. 'UKB_e_asian_MISSING_CT'\n",
       "117. 'UKB_related_MISSING_CT'\n",
       "118. 'UKB_others_MISSING_CT'\n",
       "119. 'UKB_UKBB_MISSING_CT'\n",
       "120. 'UKB_UKBL_MISSING_CT'\n",
       "121. 'UKB_HOM_REF_CT'\n",
       "122. 'UKB_white_british_HOM_REF_CT'\n",
       "123. 'UKB_non_british_white_HOM_REF_CT'\n",
       "124. 'UKB_african_HOM_REF_CT'\n",
       "125. 'UKB_s_asian_HOM_REF_CT'\n",
       "126. 'UKB_e_asian_HOM_REF_CT'\n",
       "127. 'UKB_related_HOM_REF_CT'\n",
       "128. 'UKB_others_HOM_REF_CT'\n",
       "129. 'UKB_HET_REF_ALT_CTS'\n",
       "130. 'UKB_white_british_HET_REF_ALT_CTS'\n",
       "131. 'UKB_non_british_white_HET_REF_ALT_CTS'\n",
       "132. 'UKB_african_HET_REF_ALT_CTS'\n",
       "133. 'UKB_s_asian_HET_REF_ALT_CTS'\n",
       "134. 'UKB_e_asian_HET_REF_ALT_CTS'\n",
       "135. 'UKB_related_HET_REF_ALT_CTS'\n",
       "136. 'UKB_others_HET_REF_ALT_CTS'\n",
       "137. 'UKB_TWO_ALT_GENO_CTS'\n",
       "138. 'UKB_white_british_TWO_ALT_GENO_CTS'\n",
       "139. 'UKB_non_british_white_TWO_ALT_GENO_CTS'\n",
       "140. 'UKB_african_TWO_ALT_GENO_CTS'\n",
       "141. 'UKB_s_asian_TWO_ALT_GENO_CTS'\n",
       "142. 'UKB_e_asian_TWO_ALT_GENO_CTS'\n",
       "143. 'UKB_related_TWO_ALT_GENO_CTS'\n",
       "144. 'UKB_others_TWO_ALT_GENO_CTS'\n",
       "145. 'UKB_HAP_REF_CT'\n",
       "146. 'UKB_white_british_HAP_REF_CT'\n",
       "147. 'UKB_non_british_white_HAP_REF_CT'\n",
       "148. 'UKB_african_HAP_REF_CT'\n",
       "149. 'UKB_s_asian_HAP_REF_CT'\n",
       "150. 'UKB_e_asian_HAP_REF_CT'\n",
       "151. 'UKB_related_HAP_REF_CT'\n",
       "152. 'UKB_others_HAP_REF_CT'\n",
       "153. 'UKB_HAP_ALT_CTS'\n",
       "154. 'UKB_white_british_HAP_ALT_CTS'\n",
       "155. 'UKB_non_british_white_HAP_ALT_CTS'\n",
       "156. 'UKB_african_HAP_ALT_CTS'\n",
       "157. 'UKB_s_asian_HAP_ALT_CTS'\n",
       "158. 'UKB_e_asian_HAP_ALT_CTS'\n",
       "159. 'UKB_related_HAP_ALT_CTS'\n",
       "160. 'UKB_others_HAP_ALT_CTS'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"CHROM\"                                 \n",
       "  [2] \"POS\"                                   \n",
       "  [3] \"ID\"                                    \n",
       "  [4] \"REF\"                                   \n",
       "  [5] \"ALT\"                                   \n",
       "  [6] \"FILTER\"                                \n",
       "  [7] \"POS_total\"                             \n",
       "  [8] \"Allele\"                                \n",
       "  [9] \"Csq\"                                   \n",
       " [10] \"Consequence\"                           \n",
       " [11] \"SYMBOL\"                                \n",
       " [12] \"Gene\"                                  \n",
       " [13] \"ld_indep\"                              \n",
       " [14] \"geno_data_source\"                      \n",
       " [15] \"array\"                                 \n",
       " [16] \"CNV_POS_s\"                             \n",
       " [17] \"CNV_POS_e\"                             \n",
       " [18] \"UKB_white_british_MAF\"                 \n",
       " [19] \"hwe_p\"                                 \n",
       " [20] \"mgi_notes\"                             \n",
       " [21] \"f_miss\"                                \n",
       " [22] \"f_miss_UKBB\"                           \n",
       " [23] \"f_miss_UKBL\"                           \n",
       " [24] \"LoF\"                                   \n",
       " [25] \"LoF_filter\"                            \n",
       " [26] \"LoF_flags\"                             \n",
       " [27] \"LoF_info\"                              \n",
       " [28] \"IMPACT\"                                \n",
       " [29] \"Feature_type\"                          \n",
       " [30] \"Feature\"                               \n",
       " [31] \"BIOTYPE\"                               \n",
       " [32] \"EXON\"                                  \n",
       " [33] \"INTRON\"                                \n",
       " [34] \"HGVSc\"                                 \n",
       " [35] \"HGVSp\"                                 \n",
       " [36] \"cDNA_position\"                         \n",
       " [37] \"CDS_position\"                          \n",
       " [38] \"Protein_position\"                      \n",
       " [39] \"Amino_acids\"                           \n",
       " [40] \"Codons\"                                \n",
       " [41] \"Existing_variation\"                    \n",
       " [42] \"ALLELE_NUM\"                            \n",
       " [43] \"DISTANCE\"                              \n",
       " [44] \"STRAND\"                                \n",
       " [45] \"FLAGS\"                                 \n",
       " [46] \"VARIANT_CLASS\"                         \n",
       " [47] \"SYMBOL_SOURCE\"                         \n",
       " [48] \"HGNC_ID\"                               \n",
       " [49] \"CANONICAL\"                             \n",
       " [50] \"MANE\"                                  \n",
       " [51] \"TSL\"                                   \n",
       " [52] \"APPRIS\"                                \n",
       " [53] \"CCDS\"                                  \n",
       " [54] \"ENSP\"                                  \n",
       " [55] \"SWISSPROT\"                             \n",
       " [56] \"TREMBL\"                                \n",
       " [57] \"UNIPARC\"                               \n",
       " [58] \"GENE_PHENO\"                            \n",
       " [59] \"SIFT\"                                  \n",
       " [60] \"PolyPhen\"                              \n",
       " [61] \"DOMAINS\"                               \n",
       " [62] \"miRNA\"                                 \n",
       " [63] \"HGVS_OFFSET\"                           \n",
       " [64] \"AF\"                                    \n",
       " [65] \"AFR_AF\"                                \n",
       " [66] \"AMR_AF\"                                \n",
       " [67] \"EAS_AF\"                                \n",
       " [68] \"EUR_AF\"                                \n",
       " [69] \"SAS_AF\"                                \n",
       " [70] \"AA_AF\"                                 \n",
       " [71] \"EA_AF\"                                 \n",
       " [72] \"gnomAD_AF\"                             \n",
       " [73] \"gnomAD_AFR_AF\"                         \n",
       " [74] \"gnomAD_AMR_AF\"                         \n",
       " [75] \"gnomAD_ASJ_AF\"                         \n",
       " [76] \"gnomAD_EAS_AF\"                         \n",
       " [77] \"gnomAD_FIN_AF\"                         \n",
       " [78] \"gnomAD_NFE_AF\"                         \n",
       " [79] \"gnomAD_OTH_AF\"                         \n",
       " [80] \"gnomAD_SAS_AF\"                         \n",
       " [81] \"MAX_AF\"                                \n",
       " [82] \"MAX_AF_POPS\"                           \n",
       " [83] \"CLIN_SIG\"                              \n",
       " [84] \"SOMATIC\"                               \n",
       " [85] \"PHENO\"                                 \n",
       " [86] \"PUBMED\"                                \n",
       " [87] \"VAR_SYNONYMS\"                          \n",
       " [88] \"MOTIF_NAME\"                            \n",
       " [89] \"MOTIF_POS\"                             \n",
       " [90] \"HIGH_INF_POS\"                          \n",
       " [91] \"MOTIF_SCORE_CHANGE\"                    \n",
       " [92] \"TRANSCRIPTION_FACTORS\"                 \n",
       " [93] \"UKB_AF\"                                \n",
       " [94] \"UKB_white_british_AF\"                  \n",
       " [95] \"UKB_non_british_white_AF\"              \n",
       " [96] \"UKB_african_AF\"                        \n",
       " [97] \"UKB_s_asian_AF\"                        \n",
       " [98] \"UKB_e_asian_AF\"                        \n",
       " [99] \"UKB_related_AF\"                        \n",
       "[100] \"UKB_others_AF\"                         \n",
       "[101] \"UKB_OBS_CT\"                            \n",
       "[102] \"UKB_white_british_OBS_CT\"              \n",
       "[103] \"UKB_non_british_white_OBS_CT\"          \n",
       "[104] \"UKB_african_OBS_CT\"                    \n",
       "[105] \"UKB_s_asian_OBS_CT\"                    \n",
       "[106] \"UKB_e_asian_OBS_CT\"                    \n",
       "[107] \"UKB_related_OBS_CT\"                    \n",
       "[108] \"UKB_others_OBS_CT\"                     \n",
       "[109] \"UKB_UKBB_OBS_CT\"                       \n",
       "[110] \"UKB_UKBL_OBS_CT\"                       \n",
       "[111] \"UKB_MISSING_CT\"                        \n",
       "[112] \"UKB_white_british_MISSING_CT\"          \n",
       "[113] \"UKB_non_british_white_MISSING_CT\"      \n",
       "[114] \"UKB_african_MISSING_CT\"                \n",
       "[115] \"UKB_s_asian_MISSING_CT\"                \n",
       "[116] \"UKB_e_asian_MISSING_CT\"                \n",
       "[117] \"UKB_related_MISSING_CT\"                \n",
       "[118] \"UKB_others_MISSING_CT\"                 \n",
       "[119] \"UKB_UKBB_MISSING_CT\"                   \n",
       "[120] \"UKB_UKBL_MISSING_CT\"                   \n",
       "[121] \"UKB_HOM_REF_CT\"                        \n",
       "[122] \"UKB_white_british_HOM_REF_CT\"          \n",
       "[123] \"UKB_non_british_white_HOM_REF_CT\"      \n",
       "[124] \"UKB_african_HOM_REF_CT\"                \n",
       "[125] \"UKB_s_asian_HOM_REF_CT\"                \n",
       "[126] \"UKB_e_asian_HOM_REF_CT\"                \n",
       "[127] \"UKB_related_HOM_REF_CT\"                \n",
       "[128] \"UKB_others_HOM_REF_CT\"                 \n",
       "[129] \"UKB_HET_REF_ALT_CTS\"                   \n",
       "[130] \"UKB_white_british_HET_REF_ALT_CTS\"     \n",
       "[131] \"UKB_non_british_white_HET_REF_ALT_CTS\" \n",
       "[132] \"UKB_african_HET_REF_ALT_CTS\"           \n",
       "[133] \"UKB_s_asian_HET_REF_ALT_CTS\"           \n",
       "[134] \"UKB_e_asian_HET_REF_ALT_CTS\"           \n",
       "[135] \"UKB_related_HET_REF_ALT_CTS\"           \n",
       "[136] \"UKB_others_HET_REF_ALT_CTS\"            \n",
       "[137] \"UKB_TWO_ALT_GENO_CTS\"                  \n",
       "[138] \"UKB_white_british_TWO_ALT_GENO_CTS\"    \n",
       "[139] \"UKB_non_british_white_TWO_ALT_GENO_CTS\"\n",
       "[140] \"UKB_african_TWO_ALT_GENO_CTS\"          \n",
       "[141] \"UKB_s_asian_TWO_ALT_GENO_CTS\"          \n",
       "[142] \"UKB_e_asian_TWO_ALT_GENO_CTS\"          \n",
       "[143] \"UKB_related_TWO_ALT_GENO_CTS\"          \n",
       "[144] \"UKB_others_TWO_ALT_GENO_CTS\"           \n",
       "[145] \"UKB_HAP_REF_CT\"                        \n",
       "[146] \"UKB_white_british_HAP_REF_CT\"          \n",
       "[147] \"UKB_non_british_white_HAP_REF_CT\"      \n",
       "[148] \"UKB_african_HAP_REF_CT\"                \n",
       "[149] \"UKB_s_asian_HAP_REF_CT\"                \n",
       "[150] \"UKB_e_asian_HAP_REF_CT\"                \n",
       "[151] \"UKB_related_HAP_REF_CT\"                \n",
       "[152] \"UKB_others_HAP_REF_CT\"                 \n",
       "[153] \"UKB_HAP_ALT_CTS\"                       \n",
       "[154] \"UKB_white_british_HAP_ALT_CTS\"         \n",
       "[155] \"UKB_non_british_white_HAP_ALT_CTS\"     \n",
       "[156] \"UKB_african_HAP_ALT_CTS\"               \n",
       "[157] \"UKB_s_asian_HAP_ALT_CTS\"               \n",
       "[158] \"UKB_e_asian_HAP_ALT_CTS\"               \n",
       "[159] \"UKB_related_HAP_ALT_CTS\"               \n",
       "[160] \"UKB_others_HAP_ALT_CTS\"                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "full_df %>% \n",
    "select(all_of(c(cols, setdiff(colnames(full_df), cols)))) %>%\n",
    "colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
