{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages({\n",
    "    library(tidyverse)\n",
    "    library(data.table)\n",
    "}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvar_f <- '/oak/stanford/groups/mrivas/ukbb24983/array-combined/pgen/ukb24983_cal_hla_cnv.pvar'\n",
    "afreq_f <- file.path(dirname(dirname(pvar_f)), 'annotation/afreq_20201012/ukb24983_cal_hla_cnv.afreq_20201012.pvar.zst')\n",
    "old_cal_annot_f <- '/oak/stanford/groups/mrivas/private_data/ukbb/variant_filtering/variant_filter_table.20200701.tsv.gz'\n",
    "cal_annot_f <- '/oak/stanford/groups/mrivas/ukbb24983/cal/annotation_20201002/ukb24983_cal_cALL_v2_hg19.vep101-loftee.Csq.tsv.gz'\n",
    "cnv_annot_f <- '/oak/stanford/groups/mrivas/ukbb24983/cnv/annotation_20201003/cnv.vep101-loftee.20201009.Csq.tsv.gz'\n",
    "# snp_qc_f <- '/oak/stanford/groups/mrivas/ukbb24983/snp/ukb_snp_qc.pvar.zst'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat_or_zcat <- function(f){\n",
    "    ifelse(endsWith(f, '.zst'), 'zstdcat', ifelse(endsWith(f, '.gz'), 'zcat', 'cat'))\n",
    "}\n",
    "\n",
    "fread_CHROM <- function(f){\n",
    "    fread(cmd=paste(cat_or_zcat(f), f), colClasses = c('#CHROM'='character')) %>% rename('CHROM'='#CHROM')\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_qc_f %>% fread_CHROM() -> snp_qc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_annot_f %>% fread_CHROM() -> cal_annot_df\n",
    "cnv_annot_f %>% fread_CHROM() -> cnv_annot_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "afreq_f %>% fread_CHROM() -> afreq_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'ID'</li><li>'REF'</li><li>'ALT'</li><li>'array'</li><li>'UKB_AF'</li><li>'UKB_white_british_AF'</li><li>'UKB_non_british_white_AF'</li><li>'UKB_african_AF'</li><li>'UKB_s_asian_AF'</li><li>'UKB_e_asian_AF'</li><li>'UKB_related_AF'</li><li>'UKB_others_AF'</li><li>'UKB_OBS_CT'</li><li>'UKB_white_british_OBS_CT'</li><li>'UKB_non_british_white_OBS_CT'</li><li>'UKB_african_OBS_CT'</li><li>'UKB_s_asian_OBS_CT'</li><li>'UKB_e_asian_OBS_CT'</li><li>'UKB_related_OBS_CT'</li><li>'UKB_others_OBS_CT'</li><li>'UKB_UKBB_OBS_CT'</li><li>'UKB_UKBL_OBS_CT'</li><li>'UKB_MISSING_CT'</li><li>'UKB_white_british_MISSING_CT'</li><li>'UKB_non_british_white_MISSING_CT'</li><li>'UKB_african_MISSING_CT'</li><li>'UKB_s_asian_MISSING_CT'</li><li>'UKB_e_asian_MISSING_CT'</li><li>'UKB_related_MISSING_CT'</li><li>'UKB_others_MISSING_CT'</li><li>'UKB_UKBB_MISSING_CT'</li><li>'UKB_UKBL_MISSING_CT'</li><li>'UKB_HOM_REF_CT'</li><li>'UKB_white_british_HOM_REF_CT'</li><li>'UKB_non_british_white_HOM_REF_CT'</li><li>'UKB_african_HOM_REF_CT'</li><li>'UKB_s_asian_HOM_REF_CT'</li><li>'UKB_e_asian_HOM_REF_CT'</li><li>'UKB_related_HOM_REF_CT'</li><li>'UKB_others_HOM_REF_CT'</li><li>'UKB_HET_REF_ALT_CTS'</li><li>'UKB_white_british_HET_REF_ALT_CTS'</li><li>'UKB_non_british_white_HET_REF_ALT_CTS'</li><li>'UKB_african_HET_REF_ALT_CTS'</li><li>'UKB_s_asian_HET_REF_ALT_CTS'</li><li>'UKB_e_asian_HET_REF_ALT_CTS'</li><li>'UKB_related_HET_REF_ALT_CTS'</li><li>'UKB_others_HET_REF_ALT_CTS'</li><li>'UKB_TWO_ALT_GENO_CTS'</li><li>'UKB_white_british_TWO_ALT_GENO_CTS'</li><li>'UKB_non_british_white_TWO_ALT_GENO_CTS'</li><li>'UKB_african_TWO_ALT_GENO_CTS'</li><li>'UKB_s_asian_TWO_ALT_GENO_CTS'</li><li>'UKB_e_asian_TWO_ALT_GENO_CTS'</li><li>'UKB_related_TWO_ALT_GENO_CTS'</li><li>'UKB_others_TWO_ALT_GENO_CTS'</li><li>'UKB_HAP_REF_CT'</li><li>'UKB_white_british_HAP_REF_CT'</li><li>'UKB_non_british_white_HAP_REF_CT'</li><li>'UKB_african_HAP_REF_CT'</li><li>'UKB_s_asian_HAP_REF_CT'</li><li>'UKB_e_asian_HAP_REF_CT'</li><li>'UKB_related_HAP_REF_CT'</li><li>'UKB_others_HAP_REF_CT'</li><li>'UKB_HAP_ALT_CTS'</li><li>'UKB_white_british_HAP_ALT_CTS'</li><li>'UKB_non_british_white_HAP_ALT_CTS'</li><li>'UKB_african_HAP_ALT_CTS'</li><li>'UKB_s_asian_HAP_ALT_CTS'</li><li>'UKB_e_asian_HAP_ALT_CTS'</li><li>'UKB_related_HAP_ALT_CTS'</li><li>'UKB_others_HAP_ALT_CTS'</li><li>'f_miss'</li><li>'f_miss_UKBB'</li><li>'f_miss_UKBL'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'ID'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\item 'array'\n",
       "\\item 'UKB\\_AF'\n",
       "\\item 'UKB\\_white\\_british\\_AF'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_AF'\n",
       "\\item 'UKB\\_african\\_AF'\n",
       "\\item 'UKB\\_s\\_asian\\_AF'\n",
       "\\item 'UKB\\_e\\_asian\\_AF'\n",
       "\\item 'UKB\\_related\\_AF'\n",
       "\\item 'UKB\\_others\\_AF'\n",
       "\\item 'UKB\\_OBS\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_OBS\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_OBS\\_CT'\n",
       "\\item 'UKB\\_african\\_OBS\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_OBS\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_OBS\\_CT'\n",
       "\\item 'UKB\\_related\\_OBS\\_CT'\n",
       "\\item 'UKB\\_others\\_OBS\\_CT'\n",
       "\\item 'UKB\\_UKBB\\_OBS\\_CT'\n",
       "\\item 'UKB\\_UKBL\\_OBS\\_CT'\n",
       "\\item 'UKB\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_african\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_related\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_others\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_UKBB\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_UKBL\\_MISSING\\_CT'\n",
       "\\item 'UKB\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_african\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_related\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_others\\_HOM\\_REF\\_CT'\n",
       "\\item 'UKB\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_african\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_related\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_others\\_HET\\_REF\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_african\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_related\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_others\\_TWO\\_ALT\\_GENO\\_CTS'\n",
       "\\item 'UKB\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_white\\_british\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_african\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_s\\_asian\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_e\\_asian\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_related\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_others\\_HAP\\_REF\\_CT'\n",
       "\\item 'UKB\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_white\\_british\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_non\\_british\\_white\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_african\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_s\\_asian\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_e\\_asian\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_related\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'UKB\\_others\\_HAP\\_ALT\\_CTS'\n",
       "\\item 'f\\_miss'\n",
       "\\item 'f\\_miss\\_UKBB'\n",
       "\\item 'f\\_miss\\_UKBL'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'ID'\n",
       "4. 'REF'\n",
       "5. 'ALT'\n",
       "6. 'array'\n",
       "7. 'UKB_AF'\n",
       "8. 'UKB_white_british_AF'\n",
       "9. 'UKB_non_british_white_AF'\n",
       "10. 'UKB_african_AF'\n",
       "11. 'UKB_s_asian_AF'\n",
       "12. 'UKB_e_asian_AF'\n",
       "13. 'UKB_related_AF'\n",
       "14. 'UKB_others_AF'\n",
       "15. 'UKB_OBS_CT'\n",
       "16. 'UKB_white_british_OBS_CT'\n",
       "17. 'UKB_non_british_white_OBS_CT'\n",
       "18. 'UKB_african_OBS_CT'\n",
       "19. 'UKB_s_asian_OBS_CT'\n",
       "20. 'UKB_e_asian_OBS_CT'\n",
       "21. 'UKB_related_OBS_CT'\n",
       "22. 'UKB_others_OBS_CT'\n",
       "23. 'UKB_UKBB_OBS_CT'\n",
       "24. 'UKB_UKBL_OBS_CT'\n",
       "25. 'UKB_MISSING_CT'\n",
       "26. 'UKB_white_british_MISSING_CT'\n",
       "27. 'UKB_non_british_white_MISSING_CT'\n",
       "28. 'UKB_african_MISSING_CT'\n",
       "29. 'UKB_s_asian_MISSING_CT'\n",
       "30. 'UKB_e_asian_MISSING_CT'\n",
       "31. 'UKB_related_MISSING_CT'\n",
       "32. 'UKB_others_MISSING_CT'\n",
       "33. 'UKB_UKBB_MISSING_CT'\n",
       "34. 'UKB_UKBL_MISSING_CT'\n",
       "35. 'UKB_HOM_REF_CT'\n",
       "36. 'UKB_white_british_HOM_REF_CT'\n",
       "37. 'UKB_non_british_white_HOM_REF_CT'\n",
       "38. 'UKB_african_HOM_REF_CT'\n",
       "39. 'UKB_s_asian_HOM_REF_CT'\n",
       "40. 'UKB_e_asian_HOM_REF_CT'\n",
       "41. 'UKB_related_HOM_REF_CT'\n",
       "42. 'UKB_others_HOM_REF_CT'\n",
       "43. 'UKB_HET_REF_ALT_CTS'\n",
       "44. 'UKB_white_british_HET_REF_ALT_CTS'\n",
       "45. 'UKB_non_british_white_HET_REF_ALT_CTS'\n",
       "46. 'UKB_african_HET_REF_ALT_CTS'\n",
       "47. 'UKB_s_asian_HET_REF_ALT_CTS'\n",
       "48. 'UKB_e_asian_HET_REF_ALT_CTS'\n",
       "49. 'UKB_related_HET_REF_ALT_CTS'\n",
       "50. 'UKB_others_HET_REF_ALT_CTS'\n",
       "51. 'UKB_TWO_ALT_GENO_CTS'\n",
       "52. 'UKB_white_british_TWO_ALT_GENO_CTS'\n",
       "53. 'UKB_non_british_white_TWO_ALT_GENO_CTS'\n",
       "54. 'UKB_african_TWO_ALT_GENO_CTS'\n",
       "55. 'UKB_s_asian_TWO_ALT_GENO_CTS'\n",
       "56. 'UKB_e_asian_TWO_ALT_GENO_CTS'\n",
       "57. 'UKB_related_TWO_ALT_GENO_CTS'\n",
       "58. 'UKB_others_TWO_ALT_GENO_CTS'\n",
       "59. 'UKB_HAP_REF_CT'\n",
       "60. 'UKB_white_british_HAP_REF_CT'\n",
       "61. 'UKB_non_british_white_HAP_REF_CT'\n",
       "62. 'UKB_african_HAP_REF_CT'\n",
       "63. 'UKB_s_asian_HAP_REF_CT'\n",
       "64. 'UKB_e_asian_HAP_REF_CT'\n",
       "65. 'UKB_related_HAP_REF_CT'\n",
       "66. 'UKB_others_HAP_REF_CT'\n",
       "67. 'UKB_HAP_ALT_CTS'\n",
       "68. 'UKB_white_british_HAP_ALT_CTS'\n",
       "69. 'UKB_non_british_white_HAP_ALT_CTS'\n",
       "70. 'UKB_african_HAP_ALT_CTS'\n",
       "71. 'UKB_s_asian_HAP_ALT_CTS'\n",
       "72. 'UKB_e_asian_HAP_ALT_CTS'\n",
       "73. 'UKB_related_HAP_ALT_CTS'\n",
       "74. 'UKB_others_HAP_ALT_CTS'\n",
       "75. 'f_miss'\n",
       "76. 'f_miss_UKBB'\n",
       "77. 'f_miss_UKBL'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CHROM\"                                 \n",
       " [2] \"POS\"                                   \n",
       " [3] \"ID\"                                    \n",
       " [4] \"REF\"                                   \n",
       " [5] \"ALT\"                                   \n",
       " [6] \"array\"                                 \n",
       " [7] \"UKB_AF\"                                \n",
       " [8] \"UKB_white_british_AF\"                  \n",
       " [9] \"UKB_non_british_white_AF\"              \n",
       "[10] \"UKB_african_AF\"                        \n",
       "[11] \"UKB_s_asian_AF\"                        \n",
       "[12] \"UKB_e_asian_AF\"                        \n",
       "[13] \"UKB_related_AF\"                        \n",
       "[14] \"UKB_others_AF\"                         \n",
       "[15] \"UKB_OBS_CT\"                            \n",
       "[16] \"UKB_white_british_OBS_CT\"              \n",
       "[17] \"UKB_non_british_white_OBS_CT\"          \n",
       "[18] \"UKB_african_OBS_CT\"                    \n",
       "[19] \"UKB_s_asian_OBS_CT\"                    \n",
       "[20] \"UKB_e_asian_OBS_CT\"                    \n",
       "[21] \"UKB_related_OBS_CT\"                    \n",
       "[22] \"UKB_others_OBS_CT\"                     \n",
       "[23] \"UKB_UKBB_OBS_CT\"                       \n",
       "[24] \"UKB_UKBL_OBS_CT\"                       \n",
       "[25] \"UKB_MISSING_CT\"                        \n",
       "[26] \"UKB_white_british_MISSING_CT\"          \n",
       "[27] \"UKB_non_british_white_MISSING_CT\"      \n",
       "[28] \"UKB_african_MISSING_CT\"                \n",
       "[29] \"UKB_s_asian_MISSING_CT\"                \n",
       "[30] \"UKB_e_asian_MISSING_CT\"                \n",
       "[31] \"UKB_related_MISSING_CT\"                \n",
       "[32] \"UKB_others_MISSING_CT\"                 \n",
       "[33] \"UKB_UKBB_MISSING_CT\"                   \n",
       "[34] \"UKB_UKBL_MISSING_CT\"                   \n",
       "[35] \"UKB_HOM_REF_CT\"                        \n",
       "[36] \"UKB_white_british_HOM_REF_CT\"          \n",
       "[37] \"UKB_non_british_white_HOM_REF_CT\"      \n",
       "[38] \"UKB_african_HOM_REF_CT\"                \n",
       "[39] \"UKB_s_asian_HOM_REF_CT\"                \n",
       "[40] \"UKB_e_asian_HOM_REF_CT\"                \n",
       "[41] \"UKB_related_HOM_REF_CT\"                \n",
       "[42] \"UKB_others_HOM_REF_CT\"                 \n",
       "[43] \"UKB_HET_REF_ALT_CTS\"                   \n",
       "[44] \"UKB_white_british_HET_REF_ALT_CTS\"     \n",
       "[45] \"UKB_non_british_white_HET_REF_ALT_CTS\" \n",
       "[46] \"UKB_african_HET_REF_ALT_CTS\"           \n",
       "[47] \"UKB_s_asian_HET_REF_ALT_CTS\"           \n",
       "[48] \"UKB_e_asian_HET_REF_ALT_CTS\"           \n",
       "[49] \"UKB_related_HET_REF_ALT_CTS\"           \n",
       "[50] \"UKB_others_HET_REF_ALT_CTS\"            \n",
       "[51] \"UKB_TWO_ALT_GENO_CTS\"                  \n",
       "[52] \"UKB_white_british_TWO_ALT_GENO_CTS\"    \n",
       "[53] \"UKB_non_british_white_TWO_ALT_GENO_CTS\"\n",
       "[54] \"UKB_african_TWO_ALT_GENO_CTS\"          \n",
       "[55] \"UKB_s_asian_TWO_ALT_GENO_CTS\"          \n",
       "[56] \"UKB_e_asian_TWO_ALT_GENO_CTS\"          \n",
       "[57] \"UKB_related_TWO_ALT_GENO_CTS\"          \n",
       "[58] \"UKB_others_TWO_ALT_GENO_CTS\"           \n",
       "[59] \"UKB_HAP_REF_CT\"                        \n",
       "[60] \"UKB_white_british_HAP_REF_CT\"          \n",
       "[61] \"UKB_non_british_white_HAP_REF_CT\"      \n",
       "[62] \"UKB_african_HAP_REF_CT\"                \n",
       "[63] \"UKB_s_asian_HAP_REF_CT\"                \n",
       "[64] \"UKB_e_asian_HAP_REF_CT\"                \n",
       "[65] \"UKB_related_HAP_REF_CT\"                \n",
       "[66] \"UKB_others_HAP_REF_CT\"                 \n",
       "[67] \"UKB_HAP_ALT_CTS\"                       \n",
       "[68] \"UKB_white_british_HAP_ALT_CTS\"         \n",
       "[69] \"UKB_non_british_white_HAP_ALT_CTS\"     \n",
       "[70] \"UKB_african_HAP_ALT_CTS\"               \n",
       "[71] \"UKB_s_asian_HAP_ALT_CTS\"               \n",
       "[72] \"UKB_e_asian_HAP_ALT_CTS\"               \n",
       "[73] \"UKB_related_HAP_ALT_CTS\"               \n",
       "[74] \"UKB_others_HAP_ALT_CTS\"                \n",
       "[75] \"f_miss\"                                \n",
       "[76] \"f_miss_UKBB\"                           \n",
       "[77] \"f_miss_UKBL\"                           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "afreq_df %>% colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvar_f %>% fread_CHROM() -> pvar_df\n",
    "old_cal_annot_f %>% fread_CHROM() -> old_cal_annot_f\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>array</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td> 17536</td></tr>\n",
       "\t<tr><td>1</td><td> 34197</td></tr>\n",
       "\t<tr><td>2</td><td>753693</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " array & n\\\\\n",
       " <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 &  17536\\\\\n",
       "\t 1 &  34197\\\\\n",
       "\t 2 & 753693\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 3 × 2\n",
       "\n",
       "| array &lt;int&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 |  17536 |\n",
       "| 1 |  34197 |\n",
       "| 2 | 753693 |\n",
       "\n"
      ],
      "text/plain": [
       "  array n     \n",
       "1 0      17536\n",
       "2 1      34197\n",
       "3 2     753693"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_qc_df %>% count(array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 4 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>array</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>both</td><td>753693</td></tr>\n",
       "\t<tr><td>UKBB</td><td> 34197</td></tr>\n",
       "\t<tr><td>UKBL</td><td> 17536</td></tr>\n",
       "\t<tr><td>NA  </td><td>275542</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 4 × 2\n",
       "\\begin{tabular}{ll}\n",
       " array & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t both & 753693\\\\\n",
       "\t UKBB &  34197\\\\\n",
       "\t UKBL &  17536\\\\\n",
       "\t NA   & 275542\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 4 × 2\n",
       "\n",
       "| array &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| both | 753693 |\n",
       "| UKBB |  34197 |\n",
       "| UKBL |  17536 |\n",
       "| NA   | 275542 |\n",
       "\n"
      ],
      "text/plain": [
       "  array n     \n",
       "1 both  753693\n",
       "2 UKBB   34197\n",
       "3 UKBL   17536\n",
       "4 NA    275542"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "afreq_df %>% count(array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[ytanigaw@sh02-09n54 /oak/stanford/groups/mrivas/ukbb24983/sqc/population_stratification_w24983_20200828]$ cat ukb24983_master_sqc.20200828.phe | cut -f8 | sort6 | uniq -c\n",
    "      1 genotyping_array\n",
    " 438427 UKBB\n",
    "  49950 UKBL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(colnames(cal_annot_df), colnames(cnv_annot_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'POS_s'</li><li>'POS_e'</li><li>'pvar_order'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'POS\\_s'\n",
       "\\item 'POS\\_e'\n",
       "\\item 'pvar\\_order'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'POS_s'\n",
       "2. 'POS_e'\n",
       "3. 'pvar_order'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"POS_s\"      \"POS_e\"      \"pvar_order\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(colnames(cnv_annot_df), colnames(cal_annot_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_cal_annot_df %>%\n",
    "select(ID, freq, maf) %>%\n",
    "left_join(\n",
    "    afreq_df %>% select(ID, UKB_AF, UKB_white_british_AF), \n",
    "    by='ID'\n",
    ") -> tmp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tmp %>%\n",
    "filter(maf != freq) %>%\n",
    "dim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tmp %>%\n",
    "filter(maf > .5) %>%\n",
    "dim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>101730</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 101730\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 101730\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 101730      5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tmp %>%\n",
    "filter(UKB_AF > .5) %>%\n",
    "dim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ID</th><th scope=col>freq</th><th scope=col>maf</th><th scope=col>UKB_AF</th><th scope=col>UKB_white_british_AF</th><th scope=col>UKB_MAF</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>rs28659788 </td><td>0.03614952</td><td>0.03614952</td><td>0.0368601</td><td>0.0361564</td><td>0.0368601</td></tr>\n",
       "\t<tr><td>rs116587930</td><td>0.05360614</td><td>0.05360614</td><td>0.0509576</td><td>0.0536004</td><td>0.0509576</td></tr>\n",
       "\t<tr><td>rs116720794</td><td>0.03658726</td><td>0.03658726</td><td>0.0375313</td><td>0.0365897</td><td>0.0375313</td></tr>\n",
       "\t<tr><td>rs3131972  </td><td>0.15681154</td><td>0.15681154</td><td>0.8290750</td><td>0.8431890</td><td>0.1709250</td></tr>\n",
       "\t<tr><td>rs12184325 </td><td>0.03652657</td><td>0.03652657</td><td>0.0374347</td><td>0.0365292</td><td>0.0374347</td></tr>\n",
       "\t<tr><td>rs3131962  </td><td>0.12983398</td><td>0.12983398</td><td>0.8600440</td><td>0.8701680</td><td>0.1399560</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " ID & freq & maf & UKB\\_AF & UKB\\_white\\_british\\_AF & UKB\\_MAF\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t rs28659788  & 0.03614952 & 0.03614952 & 0.0368601 & 0.0361564 & 0.0368601\\\\\n",
       "\t rs116587930 & 0.05360614 & 0.05360614 & 0.0509576 & 0.0536004 & 0.0509576\\\\\n",
       "\t rs116720794 & 0.03658726 & 0.03658726 & 0.0375313 & 0.0365897 & 0.0375313\\\\\n",
       "\t rs3131972   & 0.15681154 & 0.15681154 & 0.8290750 & 0.8431890 & 0.1709250\\\\\n",
       "\t rs12184325  & 0.03652657 & 0.03652657 & 0.0374347 & 0.0365292 & 0.0374347\\\\\n",
       "\t rs3131962   & 0.12983398 & 0.12983398 & 0.8600440 & 0.8701680 & 0.1399560\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 6\n",
       "\n",
       "| ID &lt;chr&gt; | freq &lt;dbl&gt; | maf &lt;dbl&gt; | UKB_AF &lt;dbl&gt; | UKB_white_british_AF &lt;dbl&gt; | UKB_MAF &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| rs28659788  | 0.03614952 | 0.03614952 | 0.0368601 | 0.0361564 | 0.0368601 |\n",
       "| rs116587930 | 0.05360614 | 0.05360614 | 0.0509576 | 0.0536004 | 0.0509576 |\n",
       "| rs116720794 | 0.03658726 | 0.03658726 | 0.0375313 | 0.0365897 | 0.0375313 |\n",
       "| rs3131972   | 0.15681154 | 0.15681154 | 0.8290750 | 0.8431890 | 0.1709250 |\n",
       "| rs12184325  | 0.03652657 | 0.03652657 | 0.0374347 | 0.0365292 | 0.0374347 |\n",
       "| rs3131962   | 0.12983398 | 0.12983398 | 0.8600440 | 0.8701680 | 0.1399560 |\n",
       "\n"
      ],
      "text/plain": [
       "  ID          freq       maf        UKB_AF    UKB_white_british_AF UKB_MAF  \n",
       "1 rs28659788  0.03614952 0.03614952 0.0368601 0.0361564            0.0368601\n",
       "2 rs116587930 0.05360614 0.05360614 0.0509576 0.0536004            0.0509576\n",
       "3 rs116720794 0.03658726 0.03658726 0.0375313 0.0365897            0.0375313\n",
       "4 rs3131972   0.15681154 0.15681154 0.8290750 0.8431890            0.1709250\n",
       "5 rs12184325  0.03652657 0.03652657 0.0374347 0.0365292            0.0374347\n",
       "6 rs3131962   0.12983398 0.12983398 0.8600440 0.8701680            0.1399560"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tmp %>%\n",
    "head() %>%\n",
    "mutate(\n",
    "    UKB_MAF = pmin(UKB_AF, 1 - UKB_AF)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Removed 2279 rows containing missing values (geom_point).”\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC/VBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OV\nlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqan\np6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5\nubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrL\ny8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd\n3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v\n7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///8G5K/k\nAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3deZxU1Z338QuydgMtCAiIdtQogks0\nJCogkATHZOIAKmoMyCZqEjWjcRlUkJZhpGfcMCaaaBKyjMk4OhpRHwlkjBPzgA5Gibg8ikAi\ni0KxdTe9L+f11NJVdar6fOtup7nV53w/f1Tdrrr9e10qvnNuVXdXOYIxFjon6gNgzIQIiTEN\nERJjGiIkxjRESIxpiJAY0xAhMaYhQmJMQyEgVe93q67VfR8PNR3QMORAa6OGKftr6nRMqWut\n0jGmWceQA60NOsYcqtUxpb71oI4xWh6Yg6317vvogHQw5lath3081LJXw5C9oknDlFh1nY4p\ndeKAjjGtOobsE406xtTU6phSL/brGKPlgdkvGjzsQ0jBIiQQIRGSnwgJREiE5CdCAhESIfmJ\nkECEREh+IiQQIRGSnwgJREiE5CdCAhESIfmJkECEREh+IiQQIRGSnwgJREiE5CdCAhESIfmJ\nkECEREh+IiQQIRGSnwgJREiE5CdCAhESIfmJkECEREh+IiQQIRGSnwgJREiE5CdCAhESIfmJ\nkECEREh+IiQQIRGSnwgJREiE5CdCAhESIfmJkECEREh+IiQQIRGSnwgJREiE5CdCAhESIfmJ\nkECEREh+IiQQIRWu5oG5M5fuTm1/d2q8ywgpfIQEKiZIL130Q9d9vENatnDrzvuub01uz38+\nlnjcCSl0hAQqHkgv9unRo0fPeS57eYYUm7YlvipdtDH5xaUbcu4jpMAREqhoIL3Su0einr8r\nvJtnSOtmtMUvb3gysd009eEbr1q+I7G5e228HdVuNYpa13081KpjSI1o0TGmrlHHFE0PTJuO\nITWiWceY+gYdU5rEIR1jQj8wO1OOevQ4svB+NV4hrZ6XuFz0WHIFmv3gBx/cPftQfPPlsfFe\nL/ytjHXVGtOOevQpvGNrZssN0vzEZQpSsrrL1sQv//bzeFsPudUk6l338VBrrY4pokXHlIYm\nHVOaRJ2OMW06htTqeWAatTwwzUXxwFRlHPUoKbxnrVdIr6VO7Z7K3nLdr9NbfI4UOD5HAhXH\nc6Sso57fLryn5+dI+6ZtFqJq+juJ7b/+oFmI+steJqTQERKoKCBlHfUoc9nV+8vflTdt3XH3\nzW1izSpRPXPFJzuWz28gpNAREqgYIPXx7sgHpNoVc2Ytj+9+72Ihtiz+xpXLPs3cRUiBIyRQ\nEUDysR7xV4SCR0ggUyBJ69EA/q6dIkICEZKcfF7HX1pVRUggQpKS1yP+9rcyQgIRUjbZEf+M\nQh0hgQgpU44jQlJHSCBCStc3xxEhqSMkECG1l+eIkNQREoiQUuU7IiR1hAQipGSSo/6pWwhJ\nGSGBCCmR5Khf+02EpIyQQIQUUzoiJHWEBCIktSNCUkdIIEKSHZVmbyUkZYQEIiS1I0JSR0gg\n6yEBR4SkjpBAtkNCjghJHSGBLIcEHRGSOkIC2Q2pBDoiJHWEBLIaUmnWUUn+fYSkjJBANkMq\nKeCIkNQREshiSIXWI0ICERLIXkiFHRGSOkICWQtJctRXdT8hKSMkkK2Q3BwRkjpCAlkKqZ+b\nI0JSR0ggOyG5OyIkdYQEshKSB0eEpI6QQDZCcn1+lIiQlBESyEJIXtYjQgIREsg+SAM8OSIk\ndYQEsg5Sf2+OCEkdIYFsgyStR30KDiEkZYQEsgySx/O6GCGBCAlkFyTP6xEhgQgJZBUkH44I\nSR0hgWyC5McRIakjJJBFkMr8OCIkdYQEsgeS5Ki3hyGEpIyQQNZA8umIkNQREsgWSGU9/Tki\nJHWEBLIEkuxoj6chhKSMkEB2QBrodz0iJBAhgayANEh6frTN4xBCUkZIIBsgDZLWo//ndQgh\nKSMkkAWQZEd/8DyEkJQREsh8SLKjF7wPISRlhAQyHpLsaKWPIYSkjJBApkMaIjn6Nz9DCEkZ\nIYEMhyQ7qvA1hJCUERLIbEjBHRGSOkICGQ0phCNCUkdIIJMhDQ7hiJDUERLIYEjSetTLtyNC\nUkdIIHMhDQ3liJDUERLIWEjDwzkiJHWEBDIVUlhHhKSOkECGQgrtiJDUERLITEjhHRGSOkIC\nGQlJgyNCUkdIIBMhyY6WBB1CSMoICWQgJNnRosBTCEkZIYHMg3SsFkeEpI6QQMZBKpccXRdi\nDiEpIySQaZCOkxzNCzPosEGq2utWnYd9PNSyT8OQfaJJw5S9NXU6ptSLgzrGtOoYsl806hhz\nqFbHlAZxIOQEeT2aFmrSAdHgYR8NkJpa3GoTra77eEjoGNIi2nRMadUyxcQHRsu/qC3sP0l2\nNDnkLA8PTLMOSDy1CxxP7UBhT+1yHIU8Fj5HUkZIIJMg6XRESOoICWQQJK2OCEkdIYHMgXSC\nVkeEpI6QQMZAGpN11FODI0JSR0ggUyCNltajSTqOhZCUERLIEEjyejRRywNDSMoICWQGJNnR\neD0PDCEpIySQEZBkR+dqemAISRkhgUyAJDs6LUZIIEICEVIq2dGYxA2EpIyQQISUrIMjQlJH\nSCBCStTRESGpIyQQIcUb39ERIakjJBAhxWITFY4ISR0hgQgpNklyNDpzKyEpIyQQIU1WrUeE\nBCIkkPWQJvfKOjpJup2QlBESyHZI8nlduXwHISkjJJDlkOT1KMcRIakjJJDdkLAjQlJHSCCr\nIRVwREjqCAlkM6RCjghJHSGBLIZ0RSFHhKSOkED2QporOTqu492EpIyQQNZCuqWwI0JSR0gg\nWyEtdnFESOoICWQppEWSo+HKPQhJGSGB7IS0xNURIakjJJCVkCokR8PAPoSkjJBANkKqcF+P\nCAlESCALIXlZjwgJREgg+yDJjo7GuxGSMkICWQdJfp1haIH9CEkZIYFsgySvR0MKjSEkZYQE\nsgxSRW+PjghJHSGB7ILk3REhqSMkkFWQPJ/XxQgJREggmyDJ69FRbmMISRkhgSyC9C/ez+ti\nhAQiJJA9kJZJjga5jyEkZYQEsgbSr/w5IiR1hASyBdILPh0RkjpCAlkCaZNfR4SkjpBAdkDa\nIjka6G0MISkjJJAVkLb7d0RI6ggJZAOkTwM4IiR1hASyANJuyVGZ5zGEpIyQQBZAyjrq4d0R\nIakjJJDxkPYEc0RI6ggJZDykPsEcEZI6QgKZDingekRIIEICGQ5JWo8G+BtDSMoICWQ2pOCO\nCEkdIYGMhtQ3uCNCUkdIIJMhhXFESOoICWQwpBDndTFCAhESyFxIodYjQgIREshYSCEdEZI6\nQgKZCkly1C/QGEJSRkggQyGFdhQ5pEP/cdWZI/sdc+ZV/3GIkNQREkgfpPCOIobUcP8Qp9cZ\nX7nkK2f0cobc30BIqggJpA2SBkfRQto2tvvlL9YmN2tfvLz72G2EpIiQQLogSY5KA4+JFNLA\nL78nKXnvy4MISREhgTRB0uIoWkiLW3KYtCxKXtU8MHfm0t2ZW38/dT0hhY+QQHocRQup8s+J\ny4Y/7M3xtGzh1p33Xd/a/tWB2TMISUOEpK5Ej6NoITk/SFxud56XHcWmbYmvShdtTGP76WxC\n0hAhKdO0HhUjpHUz2uKXNzzZ/tXV9YSkYwohqZLWo5Jwk4oP0up5ictFjyW/qJnzlkhBemdh\nvPcb3GoRTa77eKitUccU0apjSnOLjim6HhgdQxqL5YGRz+tCjtL0wHj4J3mGNF+C9NBDoh3S\ny2PjvS4Y05a8HkV9LJ5rzWy5QHotdWr3VGL7rTnVaUh1O+LF9rtVL6pd9/FQywENQw6IZg1T\n9h+q1zGlQc8D06pjyEHRpGNMbV247y+VHIU+GE0PTKOHfTxC2jdtsxBV099JbN87Y+bMmdMu\nX56+j8+RAsfnSPlJ61Hf8AcT7XOkW9bHW+Xcn7jK3Fp509Ydd9/cJtasEtWJ77xyTRUhhY6Q\n8pLXIw0HEy0kucyttSvmzFoe3/3examv+aqdjimElFs/retRxJAq5ISHCClwhJST7Ah+qrmf\nov4zikxe/oqCkIJHSHKyI/ip5r4qEkivX9OfkECEBAoOSXYEP9XcX8UAad9DpzvOREICERIo\nMKSc50emQGpbe0VvZ8TtH3pxREjBI6RM/XNeZzAD0vZlxzu9/8FZ64kRIYWIkNLlOjIC0jNf\nP8I546G9MUIqFCGBgkEakPe6twmQnIHfeyN+RUgFIyRQIEiSoz7JG0yAVOqc9W87CcklQgIF\ngdTBkRGQqh45yzni60/vIqRCERIoAKSOjoyAFG/Dtf2dI53fEBKOkED+IUmOeqdvMwSSEDWP\nf9Fxxv2khpBAhATyDams43pkEKR4G68rc0oJCURIIL+QlI6MgiRE7cpx8ctH9yvuIiRCUucT\nktqRYZDa79lESB0jJJA/SGWK50eJCEkZIYFsh4QcEZI6QgJZDunInllHu3PuISRlhASyG9JA\nydGnuXcRkjJCAlkNaZDkaHvefYSkjJBANkOSHb2dfychKSMkkMWQ5PO6VzvcS0jKCAlkLyR5\nPfpFx7sJSRkhgayFJDt6VHE/ISkjJJCtkI6SHFWodjAR0ktuv7lKSIGzFNIgN0emQNo7Z2j3\nvHdaJaTcCAnkAdIQV0emQLq02/g5C5IREoiQQO6QPDgyBVLpbV4AEVLobIQkOeqFHJkCqeQ5\nQnKJkEBukLysR8ZAOr+SkFwiJJALJE/rkTGQNn/u2TZCKhghgQpDOtqbIyMglZeXn1ju9ClP\nRkggQgIVhOTVkRGQpsgREoiQQIUgDffqyAhIviOkwNkFSXa0pPAUUyDV7hKibuX9WwgJRUgg\nDMn7emQMpPeHVormLzhO2ZuEBCIkEITkYz0yBtIlp38kfuU88tH4SwkJREggBOlYP45MgTT0\nCSEuPk2IJ44lJBAhgQCk4yRHt7tPMQRSr5dFy8B/EmJNL0ICERJIDUlej67zMMUQSMf+RKxx\nXhbip8MJCURIICWkcsnRTC9TDIG0YNjt5Se2iN1n8DkSipBAKkiyo8mephgCade5zuD1Qnyj\n7C+EBCIkkAKSf0emQBKiqil+seFTL44IKXhWQDrOvyNzIPmJkAJnA6QA65EZkEYtF6MyERKI\nkED5kAI5MgLSOSvEOZkICURIoDxIJwVyZAQk3xFS4IyHNCqYI1MgjX0vdf30aEICERIoB9Lo\nrKOefhyZAsnZkLxqXsrfbEAREkiGNEZyNNHXFCMgOdk+T0ggQgJJkE4Juh4ZAmnj953pyTe1\nu3rJdkICERIoC0lej871OcUISEJ89UMvgAgpdCZDkh2N8TvFEEj+IqTAGQwplCMjIPEHsl4i\nJFA7pHCOjIDEH8h6iZBAKUhnhnNkBCTfEVLgTIU0LqQjUyCNe5GQXCIkUALS+LCOTIE08gFC\ncomQQHFIk0I7MgXSc6OfbSKkghESqKZ2cnhHpkCaeLrTawTf+7tQhASqOU9yNDroFEMgTfgK\n3/vbJUICyevRZwJPMQSSvwgpcOZBmtQr66g8+BhjIP3tmR8/vuoTQoIRkrLJehyZAmn/hcnf\n/e4+8xAhgQhJlS5HpkC60pmx8qWXVl7R7VpCAhGSIm2OTIF05I2p68VHeYF0qNqtRlHruo+H\nWms0DKkRLRqmVNc36pii6YFp0zGkRjSHnHC57CjcqCYP/1V5SMsDc0g0ue5To4bU97ep67Ul\nXiDVudYsGtx3cq9Nx5A60apjSmOzjinF9MDUi5ZwA66SHYU8mBZRH3JCssP3wKghnXdv6vrR\n87xA4qld4Ew6tbtRcnRs2IMx5NTuzyf8V5MQrWtOfoOQQISU152So+GhD8YESKNGjTplpNP7\n+BNLnZHjCAlESLlVaHVkBKQJ2cbxzU9QhJTTEr2OjICU36P74V2EFDJTIOl2ZCQkZxMhdYyQ\npHLO6+CnmvuJkJQREsgMSDmO4Kea+4qQlBESyAhIsqNh8FPN/UVIyggJZAIk2dHR8FPNfUZI\nyggJZAAk2dGQGCERkr8IKVW+I0IiJF8RUrIOjgiJkHxFSIkqeuc7IiRC8hUhxXIdDW6/jZAI\nyU+ElOtoUPpGQsqp/n+fiYnm5OZLNaJwhBS4Lg1J6YiQcmzc399x1os75zW7GCKkcHVlSD9S\nOiIkmcZjzrQfxSH9ose9hASyHtJqtSNCkmmc8W1RH4ck7jiZkEC2Q3oTOCIkmUaftSlIv+tJ\nSCDLIb0rORqYcw8hSQ19PgXpPwcQEshuSLugI0KSaZw/uS4Bad9pFxASyG5I6LwuRkg5kP5w\nxGdvdK6aO6DnnwgJZDWkrKMeZfn3EZLc789KvGXx2a94cURIweuakAo5IqQ8HbvfesvtvRoI\nKXRdElJBR4Qk0xj7Xur66dGEBLIXUmFHhCTTcDYkr5qX9iIkkLWQXBwRksQoG9/XDmUrpD4u\njggpC2Pj953pCxJdvWQ7IYEshSQ5GqDeg5CkvvqhF0CEFLquBsndESH5oUNIdkLq6+6IkDJb\no5Yn3ke/PUIC2QjJiyNCymyds0Kck4mQQBZC8nBeFyMkntr5yz5I0nrUr8BuhERIfrIOkuSo\nf6H9CElq75yh3VM/SCIkkG2QPK5HhJRD5tJu4+ckf5K0gJBAlkHy7IiQZBqlt3kBREih6yqQ\nvDsiJJlGyXOE5JJVkCRHpW5jCEnq/EpCcskmSCXe1yNCyoG0+XPPthFSwSyC5Gc9IqQspPLy\n8hPLnT7lyQgJZA+kEl+OCCmzNUWOkEDWQPK3HhFS7qmdFz+EFL7ih+RzPSKk3L+QPfG656oJ\nqVCWQPLtiJBkGj+4ZLDT80uVb3p7xYGQAlfskCRHJR7HEFJObW8/HMc09EpCAlkBqdS/I0Lq\nCGTnkiH8XTuUDZCCOCKkPES/vnaU0/+CfyEkkAWQ+gVxREgyjatPcgZf9OCGFi+MCClExQxJ\nctTXxxhCknIG3/6eN0SEFKoihhTQESHJNH45/zPO0Mt++A4hwUyHVBrQESHl6dj2s9kjnSGX\nEhLIcEhB1yNCUrw+t3XRYL5qhzIbUnBHhJRjo+3dR2eOdPpecD8hgYyGFMIRIck0Lh7sdDvz\ntrUNXhgRUoiKE1L/EI4ISaZxzNwndme+eNTtc5IIKXBFCSmUI0KCTpxNhNQxcyENCOWIkAjJ\nV8ZCkhz1CTKGkAjJT6ZCCuuIkAjJV4ZCKgvriJAIyVdmQgq9HhESIfnLSEiSo95BxxASIfnJ\nREhlGhwREiH5ykBI4Z8fJSKkwpBqHpg7c2n7T2k/Xjbzm3e+T0jhKyZIR/bMrkd7go8hpMKQ\nli3cuvO+61sTm81zH9yxa8UVdYQUuiKCdKR0Xrc7xBxC6nDGtm578vqlmvhFbNqW+Kp00cbk\nHc/EDe2YuoWQQlc8kAZJ69EnYQYRUqa9jyx9tlX8ssTp9t3MbetmJN6a64Yn019XP/KdJkIK\nXdFAkh19HGoSIaXbNsxxnCl/7Xv+7V9wnkrfuHpe4nLRY6mvWi+ZesfexMarX473RptbQrju\n4iVNU/SM0VKxPDA565GOIwpbsTwwqSnuY1oUkOb1qPzTT4dO+VqbaDrz7zKQ5suQxPa3K69N\nnPFtuDLeX5rdahUtrvt4qE3HkGahZUxLq44pmh4YEfL7ZUevhpzVatIDk6zFw38xTQpIn50d\nv/h35+fxy38ekb7xtdSpXWaFEq3feIGndqErjlO7oyRHK8MeC0/t0pU8FL/Y4qyOX/64e/rG\nfdM2C1E1Pfl2KG9e0xBf7GYRUviKApLs6J7Qx0JIma3H4xcxZ2388vHsrZU3bd1x981tYs0q\nUTP7Xz/+5LEZnxBS6IoB0hDJUUX4YyGkwpBqV8yZtTy++72LhfhrxWWX37pREFLoigCSZkeE\nlN26Y/369audh+OXd/BdhFCmQJIc9dLhiJCyW3KEBDIEUs56BD/V3E+ElK5CjpBAZkDKcQQ/\n1dxXhBQ4QgpcxJByHMFPNfcXIaU7lCXSdA0hgUyANDTn+REhgYJCGn8wvRWbxOdIKAMg5Toi\nJFRQSP3OiqU2Nh3fbTEhgbo+pOF5r9cREigopNcGjt6ZuH6uX/9nvDgipOBFCGlo/uvehAQK\n/GLDpuEnbBOisvtJ73pyREjBiw5S/npESLDgr9ptOX7kW7OcCw92NENI6bo4pI6OCAkV4uXv\nnad267ao1aMjQgpeVJBkR0vabyMkUJifI+07+6hdXh0RUvAigjRCcnRX+kZCAgWFtD7R2vIJ\nyWtCAnVlSCMlR7dmbiUkEH/XThkhlUuOvpW9mZBA/F07ZdZDkh3NkG4nJBB/106Z7ZCOkxxN\nlu8gJBAhKbMcUjlyREiooJBObe+cea8SEqyLQsKOCAkVFFJ5e0c5jqenSIQUvMMNqYAjQkKF\nPrX78O+dNYQE6pKQCjkiJFT450i1x0wjJFBXhDRKcjSpw72EBNLwYsMtQwgJ1AUhnZJ11LOj\nI0JCaYD0YA9CAnU9SGMkRx3O62KEBNMA6R+HERKoy0EaXXg9IiRYeEh7B80gJFBXgyQ7mqDc\ng5BAgT/Wpb33fz2m+zpCAnUxSPJ53enqXQgJFP6XVvv+zIsjQgre4YIkOxoD9iEkUFBI32rv\nxu/v8OSIkIJ3mCB5cURIKP6unTILIXlyREgoQlJmHyRvjggJRUjKrIN0njdHhIQiJGW2QZoo\nORpdaAghgQhJmWWQJntcjwgJRkjK7II02et6REiwcJDqXtpNSAXqGpAmSY5OcRlCSKBwkLY5\nzxJSgboEJHk9OsFtCCGBCEmZRZAm98o6KncdQkggQlJmDyR/jggJRUjKrIHk0xEhocJBanzL\n64dREFLwOhGS7Og4L0MICcSXv5VZAsnvekRIMEJSZgekaX7XI0KCEZIyKyBd53s9IiQYISmz\nAdKdkqORXocQEoiQlFkA6S7J0XDPQwgJpAHSx88TEqiIIS0J5IiQUBogPc4PGkMVL6Rg6xEh\nwQhJmemQgjoiJFRQSK9mW0hIqGKFVBHUESGh+BmyysyGJDs62t8QQgIFhfT5kY+nm0tIqOKE\nJDsa5nMIIYGCQnq/76PpTT5HghUlJNnREL9DCAkU+MWGR/q8TUhuFSOkUI4ICRX8Vbt1H7Vv\nvLqAkEBFCKmidxhHhITibzYoMxZSuPWIkGCEpMxUSCHXI0KCEZIyQyHJjgYFGkJIIEJSZiak\nfw27HhESjJCUGQmpMvR6REgwQlJmIqT/1eCIkFCEpMxASBt0OCIklAZIDzzwk02EpKyIIG2R\nHA0MPoaQQBogHeFc89mHCElV8UDao2U9IiRYUEiVf05cNvxhrxAvPi9a/kZIqooG0qeSoyPD\nDCIkUOA/o/hB4nK74+nPzAkpVOEh7c466hHivC5GSLDDBqnBtRbR5L6Te206hjSIVh1Tmlt0\nTAn/wEiOysJNatT0wDTrmNIiGnWM0fJfTKPw8L+1DkjVB9yqFzWu+3io5aCGIQdFs4YpB2rr\ndUxp8PDgFUxyNCDksVSJppATktVpeWAawz4wqVp1DKkSjR720QCJp3aBC3tq10daj8IeC0/t\nQHyOpMwkSL01OiIkFCEpMwiSzvWIkGCBId2yPt4q5/7EFSGBigCSXkeEhOK7CCkzBlJf6XUG\nHcdCSKCgkCrkCAkUOSTJUT8t/70QEkjHL60eIiRQ1JD6yOsRIYGKBNLr1/QnJFDEkKT1qL+m\n/14ICRQS0r6HTneciYQEihZS3vMjQgJFDqlt7RW9nRG3f+jFESEFLyAk+flR4mtCAkUMafuy\n453e/+Cs9cSIkEIUDFK+I0JCRQrpma8f4Zzx0N4YIRUqQkgdHBESKtqfIw383hvxK0IqWHSQ\nOjoiJFSkkEqds/5tJyG5FBkkhSNCQkUKqeqRs5wjvv70LkIqVFSQJEelmRsJCRT1q3Ybru3v\nHOn8hpBwEUEqUTkiJFTUkISoefyLjjPuJzWEBIoGkuSoRLqZkEDRQ4q38boyp5SQQJFAAo4I\nCVUUkISoXTkufvnofsVdhBQBJOSIkFBFAqn9Hrf3iCSkwPmCBB0REoqQlNkNqRQ6IiQUISmz\nGpLkqG/+fYQEIiRlNkMqsB4REoyQlFkMqdB6REgwQlJmL6TCjggJRUjKrIXk4oiQUISkzFZI\nbo4ICUVIyiyF1M/NESGhCEmZnZAGuDoiJFRRQXrJ7TdXCSlwHiB5cERIqKgh7Z0ztDvfabVg\nhwtSfw+OCAkVNaRLu42fsyAZIYEOEyQv6xEhwaKGVHqbF0CEFDo3SJKjPgV2IyRQ1JBKniMk\nlw4LJG/rESHBooZ0fiUhuXQ4IHlcjwgJFjWkzZ97to2QCnYYIEmOehceQ0igSCGVl5efWO70\nKU9GSKDOh+TdESGhIoU0RY6QQJ0OqczreV2MkGBRn9r5i5AChyGVeV+PCAkWOaTaXULUrbx/\nCyGhOhlSWc+soz2uYwgJFDWk94dWiuYvOE7Zm4QE6lxIR/b0sR4REixqSJec/pH4lfPIR+Mv\nJSRQp0KS16NdHsYQEihqSEOfEOLi04R44lhCAnUmpIGSo91exhASKGpIvV4WLQP/SYg1vQgJ\n1ImQBkmOtnoaQ0igqCEd+xOxxnlZiJ8OJyRQ50GSHb3jbQwhgaKGtGDY7eUntojdZ/A5EqrT\nIMmOnvY4hpBAUUPada4zeL0Q3yj7CyGBOguS7KjC6xhCAkUNSYiqpvjFhk+9OCKk4HWAFMgR\nIaGih+QnQgpcPqTBgRwREipSSKOWi1GZCAnUKZCGBHNESKhIIZ2zQpyTiZBAnQEpqCNCQvHU\nTpnhkAI7IiRU1JDGvpe6fno0IYH0QwruiJBQUUNyNiSvmpfyNxtQ2iFJjnr5dERIqGghOdk+\nT0gg3ZCODuGIkFDRQtr4fWd68k3trl6ynZBAmiENC+OIkFBRn9p99UMvgAgpdGlI4RwREipq\nSP4ipMC1QxoezhEhofgDWWWmQpIdLQkyhpBA/IGsMkMhhXZESCie2ikzE5Ls6K5gYwgJFDWk\ncS8SkkvaII2QHC0MOIaQQFFDGvkAIbmkC9JxkqPbgo4hJFDUkJ4b/WwTIRVME6RyydH8wGMI\nCRQ1pImnO71G8L2/C8WUIi8AABkhSURBVKUHkrweXRh8DCGBooY04Ssd3vu75oG5M5fuTm3v\nu+/Ky2//gJBCJ69Hk0PMISRQ1JAULVu4ded917cmt7+3cMuu+2fVE1LIdDkiJFT0kP72zI8f\nX/VJ9uvYtC3xVemijYnt6uUfC7FnaubXiAgpWNocERIqakj7L0z+7nf3mYfSt6ybkfjksRue\nzOzy/vTE9+57Pd6ug241iEOu+3iotUrDkCrRrGHKwdqGsBMkRz0nhxvVFvZYElWJJh1j6kI/\nMIkaRbWOMVoemGoPD0y1GtKVzoyVL7208opu16ZvWT0vcbnosfTX1detTFy9PDbe66pTQeZS\nznoU9cGwkLVmtnIgHXlj6nrxUelbVs9PXGYgbb/2keRnY255ON5HdW41iwbXfTzUpmNInWjV\nMaWxOdz3nyivR2EPRssDUy9adIxpbNIxpUXU6xhz+B4YNaS+v01dry1J3/Ja6tTuqdRXG2c+\nL+3N50j+GyM7Cn0wfI4Eivo50nn3pq4fPS99y75pm4Womv5O8ot3v/mGvDch+W60VkeEhIoa\n0p9P+K+m+HnfmpOzYCpv2rrj7pvbxJpVovGa3yS+ly9/B05ejyZoOBhCAkX790ijRp0y0ul9\n/ImlzshxmVtrV8yZtTy++72LxcapyV4gpIDJjsbpOBhCAkUKaUK2cXzzE1QISLKjL8BPNfcT\nIYGiPrXL6dH98C5CCpDsaAz8VHNfERKoqCA5mwipY4Eh5TiCn2ruL0ICEZIyEyDlOIKfau4z\nQgIRkjIDIJ2W44iQUISkjJDaGy85Gh0jJBghKSOkVBPzHBESipCUEVKy8yRHpyRvISQQISkj\npESTOzgiJBQhKSOkeBN7ZR2d3H4bIYEISRkhxdcjydFn0jcSEoiQlBFSjqPyzK2EBDIGUv3/\nPhMTzcnNl2oIqWM+IakdERLKFEj393ec9eLOec0uhgjJW8ARIaEMgfSYM+1HcUi/6HEvIYF8\nQUKOCAllCKQzvi3q45DEHScTEsgPJNnRcTn3EBLIEEh91qYg/a4nIYF8QJqJ1iNCghkCaejz\nKUj/OYCQQN4h/RNcjwgJZgik8yfXJSDtO+0CQgJ5hnSn5GhE/p2EBDIE0h+O+OyNzlVzB/T8\nEyGBvEK6S3I0vMO9hAQyBJL4/VmJtyw++xUvjgipQIUdERLKFEhC7H7rLbf3aiAk15YUdkRI\nKEMgjX0vdf30aEICeYJUITkaptqBkECGQHI2JK+al/YiJJAXSLKjo5V7EBLICEhONr6vHcoD\nJHdHhIQyAtLG7zvTFyS6esl2QgK5Q5IdDQH7EBLICEhCfPXDDloIKTdXSBW93R0REsoQSP4i\nJEXyejQY7kVIIBMgjVqeeB/99ggJ5ALJy3ldjJBgJkA6Z4U4JxMhgQpDks/r8HpESDATIPmO\nkPLz9PwoESGBCEmZZZA8OyIklCGQ9s4Z2j31gyRCAhWAJDsaVHgKIYEMgXRpt/Fzkj9JWkBI\nIAzJhyNCQhkCqfQ2L4AISdUTPhwREsoQSCXPEZJLCNI7kqOBrlMICWQIpPMrCcklAGlz1lEP\nd0eEhDIE0ubPPdtGSAVTQ/pEclTmYQohgUyAVF5efmK506c8GSGBlJD2SOd1R3qZQkggEyBN\nkSMkkAqS7MjLekRIMBMgxdvsxQ8h5efzvC5GSDBDIDknXvdcNSEVSgHJvyNCQhkC6QeXDHZ6\nfqnyTW+vOBBSsgCOCAllCCQh2t5+OI5p6JWEBOoAqU8AR4SEMgZSop1LhvB37VD5kCRHA7xP\nISSQMZB2/vraUU7/C/6FkEB5kII5IiSUIZCuPskZfNGDG1q8MCKkWGBHhIQyBJIz+Pb3vCEi\npERBHRESyhBIv5z/GWfoZT98h5BgMqS+QR0REsoQSPG2/Wz2SGfIpYQEkiBJjvr7nEJIIHMg\nxdu6aDBftUNlIUmO+vmdQkggUyC1vfvozJFO3wvuJyRQBlKI9YiQYIZAuniw0+3M29Y2eGFk\nOaQw6xEhwQyBdMzcJ3ZnvnjU7XOSbIYUzhEhoQyBlJOziZA6loIkOSoNMoWQQISkzFRIIdcj\nQoIRkjJDIYVdjwgJRkjKzIRUknVUEnAKIYEISZmRkDQ4IiQUISkzEVJJ6PO6GCHBCEmZgZB0\nrEeEBCMkZeZB0uOIkFCEpMw4SKV6HBESipCUmQZJlyNCQpkE6eC67cnrl2oIKTdtjggJZQSk\nvY8sfbZV/LLE6fZdF0KpmtvcEsJ1Fy9pmhJ2TL+so75RH0v7GD1T9IzRUld7YLLvypCFtG2Y\n4zhT/tr3/Nu/4DzlBZJlK5K0HvUNezBckUAmrEjzelT+6adDp3ytTTSd+XeElF8/bed1MUKC\nmQDps7PjF//u/Dx++c8jCCmvfhrXI0KCmQCp5KH4xRZndfzyx90JKTed53UxQoKZAMl5PH4R\nc9bGLx/nezbkpnc9IiQYISkzBZLsCH6quZ8ICWQEpDvWr1+/2nk4fnkHIclJjvrATzX3FSGB\njIAkR0jZ+svndYQEIqR0FXKElCnHESGhCClwdkAakPs6AyGBCCndoSyRpmsIqT3JUZ/E14QE\nIqR04w+mt2KT+BypvXxHhIQipHT9zoqlNjYd320xISUry3dESChCSvfawNE7E9fP9ev/jBdH\nFkCSHPVuv4mQQISUadPwE7YJUdn9pHc9OTIf0oCOjggJRUjZthw/8q1ZzoUHO5qxEpJiPSIk\nGCFJ7Ty1W7dFrR4dmQ7pyJ5ZR3sytxISiJDk9p191C6vjgyHNEjpiJBQhJRufaK15ROS19ZD\nGiSd1+2UbickECFltvi7dtkGSuvRNvkOQgIRUjr+rl02+bzudzn3EBKIkAJnLiTZ0QO5dxES\niJAIKT/ZUUXefYQEIqR0p7Z3zrxX7YZ0VAFHhIQipHTl7R3lOJ6eIpkKaXAhR4SEIqQOffj3\nzhp7IRVcjwgJRkgdqz1mmrWQCq9HhAQjJEW3DLEV0hAXR4SEIiRFD/awFJLkqJfSESGhCEnR\nPw6zE5K7I0JCEVLH9g6aYSWkYe6OCAlFSOm2tff+r8d0X2cjJC+OCAlFSJmtTH1/5sWRaZCG\nenFESChCSvet9m78/g5PjgyDNFxytATvRkggQgqcUZA8OiIkFCGlm1+Xw6TuKpsgyY7uKjSG\nkECElK78jFckJa+cUW4RJK/rESHBCCnd3gucST9LPT3a8bNJzgV77YF0jORoUeExhAQipEyt\nv/qs4xx96vhTj3ack37l/mZCxkA6TnJ0q8sYQgIRklTLK4svPPuUsy9c/EqLKyNzIMmOLnYb\nQ0ggQgqcIZDKJUeTXccQEoiQLIfkzxEhoQhJavfcEd0tezsun44ICUVIUpf1mDJ3QTJbIPl1\nREgoQpI66rdeABkEybcjQkIRklTJHrsg+XdESChCkpr4B6sgnezfESGhCEnqjbM9/SGSIZAk\nRz29OiIkFCFJTTjWKUm9uZ0FkMYEcURIKEKSmjglnfmQRkuOzvM+hpBAhBS4Lg1JXo/O9TGG\nkECEZCUk2dEYP2MICURI6UYtF6MyGQ4psCNCQhFSunNWiHMymQ0puCNCQhFS4LosJNnRaJ9j\nCAlESLlVH0hmMqQQ6xEhwQhJasuFpeZ/GPP4MI4ICUVIUl8qm3XrwmTmQpoU4rwuRkgwQpIq\n/b9eAHVpSOdJjk4OMIaQQIQkNXSn6ZAmS45OCTKGkECEJHXLMsMhTQ65HhESjJCkGs+fcGtl\nMjMhyc+PTgw2hpBAhCRVmflACiMhTe6VdVQecAwhgQhJaviMP32U+pAkEyHJ53VBHRESipCk\nepv8YoOO9YiQYIQkddbGDlpqHpg7c+nu9i923DK9y0KS16Pjgo8hJBAhSf3PV/6SD2nZwq07\n77s+9Ubgf5yzostCktejY0PMISQQIUlNGOn0y/1T89i0LfFV6aLUSvXfe9Z3VUhf0uSIkFCE\nJNXxT83XzWiLX97wZPuXXRXS7bocERKKkAq2el7ictFjeZA+uCfeh/VuNYtG13081BZ2QIXk\naES4UU3NYQ8mUUuRPDCJGkSLjjG6HpgGHWMO3wODIH36f1b+YvWnEqT5Skgvj433emGDxZPs\naHjUB8NMKvspYjmQDlzWI/HT2G6zDqVveS11avdU+5dpSNXvxdt9wK16UeO6j4daD4b69hxH\nYY+ltiHshEQN6b/7ClerjiFVoknHmNp6HVMai+qBafSwjxLSvJ4LfvHCb3883fl2+pZ90zYL\nUTX9nTxIybrIc6QcR6GPhc+RQHyOJDXwF6nrhUdlbqq8aeuOu29uE2tWifi/cc30xL+1S0HS\n64iQUIQk1fuT1PXLJZmbalfMmbU8vvu9i4VYMDXRc10KkuxomIZjISQQIUl9vv0P+x6ZJDzU\nFSBpXo8ICUZIUr//wqttQrS8eOqfDYGU4wh+qrmfCAlESFLnDnFKTzihr3PsKV7eJLL4IVX0\nzjoaAj/V3FeEBCIkqc+PmyDV5SHlOIKfau4vQgIRUuCKHVKOI/ip5j4jJBAhmQop1xEhoQgJ\nREjJ8hwREoqQQISUKN8RIaEICURIMYUjQkIREoiQch0NSt1ESCBCAhGSyhEhoQgJREirFI4I\nCUVIIOsh7VA5IiQUIYGsh6R0REgoQgLZDkntiJBQhASyHFIftSNCQhESyG5IWUc9ynLuICQQ\nIYGshgQdERKKkEA2Q8KOCAlFSCCLIRVwREgoQgLZC6lvAUeEhCIkkLWQCjoiJBQhgWyFJDka\noLibkECEBLIUkosjQkIREshOSG6OCAlFSCArIbk6IiQUIYFshOTuiJBQhASyEJIHR4SEIiSQ\nfZAkR/3gToQEIiSQdZA8OSIkFCGBbINU4skRIaEICWQZJI+OCAlFSCC7IHl1REgoQgJZBUly\nVFp4CCGBCAlkEyTvjggJRUggiyCVendESChCAtkDyY8jQkIREsgaSJKjEvchhAQiJJAtkHyt\nR4QEIySQJZD8rUeEBCMkkB2Q+vl0REgoQgJZAcm3I0JCERLIBkj+HRESipBAFkCSHPX1OoSQ\nQIQEMh9SEEeEhCIkkPGQ+gdxREgoQgKZDmlAIEeEhCIkkOGQAjoiJBQhgcyGFNQRIaEICWQ0\npMCOCAlFSCCTIQV3REgoQgIZDKksuCNCQhESyFxIYRwREoqQQMZCCuWIkFCEBDIVkuSoT4Ah\nhAQiJJChkEI6IiQUIYHMhCQ56h1oCCGBCAlkJKSBPcOtR4QEIySQiZAGhXZESChCAhkISXIU\n7LwuRkgwQgKZB0l2tCfoEEICERLIOEiyow8DTyEkECGBTIMkO/rn4GMICURIIMMgyY4qQswh\nJBAhgcyCNESTI0JCERLosEGqOehWgzjkuk/hchyFmlQlmkMeS7LaBh1TGjw8eB5q1TGkSjTp\nGFNXr2NKo6jWMUbLA1Pt4YGp1gGp3rVm0ei+U6FyHIUbVS9aQw5I1tSsY0pL2AcmVZuOIQ2i\nRccYXQ9Mg44xh++B0QGp80/ttJ3XxXhqB+OpHcic50g6HRESipBAxkDS6oiQUIQEMgWSXkeE\nhCIkkCGQhut1REgoQgKZAUly1KsCfqq5jwgJREggIyDlOIKfau4nQgIREsgESDmO4Kea+4qQ\nQIQEMgCS7GgJ/FRzfxESiJBAXR9S7npESDBCAhFSonxHhIQiJBAhxTsu97wuRkgwQgIRUixW\nLjlalLqJkECEBCKkHEffab+NkECEBCIk2dHk9I2EBCIkkPWQlI4ICUVIINshqR0REoqQQJZD\nAo4ICUVIILshIUeEhCIkkNWQoCNCQhESyGZIJ2cd9cx1REgoQgJZDGkMXI8ICUZIIHshjcHr\nESHBCAlkLaSCjggJRUggWyHJjiZ1vJuQQIQEshSS7Ohcxf2EBCIkkJ2QZEdjVDsQEoiQQFZC\ncnVESChCAtkIyd0RIaEICWQhJA+OCAlFSCD7IH3OgyNCQhESyDpIk704IiQUIYFsgzS5V9bR\naLwbIYEICWQZJNnRmAL7ERKIkEB2QZIdnVJoDCGBCAlkFSTPjggJRUggmyDJjsoLjyEkECGB\nLILkwxEhoQgJZA8kP44ICUVIIGsg+XJESChCAtkCyZ8jQkIREsgSSLKj4zz8iwgJREggOyD5\ndURIKEICWQGpQnJ0rKd/ESGBCAlkA6QKv+sRIcEICWQBJNnRcI//IkICERLIfEhBHBESipBA\nxkMK5IiQUIQEMh1SMEeEhCIkkOGQZEfDfPyLCAlESCCzIVX0DuaIkFCEBDIaUmBHhIQiJJDJ\nkGRHQ/z9iwgJREgggyGFcERIKEICmQspjCNCQhESyFhIoRwREoqQQKZCCueIkFCEBDIUUkhH\nhIQiJJCZkMI6IiQUIYGMhCQ7OirQv4iQQIQEMhGS7GhQsH8RIYEICWQgpFfDOyIkFCGBDISk\nwREhoQgJZB6kPhocERKKkEDGQdLiiJBQhAQqQkg1D8yduXR3x21PkPQ4IiQUIYGKENKyhVt3\n3nd9a4dtL5A0OSIkFCGBig9SbNqW+Ep00cb8bS+Q+mYc9SgL9S8iJBAhgYoP0roZbfHLG57M\n3/YAabguR4SEIiRQ8UFaPS9xueixvO310+K91VKw0qyjwju6JkJ+f/uUNh1TWrVMaROtOsaY\n+MDomHL4Hphmz5DmS5Cy274ghXVk5H8vhKTMWEivpU7nnsrfTuTx1C7keV2Mp3YwntqBiu/U\nbt+0zUJUTX8nf9sDpNSLDT0HhP8XERKIkEDFB0lU3rR1x903t4k1q7LbHiHFynr36u3jfSBh\nhAQiJFARQqpdMWfW8vju9y7ObnuFVPBTzX1ESCBCAhUhpAIRUuAICURIyggJREggQlJGSCBC\nAhGSMkICERKIkJQREoiQQISkjJBAhAQiJGWEBCIkECEpIyQQIYEISRkhgQgJREjKCAlESCBC\nUkZIIEICEZIyQgIREoiQlBESiJBAhKSMkECEBCIkZYQEIiQQISkjJBAhgQhJGSGBCAlESMoI\nCURIIEJSRkggQgIRkjJCAhESiJCUERKIkECEpIyQQIQEIiRlhAQiJBAhKSMkECGBCEkZIYEI\nCURIyggJREggQlJGSCBCAhGSMkICERKIkJQREoiQQISkjJBAhASyEpJ7r9yzrTPH+6rxniei\nPoRsv7tnZ9SHkOnQPU+573S4euGeWNSHkGnfPc/52LtTIf1w7OudOd5XdWO/E/UhZLt/7Kao\nDyHT3rG3RH0I2ZaO3RL1IWT6eOxdPvYmpCgiJBAhKSMkECGBCEkZIYEICURIjFkdITGmIUJi\nTEOExJiGOgVSzQNzZy7d3XE7knIOYMct0yM8lJxj2XfflZff/kGRHMzHy2Z+8873i+NY4v1+\n6vrojiXnYL47Nd5lHr6pUyAtW7h1533Xt3bYjiT5AP44Z0WkkORj+d7CLbvun1VfFAfTPPfB\nHbtWXFFXDMcS78DsGVFCkg9m/vOxxG9RudcZkGLTtsRZX7QxfzuScg7gv/esjxKSfCzVyz8W\nYs/UD4viYA4+Eze0Y2pkLz3n/VdS+dPZEULKOZhLN3j8rs6AtG5GW/zyhifztyMp7wAihdTh\nwXh/+n6892E+mOpHvtNUHMey7ur6KCHJB9M09eEbr1q+w8N3dQak1fMSl4sey9+OpLwDiBRS\n/oNRfd3KyI4l72BaL5l6x97iOJaaOW+JKCHJB3Nw9oMffHD37EPu39UpkOZnj0TejqS8A4gW\nUu6xbL/2kbbiOZi3K6+tKYpjeeghES2k/P9k6y5b4/5dnQHptdTa+FT+diTlHUCkkHKPZePM\n5yM8lo7/y7R+44ViOJa35lRHC6njf7LX/dr9uzoD0r5pm4Womv5O/nYk5R1ApJByjuXdb74R\n4aHkHsyb1zQI0TYrMkjysdw7Y+bMmdMuXx7VseQczF9/0CxE/WUvu39Xp7z8XXnT1h1339wm\n1qzKbkeWfDD7Y2umJ/4MswiOpfGa3yT+xDLCl7+lg6mZ/a8ff/LYjE+K4ViqEw/LlWuqIjuW\nnIOZueKTHcvnN7h/U6dAql0xZ9by/fH/c1mc3Y4s+WAWJH68NtXPHz522rFsTB7K1MgWgdwH\n5q8Vl11+a3Q/o8g5lkRRntrlHMyWxd+4ctmnHr6JvyLEmIYIiTENERJjGiIkxjRESIxpiJAY\n0xAhMaYhQmJMQ4RkSM2zS/puj/ogLI6QDOkFZ9aq2qgPwuIIyZB+7vwx6kOwOkLqWk08749f\n7DPi3qaFI/pNSfxl+G++2Lf/2N8IMcWJty3qo7M4QupaTRn55T9vv9g5f+mO/xlwoRD/4Vz8\nwgtfc14QH1Q4P9nQGPXRWRwhda2mOBuFeNUZH9+cVSrE8q/E8VT1mCXESufVqI/N6gipazUl\nrkd85Nwav7zVqW6/ceREQoo6QupaTSmPX2xzKuOXC50Douqu0wYccYQzgZCijpC6VnmQJh1x\n5x/f3jSCkCKPkLpWuZA2O9fEN5r7EFLkEVLXKhfSe87S+MbDzrmEFHWE1LXKhdR07DHP/emW\nL32p/8uHCCnaCKlrlfccacO4kqO/VfX84IEfEFK0ERJjGiIkxjRESIxpiJAY0xAhMaYhQmJM\nQ4TEmIYIiTENERJjGiIkxjRESIxpiJAY09D/B4hjIoczkqG1AAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      },
      "text/plain": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "tmp %>%\n",
    "ggplot(aes(x = maf, y = pmin(UKB_white_british_AF, 1 - UKB_white_british_AF))) +\n",
    "geom_point(alpha=.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 6 × 69</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>CHROM</th><th scope=col>POS</th><th scope=col>ID</th><th scope=col>REF</th><th scope=col>ALT</th><th scope=col>UKB_AF</th><th scope=col>UKB_white_british_AF</th><th scope=col>UKB_non_british_white_AF</th><th scope=col>UKB_african_AF</th><th scope=col>UKB_s_asian_AF</th><th scope=col>⋯</th><th scope=col>UKB_related_HAP_REF_CT</th><th scope=col>UKB_others_HAP_REF_CT</th><th scope=col>UKB_HAP_ALT_CTS</th><th scope=col>UKB_white_british_HAP_ALT_CTS</th><th scope=col>UKB_non_british_white_HAP_ALT_CTS</th><th scope=col>UKB_african_HAP_ALT_CTS</th><th scope=col>UKB_s_asian_HAP_ALT_CTS</th><th scope=col>UKB_e_asian_HAP_ALT_CTS</th><th scope=col>UKB_related_HAP_ALT_CTS</th><th scope=col>UKB_others_HAP_ALT_CTS</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>723307</td><td>rs28659788       </td><td>C</td><td>G</td><td>3.68601e-02</td><td>3.61564e-02</td><td>3.91005e-02</td><td>         NA</td><td>       NA</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1</td><td>727841</td><td>rs116587930      </td><td>G</td><td>A</td><td>5.09576e-02</td><td>5.36004e-02</td><td>5.19909e-02</td><td>5.50474e-03</td><td>0.0144177</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1</td><td>729632</td><td>rs116720794      </td><td>C</td><td>T</td><td>3.75313e-02</td><td>3.65897e-02</td><td>4.28314e-02</td><td>2.65188e-02</td><td>0.0463466</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1</td><td>751314</td><td>1:723307-779322_-</td><td>N</td><td>+</td><td>2.07153e-06</td><td>1.50034e-06</td><td>0.00000e+00</td><td>7.77122e-05</td><td>0.0000000</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1</td><td>751315</td><td>1:723307-779322_+</td><td>N</td><td>+</td><td>1.03577e-06</td><td>0.00000e+00</td><td>2.03161e-05</td><td>0.00000e+00</td><td>0.0000000</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1</td><td>752721</td><td>rs3131972        </td><td>A</td><td>G</td><td>8.29075e-01</td><td>8.43189e-01</td><td>8.33515e-01</td><td>2.75031e-01</td><td>0.7735750</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 69\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " CHROM & POS & ID & REF & ALT & UKB\\_AF & UKB\\_white\\_british\\_AF & UKB\\_non\\_british\\_white\\_AF & UKB\\_african\\_AF & UKB\\_s\\_asian\\_AF & ⋯ & UKB\\_related\\_HAP\\_REF\\_CT & UKB\\_others\\_HAP\\_REF\\_CT & UKB\\_HAP\\_ALT\\_CTS & UKB\\_white\\_british\\_HAP\\_ALT\\_CTS & UKB\\_non\\_british\\_white\\_HAP\\_ALT\\_CTS & UKB\\_african\\_HAP\\_ALT\\_CTS & UKB\\_s\\_asian\\_HAP\\_ALT\\_CTS & UKB\\_e\\_asian\\_HAP\\_ALT\\_CTS & UKB\\_related\\_HAP\\_ALT\\_CTS & UKB\\_others\\_HAP\\_ALT\\_CTS\\\\\n",
       " <chr> & <int> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 1 & 723307 & rs28659788        & C & G & 3.68601e-02 & 3.61564e-02 & 3.91005e-02 &          NA &        NA & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t 1 & 727841 & rs116587930       & G & A & 5.09576e-02 & 5.36004e-02 & 5.19909e-02 & 5.50474e-03 & 0.0144177 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t 1 & 729632 & rs116720794       & C & T & 3.75313e-02 & 3.65897e-02 & 4.28314e-02 & 2.65188e-02 & 0.0463466 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t 1 & 751314 & 1:723307-779322\\_- & N & + & 2.07153e-06 & 1.50034e-06 & 0.00000e+00 & 7.77122e-05 & 0.0000000 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t 1 & 751315 & 1:723307-779322\\_+ & N & + & 1.03577e-06 & 0.00000e+00 & 2.03161e-05 & 0.00000e+00 & 0.0000000 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t 1 & 752721 & rs3131972         & A & G & 8.29075e-01 & 8.43189e-01 & 8.33515e-01 & 2.75031e-01 & 0.7735750 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 69\n",
       "\n",
       "| CHROM &lt;chr&gt; | POS &lt;int&gt; | ID &lt;chr&gt; | REF &lt;chr&gt; | ALT &lt;chr&gt; | UKB_AF &lt;dbl&gt; | UKB_white_british_AF &lt;dbl&gt; | UKB_non_british_white_AF &lt;dbl&gt; | UKB_african_AF &lt;dbl&gt; | UKB_s_asian_AF &lt;dbl&gt; | ⋯ ⋯ | UKB_related_HAP_REF_CT &lt;int&gt; | UKB_others_HAP_REF_CT &lt;int&gt; | UKB_HAP_ALT_CTS &lt;int&gt; | UKB_white_british_HAP_ALT_CTS &lt;int&gt; | UKB_non_british_white_HAP_ALT_CTS &lt;int&gt; | UKB_african_HAP_ALT_CTS &lt;int&gt; | UKB_s_asian_HAP_ALT_CTS &lt;int&gt; | UKB_e_asian_HAP_ALT_CTS &lt;int&gt; | UKB_related_HAP_ALT_CTS &lt;int&gt; | UKB_others_HAP_ALT_CTS &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 723307 | rs28659788        | C | G | 3.68601e-02 | 3.61564e-02 | 3.91005e-02 |          NA |        NA | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| 1 | 727841 | rs116587930       | G | A | 5.09576e-02 | 5.36004e-02 | 5.19909e-02 | 5.50474e-03 | 0.0144177 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| 1 | 729632 | rs116720794       | C | T | 3.75313e-02 | 3.65897e-02 | 4.28314e-02 | 2.65188e-02 | 0.0463466 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| 1 | 751314 | 1:723307-779322_- | N | + | 2.07153e-06 | 1.50034e-06 | 0.00000e+00 | 7.77122e-05 | 0.0000000 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| 1 | 751315 | 1:723307-779322_+ | N | + | 1.03577e-06 | 0.00000e+00 | 2.03161e-05 | 0.00000e+00 | 0.0000000 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "| 1 | 752721 | rs3131972         | A | G | 8.29075e-01 | 8.43189e-01 | 8.33515e-01 | 2.75031e-01 | 0.7735750 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  CHROM POS    ID                REF ALT UKB_AF      UKB_white_british_AF\n",
       "1 1     723307 rs28659788        C   G   3.68601e-02 3.61564e-02         \n",
       "2 1     727841 rs116587930       G   A   5.09576e-02 5.36004e-02         \n",
       "3 1     729632 rs116720794       C   T   3.75313e-02 3.65897e-02         \n",
       "4 1     751314 1:723307-779322_- N   +   2.07153e-06 1.50034e-06         \n",
       "5 1     751315 1:723307-779322_+ N   +   1.03577e-06 0.00000e+00         \n",
       "6 1     752721 rs3131972         A   G   8.29075e-01 8.43189e-01         \n",
       "  UKB_non_british_white_AF UKB_african_AF UKB_s_asian_AF ⋯\n",
       "1 3.91005e-02                       NA           NA      ⋯\n",
       "2 5.19909e-02              5.50474e-03    0.0144177      ⋯\n",
       "3 4.28314e-02              2.65188e-02    0.0463466      ⋯\n",
       "4 0.00000e+00              7.77122e-05    0.0000000      ⋯\n",
       "5 2.03161e-05              0.00000e+00    0.0000000      ⋯\n",
       "6 8.33515e-01              2.75031e-01    0.7735750      ⋯\n",
       "  UKB_related_HAP_REF_CT UKB_others_HAP_REF_CT UKB_HAP_ALT_CTS\n",
       "1 0                      0                     0              \n",
       "2 0                      0                     0              \n",
       "3 0                      0                     0              \n",
       "4 0                      0                     0              \n",
       "5 0                      0                     0              \n",
       "6 0                      0                     0              \n",
       "  UKB_white_british_HAP_ALT_CTS UKB_non_british_white_HAP_ALT_CTS\n",
       "1 0                             0                                \n",
       "2 0                             0                                \n",
       "3 0                             0                                \n",
       "4 0                             0                                \n",
       "5 0                             0                                \n",
       "6 0                             0                                \n",
       "  UKB_african_HAP_ALT_CTS UKB_s_asian_HAP_ALT_CTS UKB_e_asian_HAP_ALT_CTS\n",
       "1 0                       0                       0                      \n",
       "2 0                       0                       0                      \n",
       "3 0                       0                       0                      \n",
       "4 0                       0                       0                      \n",
       "5 0                       0                       0                      \n",
       "6 0                       0                       0                      \n",
       "  UKB_related_HAP_ALT_CTS UKB_others_HAP_ALT_CTS\n",
       "1 0                       0                     \n",
       "2 0                       0                     \n",
       "3 0                       0                     \n",
       "4 0                       0                     \n",
       "5 0                       0                     \n",
       "6 0                       0                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "afreq_df %>%\n",
    "head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'consequence_field'</li><li>'f_miss'</li><li>'f_miss_bileve'</li><li>'f_miss_wcsg'</li><li>'freq'</li><li>'hwe_p'</li><li>'maf'</li><li>'ld_indep'</li><li>'wcsg_only'</li><li>'bileve_only'</li><li>'filter'</li><li>'missingness'</li><li>'hwe'</li><li>'mcpi'</li><li>'gnomad_af'</li><li>'mgi'</li><li>'mgi_notes'</li><li>'all_filters'</li><li>'Gene_symbol'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'consequence\\_field'\n",
       "\\item 'f\\_miss'\n",
       "\\item 'f\\_miss\\_bileve'\n",
       "\\item 'f\\_miss\\_wcsg'\n",
       "\\item 'freq'\n",
       "\\item 'hwe\\_p'\n",
       "\\item 'maf'\n",
       "\\item 'ld\\_indep'\n",
       "\\item 'wcsg\\_only'\n",
       "\\item 'bileve\\_only'\n",
       "\\item 'filter'\n",
       "\\item 'missingness'\n",
       "\\item 'hwe'\n",
       "\\item 'mcpi'\n",
       "\\item 'gnomad\\_af'\n",
       "\\item 'mgi'\n",
       "\\item 'mgi\\_notes'\n",
       "\\item 'all\\_filters'\n",
       "\\item 'Gene\\_symbol'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'consequence_field'\n",
       "2. 'f_miss'\n",
       "3. 'f_miss_bileve'\n",
       "4. 'f_miss_wcsg'\n",
       "5. 'freq'\n",
       "6. 'hwe_p'\n",
       "7. 'maf'\n",
       "8. 'ld_indep'\n",
       "9. 'wcsg_only'\n",
       "10. 'bileve_only'\n",
       "11. 'filter'\n",
       "12. 'missingness'\n",
       "13. 'hwe'\n",
       "14. 'mcpi'\n",
       "15. 'gnomad_af'\n",
       "16. 'mgi'\n",
       "17. 'mgi_notes'\n",
       "18. 'all_filters'\n",
       "19. 'Gene_symbol'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"consequence_field\" \"f_miss\"            \"f_miss_bileve\"    \n",
       " [4] \"f_miss_wcsg\"       \"freq\"              \"hwe_p\"            \n",
       " [7] \"maf\"               \"ld_indep\"          \"wcsg_only\"        \n",
       "[10] \"bileve_only\"       \"filter\"            \"missingness\"      \n",
       "[13] \"hwe\"               \"mcpi\"              \"gnomad_af\"        \n",
       "[16] \"mgi\"               \"mgi_notes\"         \"all_filters\"      \n",
       "[19] \"Gene_symbol\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(colnames(old_cal_annot_df), colnames(cal_annot_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- `consequence_field`: the consequence string (from VEP?) that concatenates a lot of fields with `|`\n",
    "- `f_miss`: missing rate\n",
    "- `f_miss_bileve`: missing rate in UKBL array\n",
    "- `f_miss_wcsg`: missing rate in UKBB array\n",
    "- `freq`: allele frequency\n",
    "- `hwe_p`: HWE test p-value\n",
    "- `maf`: MAF\n",
    "- `ld_indep`: LD independence flag\n",
    "- `wcsg_only`\n",
    "- `bileve_only`\n",
    "- `filter`: QC filter? it's always `.`\n",
    "- `missingness`: QC filter - missingness\n",
    "- `hwe`: QC filter - HWE p-value\n",
    "- `mcpi`: QC filter: manual cluster plot inspection\n",
    "- `gnomad_af`: QC filter - maf comparison with gnomAD\n",
    "- `mgi`: QC filter - manual ??\n",
    "- `mgi_notes`: QC filter - notes\n",
    "- `all_filters`: QC filter summary\n",
    "- `Gene_symbol`: the gene symbol\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Allele'</li><li>'IMPACT'</li><li>'SYMBOL'</li><li>'Feature_type'</li><li>'Feature'</li><li>'BIOTYPE'</li><li>'EXON'</li><li>'INTRON'</li><li>'HGVSc'</li><li>'cDNA_position'</li><li>'CDS_position'</li><li>'Protein_position'</li><li>'Amino_acids'</li><li>'Codons'</li><li>'Existing_variation'</li><li>'ALLELE_NUM'</li><li>'DISTANCE'</li><li>'STRAND'</li><li>'FLAGS'</li><li>'VARIANT_CLASS'</li><li>'SYMBOL_SOURCE'</li><li>'HGNC_ID'</li><li>'CANONICAL'</li><li>'MANE'</li><li>'TSL'</li><li>'APPRIS'</li><li>'CCDS'</li><li>'ENSP'</li><li>'SWISSPROT'</li><li>'TREMBL'</li><li>'UNIPARC'</li><li>'GENE_PHENO'</li><li>'SIFT'</li><li>'PolyPhen'</li><li>'DOMAINS'</li><li>'miRNA'</li><li>'HGVS_OFFSET'</li><li>'AF'</li><li>'AFR_AF'</li><li>'AMR_AF'</li><li>'EAS_AF'</li><li>'EUR_AF'</li><li>'SAS_AF'</li><li>'AA_AF'</li><li>'EA_AF'</li><li>'gnomAD_AF'</li><li>'gnomAD_AFR_AF'</li><li>'gnomAD_AMR_AF'</li><li>'gnomAD_ASJ_AF'</li><li>'gnomAD_EAS_AF'</li><li>'gnomAD_FIN_AF'</li><li>'gnomAD_NFE_AF'</li><li>'gnomAD_OTH_AF'</li><li>'gnomAD_SAS_AF'</li><li>'MAX_AF'</li><li>'MAX_AF_POPS'</li><li>'CLIN_SIG'</li><li>'SOMATIC'</li><li>'PHENO'</li><li>'PUBMED'</li><li>'VAR_SYNONYMS'</li><li>'MOTIF_NAME'</li><li>'MOTIF_POS'</li><li>'HIGH_INF_POS'</li><li>'MOTIF_SCORE_CHANGE'</li><li>'TRANSCRIPTION_FACTORS'</li><li>'Csq'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Allele'\n",
       "\\item 'IMPACT'\n",
       "\\item 'SYMBOL'\n",
       "\\item 'Feature\\_type'\n",
       "\\item 'Feature'\n",
       "\\item 'BIOTYPE'\n",
       "\\item 'EXON'\n",
       "\\item 'INTRON'\n",
       "\\item 'HGVSc'\n",
       "\\item 'cDNA\\_position'\n",
       "\\item 'CDS\\_position'\n",
       "\\item 'Protein\\_position'\n",
       "\\item 'Amino\\_acids'\n",
       "\\item 'Codons'\n",
       "\\item 'Existing\\_variation'\n",
       "\\item 'ALLELE\\_NUM'\n",
       "\\item 'DISTANCE'\n",
       "\\item 'STRAND'\n",
       "\\item 'FLAGS'\n",
       "\\item 'VARIANT\\_CLASS'\n",
       "\\item 'SYMBOL\\_SOURCE'\n",
       "\\item 'HGNC\\_ID'\n",
       "\\item 'CANONICAL'\n",
       "\\item 'MANE'\n",
       "\\item 'TSL'\n",
       "\\item 'APPRIS'\n",
       "\\item 'CCDS'\n",
       "\\item 'ENSP'\n",
       "\\item 'SWISSPROT'\n",
       "\\item 'TREMBL'\n",
       "\\item 'UNIPARC'\n",
       "\\item 'GENE\\_PHENO'\n",
       "\\item 'SIFT'\n",
       "\\item 'PolyPhen'\n",
       "\\item 'DOMAINS'\n",
       "\\item 'miRNA'\n",
       "\\item 'HGVS\\_OFFSET'\n",
       "\\item 'AF'\n",
       "\\item 'AFR\\_AF'\n",
       "\\item 'AMR\\_AF'\n",
       "\\item 'EAS\\_AF'\n",
       "\\item 'EUR\\_AF'\n",
       "\\item 'SAS\\_AF'\n",
       "\\item 'AA\\_AF'\n",
       "\\item 'EA\\_AF'\n",
       "\\item 'gnomAD\\_AF'\n",
       "\\item 'gnomAD\\_AFR\\_AF'\n",
       "\\item 'gnomAD\\_AMR\\_AF'\n",
       "\\item 'gnomAD\\_ASJ\\_AF'\n",
       "\\item 'gnomAD\\_EAS\\_AF'\n",
       "\\item 'gnomAD\\_FIN\\_AF'\n",
       "\\item 'gnomAD\\_NFE\\_AF'\n",
       "\\item 'gnomAD\\_OTH\\_AF'\n",
       "\\item 'gnomAD\\_SAS\\_AF'\n",
       "\\item 'MAX\\_AF'\n",
       "\\item 'MAX\\_AF\\_POPS'\n",
       "\\item 'CLIN\\_SIG'\n",
       "\\item 'SOMATIC'\n",
       "\\item 'PHENO'\n",
       "\\item 'PUBMED'\n",
       "\\item 'VAR\\_SYNONYMS'\n",
       "\\item 'MOTIF\\_NAME'\n",
       "\\item 'MOTIF\\_POS'\n",
       "\\item 'HIGH\\_INF\\_POS'\n",
       "\\item 'MOTIF\\_SCORE\\_CHANGE'\n",
       "\\item 'TRANSCRIPTION\\_FACTORS'\n",
       "\\item 'Csq'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Allele'\n",
       "2. 'IMPACT'\n",
       "3. 'SYMBOL'\n",
       "4. 'Feature_type'\n",
       "5. 'Feature'\n",
       "6. 'BIOTYPE'\n",
       "7. 'EXON'\n",
       "8. 'INTRON'\n",
       "9. 'HGVSc'\n",
       "10. 'cDNA_position'\n",
       "11. 'CDS_position'\n",
       "12. 'Protein_position'\n",
       "13. 'Amino_acids'\n",
       "14. 'Codons'\n",
       "15. 'Existing_variation'\n",
       "16. 'ALLELE_NUM'\n",
       "17. 'DISTANCE'\n",
       "18. 'STRAND'\n",
       "19. 'FLAGS'\n",
       "20. 'VARIANT_CLASS'\n",
       "21. 'SYMBOL_SOURCE'\n",
       "22. 'HGNC_ID'\n",
       "23. 'CANONICAL'\n",
       "24. 'MANE'\n",
       "25. 'TSL'\n",
       "26. 'APPRIS'\n",
       "27. 'CCDS'\n",
       "28. 'ENSP'\n",
       "29. 'SWISSPROT'\n",
       "30. 'TREMBL'\n",
       "31. 'UNIPARC'\n",
       "32. 'GENE_PHENO'\n",
       "33. 'SIFT'\n",
       "34. 'PolyPhen'\n",
       "35. 'DOMAINS'\n",
       "36. 'miRNA'\n",
       "37. 'HGVS_OFFSET'\n",
       "38. 'AF'\n",
       "39. 'AFR_AF'\n",
       "40. 'AMR_AF'\n",
       "41. 'EAS_AF'\n",
       "42. 'EUR_AF'\n",
       "43. 'SAS_AF'\n",
       "44. 'AA_AF'\n",
       "45. 'EA_AF'\n",
       "46. 'gnomAD_AF'\n",
       "47. 'gnomAD_AFR_AF'\n",
       "48. 'gnomAD_AMR_AF'\n",
       "49. 'gnomAD_ASJ_AF'\n",
       "50. 'gnomAD_EAS_AF'\n",
       "51. 'gnomAD_FIN_AF'\n",
       "52. 'gnomAD_NFE_AF'\n",
       "53. 'gnomAD_OTH_AF'\n",
       "54. 'gnomAD_SAS_AF'\n",
       "55. 'MAX_AF'\n",
       "56. 'MAX_AF_POPS'\n",
       "57. 'CLIN_SIG'\n",
       "58. 'SOMATIC'\n",
       "59. 'PHENO'\n",
       "60. 'PUBMED'\n",
       "61. 'VAR_SYNONYMS'\n",
       "62. 'MOTIF_NAME'\n",
       "63. 'MOTIF_POS'\n",
       "64. 'HIGH_INF_POS'\n",
       "65. 'MOTIF_SCORE_CHANGE'\n",
       "66. 'TRANSCRIPTION_FACTORS'\n",
       "67. 'Csq'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Allele\"                \"IMPACT\"                \"SYMBOL\"               \n",
       " [4] \"Feature_type\"          \"Feature\"               \"BIOTYPE\"              \n",
       " [7] \"EXON\"                  \"INTRON\"                \"HGVSc\"                \n",
       "[10] \"cDNA_position\"         \"CDS_position\"          \"Protein_position\"     \n",
       "[13] \"Amino_acids\"           \"Codons\"                \"Existing_variation\"   \n",
       "[16] \"ALLELE_NUM\"            \"DISTANCE\"              \"STRAND\"               \n",
       "[19] \"FLAGS\"                 \"VARIANT_CLASS\"         \"SYMBOL_SOURCE\"        \n",
       "[22] \"HGNC_ID\"               \"CANONICAL\"             \"MANE\"                 \n",
       "[25] \"TSL\"                   \"APPRIS\"                \"CCDS\"                 \n",
       "[28] \"ENSP\"                  \"SWISSPROT\"             \"TREMBL\"               \n",
       "[31] \"UNIPARC\"               \"GENE_PHENO\"            \"SIFT\"                 \n",
       "[34] \"PolyPhen\"              \"DOMAINS\"               \"miRNA\"                \n",
       "[37] \"HGVS_OFFSET\"           \"AF\"                    \"AFR_AF\"               \n",
       "[40] \"AMR_AF\"                \"EAS_AF\"                \"EUR_AF\"               \n",
       "[43] \"SAS_AF\"                \"AA_AF\"                 \"EA_AF\"                \n",
       "[46] \"gnomAD_AF\"             \"gnomAD_AFR_AF\"         \"gnomAD_AMR_AF\"        \n",
       "[49] \"gnomAD_ASJ_AF\"         \"gnomAD_EAS_AF\"         \"gnomAD_FIN_AF\"        \n",
       "[52] \"gnomAD_NFE_AF\"         \"gnomAD_OTH_AF\"         \"gnomAD_SAS_AF\"        \n",
       "[55] \"MAX_AF\"                \"MAX_AF_POPS\"           \"CLIN_SIG\"             \n",
       "[58] \"SOMATIC\"               \"PHENO\"                 \"PUBMED\"               \n",
       "[61] \"VAR_SYNONYMS\"          \"MOTIF_NAME\"            \"MOTIF_POS\"            \n",
       "[64] \"HIGH_INF_POS\"          \"MOTIF_SCORE_CHANGE\"    \"TRANSCRIPTION_FACTORS\"\n",
       "[67] \"Csq\"                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(colnames(cal_annot_df), colnames(old_cal_annot_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>CHROM %in% c(\"X\", \"Y\", \"MT\", \"XY\")</th><th scope=col>is.na(all_filters)</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>FALSE</td><td>FALSE</td><td>784256</td></tr>\n",
       "\t<tr><td> TRUE</td><td> TRUE</td><td> 19813</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " CHROM \\%in\\% c(\"X\", \"Y\", \"MT\", \"XY\") & is.na(all\\_filters) & n\\\\\n",
       " <lgl> & <lgl> & <int>\\\\\n",
       "\\hline\n",
       "\t FALSE & FALSE & 784256\\\\\n",
       "\t  TRUE &  TRUE &  19813\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 3\n",
       "\n",
       "| CHROM %in% c(\"X\", \"Y\", \"MT\", \"XY\") &lt;lgl&gt; | is.na(all_filters) &lt;lgl&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| FALSE | FALSE | 784256 |\n",
       "|  TRUE |  TRUE |  19813 |\n",
       "\n"
      ],
      "text/plain": [
       "  CHROM %in% c(\"X\", \"Y\", \"MT\", \"XY\") is.na(all_filters) n     \n",
       "1 FALSE                              FALSE              784256\n",
       "2  TRUE                               TRUE               19813"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_cal_annot_df %>%\n",
    "count(CHROM %in% c('X', 'Y', 'MT', 'XY'), is.na(all_filters))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 26 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>all_filters</th><th scope=col>missingness</th><th scope=col>hwe</th><th scope=col>mcpi</th><th scope=col>gnomad_af</th><th scope=col>mgi</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>    </td><td>       </td><td>651983</td></tr>\n",
       "\t<tr><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>    </td><td>NOT_PTV</td><td>    29</td></tr>\n",
       "\t<tr><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>    </td><td>PASS   </td><td>    53</td></tr>\n",
       "\t<tr><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>PASS</td><td>       </td><td>  3589</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 0</td><td> 0</td><td>    </td><td>FAIL   </td><td>    41</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 0</td><td> 0</td><td>FAIL</td><td>       </td><td>   166</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 0</td><td> 1</td><td>    </td><td>       </td><td>     5</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 0</td><td> 1</td><td>PASS</td><td>       </td><td>     5</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 1</td><td> 0</td><td>    </td><td>       </td><td> 34200</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 1</td><td> 0</td><td>    </td><td>NOT_PTV</td><td>     3</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 1</td><td> 0</td><td>    </td><td>PASS   </td><td>     1</td></tr>\n",
       "\t<tr><td> 1</td><td> 0</td><td> 1</td><td> 0</td><td>PASS</td><td>       </td><td>    58</td></tr>\n",
       "\t<tr><td> 1</td><td> 1</td><td> 0</td><td> 0</td><td>    </td><td>       </td><td> 71495</td></tr>\n",
       "\t<tr><td> 1</td><td> 1</td><td> 0</td><td> 0</td><td>    </td><td>NOT_PTV</td><td>     1</td></tr>\n",
       "\t<tr><td> 1</td><td> 1</td><td> 0</td><td> 0</td><td>    </td><td>PASS   </td><td>     1</td></tr>\n",
       "\t<tr><td> 1</td><td> 1</td><td> 0</td><td> 0</td><td>PASS</td><td>       </td><td>   271</td></tr>\n",
       "\t<tr><td> 2</td><td> 0</td><td> 1</td><td> 0</td><td>    </td><td>FAIL   </td><td>     3</td></tr>\n",
       "\t<tr><td> 2</td><td> 0</td><td> 1</td><td> 0</td><td>FAIL</td><td>       </td><td>     4</td></tr>\n",
       "\t<tr><td> 2</td><td> 1</td><td> 0</td><td> 0</td><td>FAIL</td><td>       </td><td>    38</td></tr>\n",
       "\t<tr><td> 2</td><td> 1</td><td> 0</td><td> 1</td><td>PASS</td><td>       </td><td>     1</td></tr>\n",
       "\t<tr><td> 2</td><td> 1</td><td> 1</td><td> 0</td><td>    </td><td>       </td><td> 22258</td></tr>\n",
       "\t<tr><td> 2</td><td> 1</td><td> 1</td><td> 0</td><td>PASS</td><td>       </td><td>    42</td></tr>\n",
       "\t<tr><td> 3</td><td> 1</td><td> 0</td><td> 1</td><td>FAIL</td><td>       </td><td>     2</td></tr>\n",
       "\t<tr><td> 3</td><td> 1</td><td> 1</td><td> 0</td><td>    </td><td>FAIL   </td><td>     2</td></tr>\n",
       "\t<tr><td> 3</td><td> 1</td><td> 1</td><td> 0</td><td>FAIL</td><td>       </td><td>     5</td></tr>\n",
       "\t<tr><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>    </td><td>       </td><td> 19813</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 26 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " all\\_filters & missingness & hwe & mcpi & gnomad\\_af & mgi & n\\\\\n",
       " <int> & <int> & <int> & <int> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t  0 &  0 &  0 &  0 &      &         & 651983\\\\\n",
       "\t  0 &  0 &  0 &  0 &      & NOT\\_PTV &     29\\\\\n",
       "\t  0 &  0 &  0 &  0 &      & PASS    &     53\\\\\n",
       "\t  0 &  0 &  0 &  0 & PASS &         &   3589\\\\\n",
       "\t  1 &  0 &  0 &  0 &      & FAIL    &     41\\\\\n",
       "\t  1 &  0 &  0 &  0 & FAIL &         &    166\\\\\n",
       "\t  1 &  0 &  0 &  1 &      &         &      5\\\\\n",
       "\t  1 &  0 &  0 &  1 & PASS &         &      5\\\\\n",
       "\t  1 &  0 &  1 &  0 &      &         &  34200\\\\\n",
       "\t  1 &  0 &  1 &  0 &      & NOT\\_PTV &      3\\\\\n",
       "\t  1 &  0 &  1 &  0 &      & PASS    &      1\\\\\n",
       "\t  1 &  0 &  1 &  0 & PASS &         &     58\\\\\n",
       "\t  1 &  1 &  0 &  0 &      &         &  71495\\\\\n",
       "\t  1 &  1 &  0 &  0 &      & NOT\\_PTV &      1\\\\\n",
       "\t  1 &  1 &  0 &  0 &      & PASS    &      1\\\\\n",
       "\t  1 &  1 &  0 &  0 & PASS &         &    271\\\\\n",
       "\t  2 &  0 &  1 &  0 &      & FAIL    &      3\\\\\n",
       "\t  2 &  0 &  1 &  0 & FAIL &         &      4\\\\\n",
       "\t  2 &  1 &  0 &  0 & FAIL &         &     38\\\\\n",
       "\t  2 &  1 &  0 &  1 & PASS &         &      1\\\\\n",
       "\t  2 &  1 &  1 &  0 &      &         &  22258\\\\\n",
       "\t  2 &  1 &  1 &  0 & PASS &         &     42\\\\\n",
       "\t  3 &  1 &  0 &  1 & FAIL &         &      2\\\\\n",
       "\t  3 &  1 &  1 &  0 &      & FAIL    &      2\\\\\n",
       "\t  3 &  1 &  1 &  0 & FAIL &         &      5\\\\\n",
       "\t NA & NA & NA & NA &      &         &  19813\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 26 × 7\n",
       "\n",
       "| all_filters &lt;int&gt; | missingness &lt;int&gt; | hwe &lt;int&gt; | mcpi &lt;int&gt; | gnomad_af &lt;chr&gt; | mgi &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "|  0 |  0 |  0 |  0 | <!----> | <!----> | 651983 |\n",
       "|  0 |  0 |  0 |  0 | <!----> | NOT_PTV |     29 |\n",
       "|  0 |  0 |  0 |  0 | <!----> | PASS    |     53 |\n",
       "|  0 |  0 |  0 |  0 | PASS | <!----> |   3589 |\n",
       "|  1 |  0 |  0 |  0 | <!----> | FAIL    |     41 |\n",
       "|  1 |  0 |  0 |  0 | FAIL | <!----> |    166 |\n",
       "|  1 |  0 |  0 |  1 | <!----> | <!----> |      5 |\n",
       "|  1 |  0 |  0 |  1 | PASS | <!----> |      5 |\n",
       "|  1 |  0 |  1 |  0 | <!----> | <!----> |  34200 |\n",
       "|  1 |  0 |  1 |  0 | <!----> | NOT_PTV |      3 |\n",
       "|  1 |  0 |  1 |  0 | <!----> | PASS    |      1 |\n",
       "|  1 |  0 |  1 |  0 | PASS | <!----> |     58 |\n",
       "|  1 |  1 |  0 |  0 | <!----> | <!----> |  71495 |\n",
       "|  1 |  1 |  0 |  0 | <!----> | NOT_PTV |      1 |\n",
       "|  1 |  1 |  0 |  0 | <!----> | PASS    |      1 |\n",
       "|  1 |  1 |  0 |  0 | PASS | <!----> |    271 |\n",
       "|  2 |  0 |  1 |  0 | <!----> | FAIL    |      3 |\n",
       "|  2 |  0 |  1 |  0 | FAIL | <!----> |      4 |\n",
       "|  2 |  1 |  0 |  0 | FAIL | <!----> |     38 |\n",
       "|  2 |  1 |  0 |  1 | PASS | <!----> |      1 |\n",
       "|  2 |  1 |  1 |  0 | <!----> | <!----> |  22258 |\n",
       "|  2 |  1 |  1 |  0 | PASS | <!----> |     42 |\n",
       "|  3 |  1 |  0 |  1 | FAIL | <!----> |      2 |\n",
       "|  3 |  1 |  1 |  0 | <!----> | FAIL    |      2 |\n",
       "|  3 |  1 |  1 |  0 | FAIL | <!----> |      5 |\n",
       "| NA | NA | NA | NA | <!----> | <!----> |  19813 |\n",
       "\n"
      ],
      "text/plain": [
       "   all_filters missingness hwe mcpi gnomad_af mgi     n     \n",
       "1   0           0           0   0                     651983\n",
       "2   0           0           0   0             NOT_PTV     29\n",
       "3   0           0           0   0             PASS        53\n",
       "4   0           0           0   0   PASS                3589\n",
       "5   1           0           0   0             FAIL        41\n",
       "6   1           0           0   0   FAIL                 166\n",
       "7   1           0           0   1                          5\n",
       "8   1           0           0   1   PASS                   5\n",
       "9   1           0           1   0                      34200\n",
       "10  1           0           1   0             NOT_PTV      3\n",
       "11  1           0           1   0             PASS         1\n",
       "12  1           0           1   0   PASS                  58\n",
       "13  1           1           0   0                      71495\n",
       "14  1           1           0   0             NOT_PTV      1\n",
       "15  1           1           0   0             PASS         1\n",
       "16  1           1           0   0   PASS                 271\n",
       "17  2           0           1   0             FAIL         3\n",
       "18  2           0           1   0   FAIL                   4\n",
       "19  2           1           0   0   FAIL                  38\n",
       "20  2           1           0   1   PASS                   1\n",
       "21  2           1           1   0                      22258\n",
       "22  2           1           1   0   PASS                  42\n",
       "23  3           1           0   1   FAIL                   2\n",
       "24  3           1           1   0             FAIL         2\n",
       "25  3           1           1   0   FAIL                   5\n",
       "26 NA          NA          NA  NA                      19813"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_cal_annot_df %>%\n",
    "count(all_filters, missingness, hwe, mcpi, gnomad_af, mgi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 15 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>all_filters</th><th scope=col>FILTER</th><th scope=col>filter</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td> 0</td><td>                          </td><td>NA</td><td>655654</td></tr>\n",
       "\t<tr><td> 1</td><td>gnomad_af                 </td><td>NA</td><td>   166</td></tr>\n",
       "\t<tr><td> 1</td><td>hwe                       </td><td>NA</td><td> 34262</td></tr>\n",
       "\t<tr><td> 1</td><td>mcpi                      </td><td>NA</td><td>    10</td></tr>\n",
       "\t<tr><td> 1</td><td>mgi                       </td><td>NA</td><td>    41</td></tr>\n",
       "\t<tr><td> 1</td><td>missingness               </td><td>NA</td><td> 71768</td></tr>\n",
       "\t<tr><td> 2</td><td>hwe;gnomad_af             </td><td>NA</td><td>     4</td></tr>\n",
       "\t<tr><td> 2</td><td>hwe;mgi                   </td><td>NA</td><td>     3</td></tr>\n",
       "\t<tr><td> 2</td><td>missingness;gnomad_af     </td><td>NA</td><td>    38</td></tr>\n",
       "\t<tr><td> 2</td><td>missingness;hwe           </td><td>NA</td><td> 22300</td></tr>\n",
       "\t<tr><td> 2</td><td>missingness;mcpi          </td><td>NA</td><td>     1</td></tr>\n",
       "\t<tr><td> 3</td><td>missingness;hwe;gnomad_af </td><td>NA</td><td>     5</td></tr>\n",
       "\t<tr><td> 3</td><td>missingness;hwe;mgi       </td><td>NA</td><td>     2</td></tr>\n",
       "\t<tr><td> 3</td><td>missingness;mcpi;gnomad_af</td><td>NA</td><td>     2</td></tr>\n",
       "\t<tr><td>NA</td><td>                          </td><td>NA</td><td> 19813</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 15 × 4\n",
       "\\begin{tabular}{llll}\n",
       " all\\_filters & FILTER & filter & n\\\\\n",
       " <int> & <chr> & <lgl> & <int>\\\\\n",
       "\\hline\n",
       "\t  0 &                            & NA & 655654\\\\\n",
       "\t  1 & gnomad\\_af                  & NA &    166\\\\\n",
       "\t  1 & hwe                        & NA &  34262\\\\\n",
       "\t  1 & mcpi                       & NA &     10\\\\\n",
       "\t  1 & mgi                        & NA &     41\\\\\n",
       "\t  1 & missingness                & NA &  71768\\\\\n",
       "\t  2 & hwe;gnomad\\_af              & NA &      4\\\\\n",
       "\t  2 & hwe;mgi                    & NA &      3\\\\\n",
       "\t  2 & missingness;gnomad\\_af      & NA &     38\\\\\n",
       "\t  2 & missingness;hwe            & NA &  22300\\\\\n",
       "\t  2 & missingness;mcpi           & NA &      1\\\\\n",
       "\t  3 & missingness;hwe;gnomad\\_af  & NA &      5\\\\\n",
       "\t  3 & missingness;hwe;mgi        & NA &      2\\\\\n",
       "\t  3 & missingness;mcpi;gnomad\\_af & NA &      2\\\\\n",
       "\t NA &                            & NA &  19813\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 15 × 4\n",
       "\n",
       "| all_filters &lt;int&gt; | FILTER &lt;chr&gt; | filter &lt;lgl&gt; | n &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "|  0 | <!----> | NA | 655654 |\n",
       "|  1 | gnomad_af                  | NA |    166 |\n",
       "|  1 | hwe                        | NA |  34262 |\n",
       "|  1 | mcpi                       | NA |     10 |\n",
       "|  1 | mgi                        | NA |     41 |\n",
       "|  1 | missingness                | NA |  71768 |\n",
       "|  2 | hwe;gnomad_af              | NA |      4 |\n",
       "|  2 | hwe;mgi                    | NA |      3 |\n",
       "|  2 | missingness;gnomad_af      | NA |     38 |\n",
       "|  2 | missingness;hwe            | NA |  22300 |\n",
       "|  2 | missingness;mcpi           | NA |      1 |\n",
       "|  3 | missingness;hwe;gnomad_af  | NA |      5 |\n",
       "|  3 | missingness;hwe;mgi        | NA |      2 |\n",
       "|  3 | missingness;mcpi;gnomad_af | NA |      2 |\n",
       "| NA | <!----> | NA |  19813 |\n",
       "\n"
      ],
      "text/plain": [
       "   all_filters FILTER                     filter n     \n",
       "1   0                                     NA     655654\n",
       "2   1          gnomad_af                  NA        166\n",
       "3   1          hwe                        NA      34262\n",
       "4   1          mcpi                       NA         10\n",
       "5   1          mgi                        NA         41\n",
       "6   1          missingness                NA      71768\n",
       "7   2          hwe;gnomad_af              NA          4\n",
       "8   2          hwe;mgi                    NA          3\n",
       "9   2          missingness;gnomad_af      NA         38\n",
       "10  2          missingness;hwe            NA      22300\n",
       "11  2          missingness;mcpi           NA          1\n",
       "12  3          missingness;hwe;gnomad_af  NA          5\n",
       "13  3          missingness;hwe;mgi        NA          2\n",
       "14  3          missingness;mcpi;gnomad_af NA          2\n",
       "15 NA                                     NA      19813"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_cal_annot_df %>%\n",
    "mutate(\n",
    "    FILTER = str_replace(paste0(\n",
    "        if_else(is.na(missingness) | missingness == 0, '', 'missingness;'),\n",
    "        if_else(is.na(hwe) | hwe == 0, '', 'hwe;'),\n",
    "        if_else(is.na(mcpi) | mcpi == 0, '', 'mcpi;'),\n",
    "        if_else(is.na(gnomad_af) | gnomad_af %in% c('', 'PASS'), '', 'gnomad_af;'),\n",
    "        if_else(is.na(mgi) | mgi %in% c('', 'PASS', 'NOT_PTV'), '', 'mgi;')\n",
    "    ), ';$', '')\n",
    ") %>%\n",
    "count(all_filters, FILTER, filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 38 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>mgi_notes</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>                                                                 </td><td>803983</td></tr>\n",
       "\t<tr><td>does not match anything in gnomad; possibly typing another allele</td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing 11:704604 GA / G (rs769840755)                     </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing different PTV                                      </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing frameshift indel 1bp away                          </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing inframe deletion                                   </td><td>     2</td></tr>\n",
       "\t<tr><td>likely typing missense                                           </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing missense variant                                   </td><td>     4</td></tr>\n",
       "\t<tr><td>likely typing neighboring frameshift indel 1:145586671           </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing non-PTV                                            </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing overlapping missense allele                        </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing shifted frameshift indel                           </td><td>     3</td></tr>\n",
       "\t<tr><td>likely typing splice region variant                              </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing start lost variant instead of frameshift indel     </td><td>     1</td></tr>\n",
       "\t<tr><td>likely typing synonymous                                         </td><td>     8</td></tr>\n",
       "\t<tr><td>likely typing synonymous or missense                             </td><td>     2</td></tr>\n",
       "\t<tr><td>likely typing synonymous variant                                 </td><td>     9</td></tr>\n",
       "\t<tr><td>likely typing synonymous variant, maybe PTV                      </td><td>     1</td></tr>\n",
       "\t<tr><td>matches                                                          </td><td>     1</td></tr>\n",
       "\t<tr><td>missense                                                         </td><td>     1</td></tr>\n",
       "\t<tr><td>multi-allelic                                                    </td><td>     3</td></tr>\n",
       "\t<tr><td>multi-allelic splice                                             </td><td>     1</td></tr>\n",
       "\t<tr><td>no match                                                         </td><td>    24</td></tr>\n",
       "\t<tr><td>no match ; likely typing synonymous variant                      </td><td>     2</td></tr>\n",
       "\t<tr><td>no match, large allele                                           </td><td>     1</td></tr>\n",
       "\t<tr><td>no match, maybe typing synonymous                                </td><td>     1</td></tr>\n",
       "\t<tr><td>no match, multiple alleles nearby                                </td><td>     1</td></tr>\n",
       "\t<tr><td>no match, typing neighbor                                        </td><td>     1</td></tr>\n",
       "\t<tr><td>no match; likely typing synonymous variant                       </td><td>     1</td></tr>\n",
       "\t<tr><td>no match; likely typing the synonymous variant                   </td><td>     2</td></tr>\n",
       "\t<tr><td>no match; no common PTVs in gene symbol                          </td><td>     1</td></tr>\n",
       "\t<tr><td>no match; typing wrong allele                                    </td><td>     1</td></tr>\n",
       "\t<tr><td>synonymous                                                       </td><td>     1</td></tr>\n",
       "\t<tr><td>Two PTVs                                                         </td><td>     1</td></tr>\n",
       "\t<tr><td>typing synonymous                                                </td><td>     1</td></tr>\n",
       "\t<tr><td>typing wrong allele                                              </td><td>     1</td></tr>\n",
       "\t<tr><td>wrong alleles but likely typing PTV                              </td><td>     1</td></tr>\n",
       "\t<tr><td>wrong alleles, complex alleles                                   </td><td>     1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 38 × 2\n",
       "\\begin{tabular}{ll}\n",
       " mgi\\_notes & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t                                                                   & 803983\\\\\n",
       "\t does not match anything in gnomad; possibly typing another allele &      1\\\\\n",
       "\t likely typing 11:704604 GA / G (rs769840755)                      &      1\\\\\n",
       "\t likely typing different PTV                                       &      1\\\\\n",
       "\t likely typing frameshift indel 1bp away                           &      1\\\\\n",
       "\t likely typing inframe deletion                                    &      2\\\\\n",
       "\t likely typing missense                                            &      1\\\\\n",
       "\t likely typing missense variant                                    &      4\\\\\n",
       "\t likely typing neighboring frameshift indel 1:145586671            &      1\\\\\n",
       "\t likely typing non-PTV                                             &      1\\\\\n",
       "\t likely typing overlapping missense allele                         &      1\\\\\n",
       "\t likely typing shifted frameshift indel                            &      3\\\\\n",
       "\t likely typing splice region variant                               &      1\\\\\n",
       "\t likely typing start lost variant instead of frameshift indel      &      1\\\\\n",
       "\t likely typing synonymous                                          &      8\\\\\n",
       "\t likely typing synonymous or missense                              &      2\\\\\n",
       "\t likely typing synonymous variant                                  &      9\\\\\n",
       "\t likely typing synonymous variant, maybe PTV                       &      1\\\\\n",
       "\t matches                                                           &      1\\\\\n",
       "\t missense                                                          &      1\\\\\n",
       "\t multi-allelic                                                     &      3\\\\\n",
       "\t multi-allelic splice                                              &      1\\\\\n",
       "\t no match                                                          &     24\\\\\n",
       "\t no match ; likely typing synonymous variant                       &      2\\\\\n",
       "\t no match, large allele                                            &      1\\\\\n",
       "\t no match, maybe typing synonymous                                 &      1\\\\\n",
       "\t no match, multiple alleles nearby                                 &      1\\\\\n",
       "\t no match, typing neighbor                                         &      1\\\\\n",
       "\t no match; likely typing synonymous variant                        &      1\\\\\n",
       "\t no match; likely typing the synonymous variant                    &      2\\\\\n",
       "\t no match; no common PTVs in gene symbol                           &      1\\\\\n",
       "\t no match; typing wrong allele                                     &      1\\\\\n",
       "\t synonymous                                                        &      1\\\\\n",
       "\t Two PTVs                                                          &      1\\\\\n",
       "\t typing synonymous                                                 &      1\\\\\n",
       "\t typing wrong allele                                               &      1\\\\\n",
       "\t wrong alleles but likely typing PTV                               &      1\\\\\n",
       "\t wrong alleles, complex alleles                                    &      1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 38 × 2\n",
       "\n",
       "| mgi_notes &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| <!----> | 803983 |\n",
       "| does not match anything in gnomad; possibly typing another allele |      1 |\n",
       "| likely typing 11:704604 GA / G (rs769840755)                      |      1 |\n",
       "| likely typing different PTV                                       |      1 |\n",
       "| likely typing frameshift indel 1bp away                           |      1 |\n",
       "| likely typing inframe deletion                                    |      2 |\n",
       "| likely typing missense                                            |      1 |\n",
       "| likely typing missense variant                                    |      4 |\n",
       "| likely typing neighboring frameshift indel 1:145586671            |      1 |\n",
       "| likely typing non-PTV                                             |      1 |\n",
       "| likely typing overlapping missense allele                         |      1 |\n",
       "| likely typing shifted frameshift indel                            |      3 |\n",
       "| likely typing splice region variant                               |      1 |\n",
       "| likely typing start lost variant instead of frameshift indel      |      1 |\n",
       "| likely typing synonymous                                          |      8 |\n",
       "| likely typing synonymous or missense                              |      2 |\n",
       "| likely typing synonymous variant                                  |      9 |\n",
       "| likely typing synonymous variant, maybe PTV                       |      1 |\n",
       "| matches                                                           |      1 |\n",
       "| missense                                                          |      1 |\n",
       "| multi-allelic                                                     |      3 |\n",
       "| multi-allelic splice                                              |      1 |\n",
       "| no match                                                          |     24 |\n",
       "| no match ; likely typing synonymous variant                       |      2 |\n",
       "| no match, large allele                                            |      1 |\n",
       "| no match, maybe typing synonymous                                 |      1 |\n",
       "| no match, multiple alleles nearby                                 |      1 |\n",
       "| no match, typing neighbor                                         |      1 |\n",
       "| no match; likely typing synonymous variant                        |      1 |\n",
       "| no match; likely typing the synonymous variant                    |      2 |\n",
       "| no match; no common PTVs in gene symbol                           |      1 |\n",
       "| no match; typing wrong allele                                     |      1 |\n",
       "| synonymous                                                        |      1 |\n",
       "| Two PTVs                                                          |      1 |\n",
       "| typing synonymous                                                 |      1 |\n",
       "| typing wrong allele                                               |      1 |\n",
       "| wrong alleles but likely typing PTV                               |      1 |\n",
       "| wrong alleles, complex alleles                                    |      1 |\n",
       "\n"
      ],
      "text/plain": [
       "   mgi_notes                                                         n     \n",
       "1                                                                    803983\n",
       "2  does not match anything in gnomad; possibly typing another allele      1\n",
       "3  likely typing 11:704604 GA / G (rs769840755)                           1\n",
       "4  likely typing different PTV                                            1\n",
       "5  likely typing frameshift indel 1bp away                                1\n",
       "6  likely typing inframe deletion                                         2\n",
       "7  likely typing missense                                                 1\n",
       "8  likely typing missense variant                                         4\n",
       "9  likely typing neighboring frameshift indel 1:145586671                 1\n",
       "10 likely typing non-PTV                                                  1\n",
       "11 likely typing overlapping missense allele                              1\n",
       "12 likely typing shifted frameshift indel                                 3\n",
       "13 likely typing splice region variant                                    1\n",
       "14 likely typing start lost variant instead of frameshift indel           1\n",
       "15 likely typing synonymous                                               8\n",
       "16 likely typing synonymous or missense                                   2\n",
       "17 likely typing synonymous variant                                       9\n",
       "18 likely typing synonymous variant, maybe PTV                            1\n",
       "19 matches                                                                1\n",
       "20 missense                                                               1\n",
       "21 multi-allelic                                                          3\n",
       "22 multi-allelic splice                                                   1\n",
       "23 no match                                                              24\n",
       "24 no match ; likely typing synonymous variant                            2\n",
       "25 no match, large allele                                                 1\n",
       "26 no match, maybe typing synonymous                                      1\n",
       "27 no match, multiple alleles nearby                                      1\n",
       "28 no match, typing neighbor                                              1\n",
       "29 no match; likely typing synonymous variant                             1\n",
       "30 no match; likely typing the synonymous variant                         2\n",
       "31 no match; no common PTVs in gene symbol                                1\n",
       "32 no match; typing wrong allele                                          1\n",
       "33 synonymous                                                             1\n",
       "34 Two PTVs                                                               1\n",
       "35 typing synonymous                                                      1\n",
       "36 typing wrong allele                                                    1\n",
       "37 wrong alleles but likely typing PTV                                    1\n",
       "38 wrong alleles, complex alleles                                         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>%\n",
    "count(mgi_notes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 4 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>wcsg_only</th><th scope=col>bileve_only</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>FALSE</td><td>FALSE</td><td>732880</td></tr>\n",
       "\t<tr><td>FALSE</td><td> TRUE</td><td> 17103</td></tr>\n",
       "\t<tr><td> TRUE</td><td>FALSE</td><td> 34273</td></tr>\n",
       "\t<tr><td>   NA</td><td>   NA</td><td> 19813</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 4 × 3\n",
       "\\begin{tabular}{lll}\n",
       " wcsg\\_only & bileve\\_only & n\\\\\n",
       " <lgl> & <lgl> & <int>\\\\\n",
       "\\hline\n",
       "\t FALSE & FALSE & 732880\\\\\n",
       "\t FALSE &  TRUE &  17103\\\\\n",
       "\t  TRUE & FALSE &  34273\\\\\n",
       "\t    NA &    NA &  19813\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 4 × 3\n",
       "\n",
       "| wcsg_only &lt;lgl&gt; | bileve_only &lt;lgl&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| FALSE | FALSE | 732880 |\n",
       "| FALSE |  TRUE |  17103 |\n",
       "|  TRUE | FALSE |  34273 |\n",
       "|    NA |    NA |  19813 |\n",
       "\n"
      ],
      "text/plain": [
       "  wcsg_only bileve_only n     \n",
       "1 FALSE     FALSE       732880\n",
       "2 FALSE      TRUE        17103\n",
       "3  TRUE     FALSE        34273\n",
       "4    NA        NA        19813"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>%\n",
    "count(wcsg_only, bileve_only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>array</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td> 17536</td></tr>\n",
       "\t<tr><td>1</td><td> 34197</td></tr>\n",
       "\t<tr><td>2</td><td>753693</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " array & n\\\\\n",
       " <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 &  17536\\\\\n",
       "\t 1 &  34197\\\\\n",
       "\t 2 & 753693\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 3 × 2\n",
       "\n",
       "| array &lt;int&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 |  17536 |\n",
       "| 1 |  34197 |\n",
       "| 2 | 753693 |\n",
       "\n"
      ],
      "text/plain": [
       "  array n     \n",
       "1 0      17536\n",
       "2 1      34197\n",
       "3 2     753693"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_qc_df %>% count(array)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 25 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>CHROM</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1 </td><td>63487</td></tr>\n",
       "\t<tr><td>10</td><td>38308</td></tr>\n",
       "\t<tr><td>11</td><td>40824</td></tr>\n",
       "\t<tr><td>12</td><td>37302</td></tr>\n",
       "\t<tr><td>13</td><td>26806</td></tr>\n",
       "\t<tr><td>14</td><td>25509</td></tr>\n",
       "\t<tr><td>15</td><td>24467</td></tr>\n",
       "\t<tr><td>16</td><td>28960</td></tr>\n",
       "\t<tr><td>17</td><td>28835</td></tr>\n",
       "\t<tr><td>18</td><td>21962</td></tr>\n",
       "\t<tr><td>19</td><td>26186</td></tr>\n",
       "\t<tr><td>2 </td><td>61966</td></tr>\n",
       "\t<tr><td>20</td><td>19959</td></tr>\n",
       "\t<tr><td>21</td><td>11342</td></tr>\n",
       "\t<tr><td>22</td><td>12968</td></tr>\n",
       "\t<tr><td>3 </td><td>52300</td></tr>\n",
       "\t<tr><td>4 </td><td>47443</td></tr>\n",
       "\t<tr><td>5 </td><td>46314</td></tr>\n",
       "\t<tr><td>6 </td><td>53695</td></tr>\n",
       "\t<tr><td>7 </td><td>42722</td></tr>\n",
       "\t<tr><td>8 </td><td>38591</td></tr>\n",
       "\t<tr><td>9 </td><td>34310</td></tr>\n",
       "\t<tr><td>MT</td><td>  265</td></tr>\n",
       "\t<tr><td>X </td><td>18857</td></tr>\n",
       "\t<tr><td>Y </td><td>  691</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 25 × 2\n",
       "\\begin{tabular}{ll}\n",
       " CHROM & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 1  & 63487\\\\\n",
       "\t 10 & 38308\\\\\n",
       "\t 11 & 40824\\\\\n",
       "\t 12 & 37302\\\\\n",
       "\t 13 & 26806\\\\\n",
       "\t 14 & 25509\\\\\n",
       "\t 15 & 24467\\\\\n",
       "\t 16 & 28960\\\\\n",
       "\t 17 & 28835\\\\\n",
       "\t 18 & 21962\\\\\n",
       "\t 19 & 26186\\\\\n",
       "\t 2  & 61966\\\\\n",
       "\t 20 & 19959\\\\\n",
       "\t 21 & 11342\\\\\n",
       "\t 22 & 12968\\\\\n",
       "\t 3  & 52300\\\\\n",
       "\t 4  & 47443\\\\\n",
       "\t 5  & 46314\\\\\n",
       "\t 6  & 53695\\\\\n",
       "\t 7  & 42722\\\\\n",
       "\t 8  & 38591\\\\\n",
       "\t 9  & 34310\\\\\n",
       "\t MT &   265\\\\\n",
       "\t X  & 18857\\\\\n",
       "\t Y  &   691\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 25 × 2\n",
       "\n",
       "| CHROM &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| 1  | 63487 |\n",
       "| 10 | 38308 |\n",
       "| 11 | 40824 |\n",
       "| 12 | 37302 |\n",
       "| 13 | 26806 |\n",
       "| 14 | 25509 |\n",
       "| 15 | 24467 |\n",
       "| 16 | 28960 |\n",
       "| 17 | 28835 |\n",
       "| 18 | 21962 |\n",
       "| 19 | 26186 |\n",
       "| 2  | 61966 |\n",
       "| 20 | 19959 |\n",
       "| 21 | 11342 |\n",
       "| 22 | 12968 |\n",
       "| 3  | 52300 |\n",
       "| 4  | 47443 |\n",
       "| 5  | 46314 |\n",
       "| 6  | 53695 |\n",
       "| 7  | 42722 |\n",
       "| 8  | 38591 |\n",
       "| 9  | 34310 |\n",
       "| MT |   265 |\n",
       "| X  | 18857 |\n",
       "| Y  |   691 |\n",
       "\n"
      ],
      "text/plain": [
       "   CHROM n    \n",
       "1  1     63487\n",
       "2  10    38308\n",
       "3  11    40824\n",
       "4  12    37302\n",
       "5  13    26806\n",
       "6  14    25509\n",
       "7  15    24467\n",
       "8  16    28960\n",
       "9  17    28835\n",
       "10 18    21962\n",
       "11 19    26186\n",
       "12 2     61966\n",
       "13 20    19959\n",
       "14 21    11342\n",
       "15 22    12968\n",
       "16 3     52300\n",
       "17 4     47443\n",
       "18 5     46314\n",
       "19 6     53695\n",
       "20 7     42722\n",
       "21 8     38591\n",
       "22 9     34310\n",
       "23 MT      265\n",
       "24 X     18857\n",
       "25 Y       691"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>% count(CHROM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 26 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chromosome</th><th scope=col>CHROM</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1 </td><td>1 </td><td>63487</td></tr>\n",
       "\t<tr><td>10</td><td>10</td><td>38308</td></tr>\n",
       "\t<tr><td>11</td><td>11</td><td>40824</td></tr>\n",
       "\t<tr><td>12</td><td>12</td><td>37302</td></tr>\n",
       "\t<tr><td>13</td><td>13</td><td>26806</td></tr>\n",
       "\t<tr><td>14</td><td>14</td><td>25509</td></tr>\n",
       "\t<tr><td>15</td><td>15</td><td>24467</td></tr>\n",
       "\t<tr><td>16</td><td>16</td><td>28960</td></tr>\n",
       "\t<tr><td>17</td><td>17</td><td>28835</td></tr>\n",
       "\t<tr><td>18</td><td>18</td><td>21962</td></tr>\n",
       "\t<tr><td>19</td><td>19</td><td>26186</td></tr>\n",
       "\t<tr><td>2 </td><td>2 </td><td>61966</td></tr>\n",
       "\t<tr><td>20</td><td>20</td><td>19959</td></tr>\n",
       "\t<tr><td>21</td><td>21</td><td>11342</td></tr>\n",
       "\t<tr><td>22</td><td>22</td><td>12968</td></tr>\n",
       "\t<tr><td>23</td><td>X </td><td>18857</td></tr>\n",
       "\t<tr><td>24</td><td>Y </td><td>  691</td></tr>\n",
       "\t<tr><td>25</td><td>X </td><td> 1357</td></tr>\n",
       "\t<tr><td>26</td><td>MT</td><td>  265</td></tr>\n",
       "\t<tr><td>3 </td><td>3 </td><td>52300</td></tr>\n",
       "\t<tr><td>4 </td><td>4 </td><td>47443</td></tr>\n",
       "\t<tr><td>5 </td><td>5 </td><td>46314</td></tr>\n",
       "\t<tr><td>6 </td><td>6 </td><td>53695</td></tr>\n",
       "\t<tr><td>7 </td><td>7 </td><td>42722</td></tr>\n",
       "\t<tr><td>8 </td><td>8 </td><td>38591</td></tr>\n",
       "\t<tr><td>9 </td><td>9 </td><td>34310</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 26 × 3\n",
       "\\begin{tabular}{lll}\n",
       " chromosome & CHROM & n\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 1  & 1  & 63487\\\\\n",
       "\t 10 & 10 & 38308\\\\\n",
       "\t 11 & 11 & 40824\\\\\n",
       "\t 12 & 12 & 37302\\\\\n",
       "\t 13 & 13 & 26806\\\\\n",
       "\t 14 & 14 & 25509\\\\\n",
       "\t 15 & 15 & 24467\\\\\n",
       "\t 16 & 16 & 28960\\\\\n",
       "\t 17 & 17 & 28835\\\\\n",
       "\t 18 & 18 & 21962\\\\\n",
       "\t 19 & 19 & 26186\\\\\n",
       "\t 2  & 2  & 61966\\\\\n",
       "\t 20 & 20 & 19959\\\\\n",
       "\t 21 & 21 & 11342\\\\\n",
       "\t 22 & 22 & 12968\\\\\n",
       "\t 23 & X  & 18857\\\\\n",
       "\t 24 & Y  &   691\\\\\n",
       "\t 25 & X  &  1357\\\\\n",
       "\t 26 & MT &   265\\\\\n",
       "\t 3  & 3  & 52300\\\\\n",
       "\t 4  & 4  & 47443\\\\\n",
       "\t 5  & 5  & 46314\\\\\n",
       "\t 6  & 6  & 53695\\\\\n",
       "\t 7  & 7  & 42722\\\\\n",
       "\t 8  & 8  & 38591\\\\\n",
       "\t 9  & 9  & 34310\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 26 × 3\n",
       "\n",
       "| chromosome &lt;chr&gt; | CHROM &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 1  | 1  | 63487 |\n",
       "| 10 | 10 | 38308 |\n",
       "| 11 | 11 | 40824 |\n",
       "| 12 | 12 | 37302 |\n",
       "| 13 | 13 | 26806 |\n",
       "| 14 | 14 | 25509 |\n",
       "| 15 | 15 | 24467 |\n",
       "| 16 | 16 | 28960 |\n",
       "| 17 | 17 | 28835 |\n",
       "| 18 | 18 | 21962 |\n",
       "| 19 | 19 | 26186 |\n",
       "| 2  | 2  | 61966 |\n",
       "| 20 | 20 | 19959 |\n",
       "| 21 | 21 | 11342 |\n",
       "| 22 | 22 | 12968 |\n",
       "| 23 | X  | 18857 |\n",
       "| 24 | Y  |   691 |\n",
       "| 25 | X  |  1357 |\n",
       "| 26 | MT |   265 |\n",
       "| 3  | 3  | 52300 |\n",
       "| 4  | 4  | 47443 |\n",
       "| 5  | 5  | 46314 |\n",
       "| 6  | 6  | 53695 |\n",
       "| 7  | 7  | 42722 |\n",
       "| 8  | 8  | 38591 |\n",
       "| 9  | 9  | 34310 |\n",
       "\n"
      ],
      "text/plain": [
       "   chromosome CHROM n    \n",
       "1  1          1     63487\n",
       "2  10         10    38308\n",
       "3  11         11    40824\n",
       "4  12         12    37302\n",
       "5  13         13    26806\n",
       "6  14         14    25509\n",
       "7  15         15    24467\n",
       "8  16         16    28960\n",
       "9  17         17    28835\n",
       "10 18         18    21962\n",
       "11 19         19    26186\n",
       "12 2          2     61966\n",
       "13 20         20    19959\n",
       "14 21         21    11342\n",
       "15 22         22    12968\n",
       "16 23         X     18857\n",
       "17 24         Y       691\n",
       "18 25         X      1357\n",
       "19 26         MT      265\n",
       "20 3          3     52300\n",
       "21 4          4     47443\n",
       "22 5          5     46314\n",
       "23 6          6     53695\n",
       "24 7          7     42722\n",
       "25 8          8     38591\n",
       "26 9          9     34310"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_qc_df %>% \n",
    "mutate(CHROM = if_else(chromosome == '23', 'X', if_else(chromosome == '24', 'Y', if_else(chromosome == '25', 'X', if_else(chromosome == '26', 'MT', chromosome))))) %>%\n",
    "count(chromosome, CHROM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 1 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>CHROM</th><th scope=col>POS</th><th scope=col>REF</th><th scope=col>ALT</th><th scope=col>ID</th><th scope=col>Gene</th><th scope=col>Consequence</th><th scope=col>HGVSp</th><th scope=col>LoF</th><th scope=col>LoF_filter</th><th scope=col>⋯</th><th scope=col>bileve_only</th><th scope=col>filter</th><th scope=col>missingness</th><th scope=col>hwe</th><th scope=col>mcpi</th><th scope=col>gnomad_af</th><th scope=col>mgi</th><th scope=col>mgi_notes</th><th scope=col>all_filters</th><th scope=col>Gene_symbol</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>MT</td><td>73</td><td>A</td><td>G</td><td>rs869183622</td><td>ENSG00000210100,ENSG00000210112,ENSG00000210077,ENSG00000211459,ENSG00000209082,ENSG00000198763,ENSG00000210082,ENSG00000210049,ENSG00000198888</td><td>upstream_gene_variant</td><td></td><td></td><td></td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td></td><td></td><td></td><td>NA</td><td>MT-ND1,MT-ND2,MT-TL1,MT-TV,MT-RNR1,MT-RNR2,MT-TM,MT-TI,MT-TF</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 1 × 31\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " CHROM & POS & REF & ALT & ID & Gene & Consequence & HGVSp & LoF & LoF\\_filter & ⋯ & bileve\\_only & filter & missingness & hwe & mcpi & gnomad\\_af & mgi & mgi\\_notes & all\\_filters & Gene\\_symbol\\\\\n",
       " <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t MT & 73 & A & G & rs869183622 & ENSG00000210100,ENSG00000210112,ENSG00000210077,ENSG00000211459,ENSG00000209082,ENSG00000198763,ENSG00000210082,ENSG00000210049,ENSG00000198888 & upstream\\_gene\\_variant &  &  &  & ⋯ & NA & NA & NA & NA & NA &  &  &  & NA & MT-ND1,MT-ND2,MT-TL1,MT-TV,MT-RNR1,MT-RNR2,MT-TM,MT-TI,MT-TF\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 1 × 31\n",
       "\n",
       "| CHROM &lt;chr&gt; | POS &lt;int&gt; | REF &lt;chr&gt; | ALT &lt;chr&gt; | ID &lt;chr&gt; | Gene &lt;chr&gt; | Consequence &lt;chr&gt; | HGVSp &lt;chr&gt; | LoF &lt;chr&gt; | LoF_filter &lt;chr&gt; | ⋯ ⋯ | bileve_only &lt;lgl&gt; | filter &lt;lgl&gt; | missingness &lt;int&gt; | hwe &lt;int&gt; | mcpi &lt;int&gt; | gnomad_af &lt;chr&gt; | mgi &lt;chr&gt; | mgi_notes &lt;chr&gt; | all_filters &lt;int&gt; | Gene_symbol &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| MT | 73 | A | G | rs869183622 | ENSG00000210100,ENSG00000210112,ENSG00000210077,ENSG00000211459,ENSG00000209082,ENSG00000198763,ENSG00000210082,ENSG00000210049,ENSG00000198888 | upstream_gene_variant | <!----> | <!----> | <!----> | ⋯ | NA | NA | NA | NA | NA | <!----> | <!----> | <!----> | NA | MT-ND1,MT-ND2,MT-TL1,MT-TV,MT-RNR1,MT-RNR2,MT-TM,MT-TI,MT-TF |\n",
       "\n"
      ],
      "text/plain": [
       "  CHROM POS REF ALT ID         \n",
       "1 MT    73  A   G   rs869183622\n",
       "  Gene                                                                                                                                           \n",
       "1 ENSG00000210100,ENSG00000210112,ENSG00000210077,ENSG00000211459,ENSG00000209082,ENSG00000198763,ENSG00000210082,ENSG00000210049,ENSG00000198888\n",
       "  Consequence           HGVSp LoF LoF_filter ⋯ bileve_only filter missingness\n",
       "1 upstream_gene_variant                      ⋯ NA          NA     NA         \n",
       "  hwe mcpi gnomad_af mgi mgi_notes all_filters\n",
       "1 NA  NA                           NA         \n",
       "  Gene_symbol                                                 \n",
       "1 MT-ND1,MT-ND2,MT-TL1,MT-TV,MT-RNR1,MT-RNR2,MT-TM,MT-TI,MT-TF"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df%>% filter(ID == 'rs869183622') %>% head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'rs869183622'</li><li>'rs62581312'</li><li>'rs41323649'</li><li>'rs3937037'</li><li>''</li><li>'rs2853515'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'rs869183622'\n",
       "\\item 'rs62581312'\n",
       "\\item 'rs41323649'\n",
       "\\item 'rs3937037'\n",
       "\\item ''\n",
       "\\item 'rs2853515'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'rs869183622'\n",
       "2. 'rs62581312'\n",
       "3. 'rs41323649'\n",
       "4. 'rs3937037'\n",
       "5. ''\n",
       "6. 'rs2853515'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"rs869183622\" \"rs62581312\"  \"rs41323649\"  \"rs3937037\"   \"\"           \n",
       "[6] \"rs2853515\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(old_annot_df$ID, snp_qc_df$rs_id) %>% head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ld_indep</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>FALSE</td><td>422831</td></tr>\n",
       "\t<tr><td> TRUE</td><td>361425</td></tr>\n",
       "\t<tr><td>   NA</td><td> 19813</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " ld\\_indep & n\\\\\n",
       " <lgl> & <int>\\\\\n",
       "\\hline\n",
       "\t FALSE & 422831\\\\\n",
       "\t  TRUE & 361425\\\\\n",
       "\t    NA &  19813\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 3 × 2\n",
       "\n",
       "| ld_indep &lt;lgl&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| FALSE | 422831 |\n",
       "|  TRUE | 361425 |\n",
       "|    NA |  19813 |\n",
       "\n"
      ],
      "text/plain": [
       "  ld_indep n     \n",
       "1 FALSE    422831\n",
       "2  TRUE    361425\n",
       "3    NA     19813"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>%\n",
    "count(ld_indep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CHROM'</li><li>'POS'</li><li>'REF'</li><li>'ALT'</li><li>'ID'</li><li>'Gene'</li><li>'Consequence'</li><li>'HGVSp'</li><li>'LoF'</li><li>'LoF_filter'</li><li>'LoF_flags'</li><li>'LoF_info'</li><li>'consequence_field'</li><li>'f_miss'</li><li>'f_miss_bileve'</li><li>'f_miss_wcsg'</li><li>'freq'</li><li>'hwe_p'</li><li>'maf'</li><li>'ld_indep'</li><li>'wcsg_only'</li><li>'bileve_only'</li><li>'filter'</li><li>'missingness'</li><li>'hwe'</li><li>'mcpi'</li><li>'gnomad_af'</li><li>'mgi'</li><li>'mgi_notes'</li><li>'all_filters'</li><li>'Gene_symbol'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CHROM'\n",
       "\\item 'POS'\n",
       "\\item 'REF'\n",
       "\\item 'ALT'\n",
       "\\item 'ID'\n",
       "\\item 'Gene'\n",
       "\\item 'Consequence'\n",
       "\\item 'HGVSp'\n",
       "\\item 'LoF'\n",
       "\\item 'LoF\\_filter'\n",
       "\\item 'LoF\\_flags'\n",
       "\\item 'LoF\\_info'\n",
       "\\item 'consequence\\_field'\n",
       "\\item 'f\\_miss'\n",
       "\\item 'f\\_miss\\_bileve'\n",
       "\\item 'f\\_miss\\_wcsg'\n",
       "\\item 'freq'\n",
       "\\item 'hwe\\_p'\n",
       "\\item 'maf'\n",
       "\\item 'ld\\_indep'\n",
       "\\item 'wcsg\\_only'\n",
       "\\item 'bileve\\_only'\n",
       "\\item 'filter'\n",
       "\\item 'missingness'\n",
       "\\item 'hwe'\n",
       "\\item 'mcpi'\n",
       "\\item 'gnomad\\_af'\n",
       "\\item 'mgi'\n",
       "\\item 'mgi\\_notes'\n",
       "\\item 'all\\_filters'\n",
       "\\item 'Gene\\_symbol'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CHROM'\n",
       "2. 'POS'\n",
       "3. 'REF'\n",
       "4. 'ALT'\n",
       "5. 'ID'\n",
       "6. 'Gene'\n",
       "7. 'Consequence'\n",
       "8. 'HGVSp'\n",
       "9. 'LoF'\n",
       "10. 'LoF_filter'\n",
       "11. 'LoF_flags'\n",
       "12. 'LoF_info'\n",
       "13. 'consequence_field'\n",
       "14. 'f_miss'\n",
       "15. 'f_miss_bileve'\n",
       "16. 'f_miss_wcsg'\n",
       "17. 'freq'\n",
       "18. 'hwe_p'\n",
       "19. 'maf'\n",
       "20. 'ld_indep'\n",
       "21. 'wcsg_only'\n",
       "22. 'bileve_only'\n",
       "23. 'filter'\n",
       "24. 'missingness'\n",
       "25. 'hwe'\n",
       "26. 'mcpi'\n",
       "27. 'gnomad_af'\n",
       "28. 'mgi'\n",
       "29. 'mgi_notes'\n",
       "30. 'all_filters'\n",
       "31. 'Gene_symbol'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CHROM\"             \"POS\"               \"REF\"              \n",
       " [4] \"ALT\"               \"ID\"                \"Gene\"             \n",
       " [7] \"Consequence\"       \"HGVSp\"             \"LoF\"              \n",
       "[10] \"LoF_filter\"        \"LoF_flags\"         \"LoF_info\"         \n",
       "[13] \"consequence_field\" \"f_miss\"            \"f_miss_bileve\"    \n",
       "[16] \"f_miss_wcsg\"       \"freq\"              \"hwe_p\"            \n",
       "[19] \"maf\"               \"ld_indep\"          \"wcsg_only\"        \n",
       "[22] \"bileve_only\"       \"filter\"            \"missingness\"      \n",
       "[25] \"hwe\"               \"mcpi\"              \"gnomad_af\"        \n",
       "[28] \"mgi\"               \"mgi_notes\"         \"all_filters\"      \n",
       "[31] \"Gene_symbol\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>%\n",
    "colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>array</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td> 17526</td></tr>\n",
       "\t<tr><td>1</td><td> 34192</td></tr>\n",
       "\t<tr><td>2</td><td>752351</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " array & n\\\\\n",
       " <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 &  17526\\\\\n",
       "\t 1 &  34192\\\\\n",
       "\t 2 & 752351\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 3 × 2\n",
       "\n",
       "| array &lt;int&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 |  17526 |\n",
       "| 1 |  34192 |\n",
       "| 2 | 752351 |\n",
       "\n"
      ],
      "text/plain": [
       "  array n     \n",
       "1 0      17526\n",
       "2 1      34192\n",
       "3 2     752351"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>%\n",
    "select(CHROM, POS, REF, ALT, wcsg_only, bileve_only) %>%\n",
    "left_join(\n",
    "    snp_qc_df %>% \n",
    "    mutate(CHROM = if_else(chromosome == '23', 'X', if_else(chromosome == '24', 'Y', if_else(chromosome == '25', 'X', if_else(chromosome == '26', 'MT', chromosome))))) %>%\n",
    "    select(CHROM, position, allele1_ref, allele2_alt, array),\n",
    "    by=c('CHROM'='CHROM', 'POS'='position', 'REF'='allele1_ref', 'ALT'='allele2_alt')\n",
    ") %>%\n",
    "count(array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 8 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>wcsg_only</th><th scope=col>bileve_only</th><th scope=col>array</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>FALSE</td><td>FALSE</td><td> 2</td><td>732880</td></tr>\n",
       "\t<tr><td>FALSE</td><td> TRUE</td><td> 0</td><td> 17103</td></tr>\n",
       "\t<tr><td> TRUE</td><td>FALSE</td><td> 1</td><td> 33831</td></tr>\n",
       "\t<tr><td> TRUE</td><td>FALSE</td><td> 2</td><td>   442</td></tr>\n",
       "\t<tr><td>   NA</td><td>   NA</td><td> 0</td><td>   278</td></tr>\n",
       "\t<tr><td>   NA</td><td>   NA</td><td> 1</td><td>    99</td></tr>\n",
       "\t<tr><td>   NA</td><td>   NA</td><td> 2</td><td> 17157</td></tr>\n",
       "\t<tr><td>   NA</td><td>   NA</td><td>NA</td><td>  2279</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 8 × 4\n",
       "\\begin{tabular}{llll}\n",
       " wcsg\\_only & bileve\\_only & array & n\\\\\n",
       " <lgl> & <lgl> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t FALSE & FALSE &  2 & 732880\\\\\n",
       "\t FALSE &  TRUE &  0 &  17103\\\\\n",
       "\t  TRUE & FALSE &  1 &  33831\\\\\n",
       "\t  TRUE & FALSE &  2 &    442\\\\\n",
       "\t    NA &    NA &  0 &    278\\\\\n",
       "\t    NA &    NA &  1 &     99\\\\\n",
       "\t    NA &    NA &  2 &  17157\\\\\n",
       "\t    NA &    NA & NA &   2279\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 8 × 4\n",
       "\n",
       "| wcsg_only &lt;lgl&gt; | bileve_only &lt;lgl&gt; | array &lt;int&gt; | n &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| FALSE | FALSE |  2 | 732880 |\n",
       "| FALSE |  TRUE |  0 |  17103 |\n",
       "|  TRUE | FALSE |  1 |  33831 |\n",
       "|  TRUE | FALSE |  2 |    442 |\n",
       "|    NA |    NA |  0 |    278 |\n",
       "|    NA |    NA |  1 |     99 |\n",
       "|    NA |    NA |  2 |  17157 |\n",
       "|    NA |    NA | NA |   2279 |\n",
       "\n"
      ],
      "text/plain": [
       "  wcsg_only bileve_only array n     \n",
       "1 FALSE     FALSE        2    732880\n",
       "2 FALSE      TRUE        0     17103\n",
       "3  TRUE     FALSE        1     33831\n",
       "4  TRUE     FALSE        2       442\n",
       "5    NA        NA        0       278\n",
       "6    NA        NA        1        99\n",
       "7    NA        NA        2     17157\n",
       "8    NA        NA       NA      2279"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_annot_df %>%\n",
    "select(ID, wcsg_only, bileve_only) %>%\n",
    "left_join(\n",
    "    snp_qc_df %>% select(rs_id, array),\n",
    "    by=c('ID'='rs_id')\n",
    ") %>%\n",
    "count(wcsg_only, bileve_only, array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 6 × 18</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>f_miss</th><th scope=col>f_miss_bileve</th><th scope=col>f_miss_wcsg</th><th scope=col>freq</th><th scope=col>hwe_p</th><th scope=col>maf</th><th scope=col>ld_indep</th><th scope=col>wcsg_only</th><th scope=col>bileve_only</th><th scope=col>filter</th><th scope=col>missingness</th><th scope=col>hwe</th><th scope=col>mcpi</th><th scope=col>gnomad_af</th><th scope=col>mgi</th><th scope=col>mgi_notes</th><th scope=col>all_filters</th><th scope=col>Gene_symbol</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.891500</td><td>0.011990</td><td>1.000000</td><td>0.03614952</td><td> 4.459e-22</td><td>0.03614952</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td>NA</td><td>1</td><td>1</td><td>0</td><td></td><td></td><td></td><td>2</td><td>AL669831.5    </td></tr>\n",
       "\t<tr><td>0.099850</td><td>0.126300</td><td>0.096580</td><td>0.05360614</td><td> 0.000e+00</td><td>0.05360614</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>1</td><td>1</td><td>0</td><td></td><td></td><td></td><td>2</td><td>AL669831.5    </td></tr>\n",
       "\t<tr><td>0.044820</td><td>0.107900</td><td>0.037040</td><td>0.03658726</td><td>7.009e-195</td><td>0.03658726</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>1</td><td>1</td><td>0</td><td></td><td></td><td></td><td>2</td><td>AL669831.5    </td></tr>\n",
       "\t<tr><td>0.005184</td><td>0.003537</td><td>0.005387</td><td>0.15681154</td><td>1.554e-266</td><td>0.15681154</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>0</td><td>1</td><td>0</td><td></td><td></td><td></td><td>1</td><td>RP11-206L10.10</td></tr>\n",
       "\t<tr><td>0.001100</td><td>0.001269</td><td>0.001079</td><td>0.03652657</td><td>4.046e-198</td><td>0.03652657</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>0</td><td>1</td><td>0</td><td></td><td></td><td></td><td>1</td><td>FAM87B        </td></tr>\n",
       "\t<tr><td>0.001874</td><td>0.004130</td><td>0.001596</td><td>0.12983398</td><td> 2.418e-02</td><td>0.12983398</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>0</td><td>0</td><td>0</td><td></td><td></td><td></td><td>0</td><td>RP11-206L10.10</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 18\n",
       "\\begin{tabular}{llllllllllllllllll}\n",
       " f\\_miss & f\\_miss\\_bileve & f\\_miss\\_wcsg & freq & hwe\\_p & maf & ld\\_indep & wcsg\\_only & bileve\\_only & filter & missingness & hwe & mcpi & gnomad\\_af & mgi & mgi\\_notes & all\\_filters & Gene\\_symbol\\\\\n",
       " <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0.891500 & 0.011990 & 1.000000 & 0.03614952 &  4.459e-22 & 0.03614952 & FALSE & FALSE &  TRUE & NA & 1 & 1 & 0 &  &  &  & 2 & AL669831.5    \\\\\n",
       "\t 0.099850 & 0.126300 & 0.096580 & 0.05360614 &  0.000e+00 & 0.05360614 & FALSE & FALSE & FALSE & NA & 1 & 1 & 0 &  &  &  & 2 & AL669831.5    \\\\\n",
       "\t 0.044820 & 0.107900 & 0.037040 & 0.03658726 & 7.009e-195 & 0.03658726 & FALSE & FALSE & FALSE & NA & 1 & 1 & 0 &  &  &  & 2 & AL669831.5    \\\\\n",
       "\t 0.005184 & 0.003537 & 0.005387 & 0.15681154 & 1.554e-266 & 0.15681154 & FALSE & FALSE & FALSE & NA & 0 & 1 & 0 &  &  &  & 1 & RP11-206L10.10\\\\\n",
       "\t 0.001100 & 0.001269 & 0.001079 & 0.03652657 & 4.046e-198 & 0.03652657 & FALSE & FALSE & FALSE & NA & 0 & 1 & 0 &  &  &  & 1 & FAM87B        \\\\\n",
       "\t 0.001874 & 0.004130 & 0.001596 & 0.12983398 &  2.418e-02 & 0.12983398 & FALSE & FALSE & FALSE & NA & 0 & 0 & 0 &  &  &  & 0 & RP11-206L10.10\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 18\n",
       "\n",
       "| f_miss &lt;dbl&gt; | f_miss_bileve &lt;dbl&gt; | f_miss_wcsg &lt;dbl&gt; | freq &lt;dbl&gt; | hwe_p &lt;dbl&gt; | maf &lt;dbl&gt; | ld_indep &lt;lgl&gt; | wcsg_only &lt;lgl&gt; | bileve_only &lt;lgl&gt; | filter &lt;lgl&gt; | missingness &lt;int&gt; | hwe &lt;int&gt; | mcpi &lt;int&gt; | gnomad_af &lt;chr&gt; | mgi &lt;chr&gt; | mgi_notes &lt;chr&gt; | all_filters &lt;int&gt; | Gene_symbol &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0.891500 | 0.011990 | 1.000000 | 0.03614952 |  4.459e-22 | 0.03614952 | FALSE | FALSE |  TRUE | NA | 1 | 1 | 0 | <!----> | <!----> | <!----> | 2 | AL669831.5     |\n",
       "| 0.099850 | 0.126300 | 0.096580 | 0.05360614 |  0.000e+00 | 0.05360614 | FALSE | FALSE | FALSE | NA | 1 | 1 | 0 | <!----> | <!----> | <!----> | 2 | AL669831.5     |\n",
       "| 0.044820 | 0.107900 | 0.037040 | 0.03658726 | 7.009e-195 | 0.03658726 | FALSE | FALSE | FALSE | NA | 1 | 1 | 0 | <!----> | <!----> | <!----> | 2 | AL669831.5     |\n",
       "| 0.005184 | 0.003537 | 0.005387 | 0.15681154 | 1.554e-266 | 0.15681154 | FALSE | FALSE | FALSE | NA | 0 | 1 | 0 | <!----> | <!----> | <!----> | 1 | RP11-206L10.10 |\n",
       "| 0.001100 | 0.001269 | 0.001079 | 0.03652657 | 4.046e-198 | 0.03652657 | FALSE | FALSE | FALSE | NA | 0 | 1 | 0 | <!----> | <!----> | <!----> | 1 | FAM87B         |\n",
       "| 0.001874 | 0.004130 | 0.001596 | 0.12983398 |  2.418e-02 | 0.12983398 | FALSE | FALSE | FALSE | NA | 0 | 0 | 0 | <!----> | <!----> | <!----> | 0 | RP11-206L10.10 |\n",
       "\n"
      ],
      "text/plain": [
       "  f_miss   f_miss_bileve f_miss_wcsg freq       hwe_p      maf        ld_indep\n",
       "1 0.891500 0.011990      1.000000    0.03614952  4.459e-22 0.03614952 FALSE   \n",
       "2 0.099850 0.126300      0.096580    0.05360614  0.000e+00 0.05360614 FALSE   \n",
       "3 0.044820 0.107900      0.037040    0.03658726 7.009e-195 0.03658726 FALSE   \n",
       "4 0.005184 0.003537      0.005387    0.15681154 1.554e-266 0.15681154 FALSE   \n",
       "5 0.001100 0.001269      0.001079    0.03652657 4.046e-198 0.03652657 FALSE   \n",
       "6 0.001874 0.004130      0.001596    0.12983398  2.418e-02 0.12983398 FALSE   \n",
       "  wcsg_only bileve_only filter missingness hwe mcpi gnomad_af mgi mgi_notes\n",
       "1 FALSE      TRUE       NA     1           1   0                           \n",
       "2 FALSE     FALSE       NA     1           1   0                           \n",
       "3 FALSE     FALSE       NA     1           1   0                           \n",
       "4 FALSE     FALSE       NA     0           1   0                           \n",
       "5 FALSE     FALSE       NA     0           1   0                           \n",
       "6 FALSE     FALSE       NA     0           0   0                           \n",
       "  all_filters Gene_symbol   \n",
       "1 2           AL669831.5    \n",
       "2 2           AL669831.5    \n",
       "3 2           AL669831.5    \n",
       "4 1           RP11-206L10.10\n",
       "5 1           FAM87B        \n",
       "6 0           RP11-206L10.10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "old_cal_annot_df %>%\n",
    "select(all_of(setdiff(colnames(old_cal_annot_df), colnames(cal_annot_df)))) %>%\n",
    "select(-consequence_field) %>%\n",
    "head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.table: 6 × 159</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>rs_id</th><th scope=col>affymetrix_snp_id</th><th scope=col>affymetrix_probeset_id</th><th scope=col>chromosome</th><th scope=col>position</th><th scope=col>allele1_ref</th><th scope=col>allele2_alt</th><th scope=col>strand</th><th scope=col>array</th><th scope=col>Batch_b001_qc</th><th scope=col>⋯</th><th scope=col>PC32_loading</th><th scope=col>PC33_loading</th><th scope=col>PC34_loading</th><th scope=col>PC35_loading</th><th scope=col>PC36_loading</th><th scope=col>PC37_loading</th><th scope=col>PC38_loading</th><th scope=col>PC9_loading</th><th scope=col>PC40_loading</th><th scope=col>in_Phasing_Input</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>rs28659788 </td><td>Affx-13546538</td><td>AX-32115783</td><td>1</td><td>723307</td><td>C</td><td>G</td><td>+</td><td>0</td><td>1</td><td>⋯</td><td>         NA</td><td>        NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>        NA</td><td>          NA</td><td>0</td></tr>\n",
       "\t<tr><td>rs116587930</td><td>Affx-35298040</td><td>AX-37361813</td><td>1</td><td>727841</td><td>G</td><td>A</td><td>+</td><td>2</td><td>1</td><td>⋯</td><td>         NA</td><td>        NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>        NA</td><td>          NA</td><td>0</td></tr>\n",
       "\t<tr><td>rs116720794</td><td>Affx-13637449</td><td>AX-32137419</td><td>1</td><td>729632</td><td>C</td><td>T</td><td>+</td><td>2</td><td>1</td><td>⋯</td><td>         NA</td><td>        NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>        NA</td><td>          NA</td><td>0</td></tr>\n",
       "\t<tr><td>rs3131972  </td><td>Affx-13945728</td><td>AX-13191280</td><td>1</td><td>752721</td><td>A</td><td>G</td><td>+</td><td>2</td><td>1</td><td>⋯</td><td>         NA</td><td>        NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>        NA</td><td>          NA</td><td>1</td></tr>\n",
       "\t<tr><td>rs12184325 </td><td>Affx-13963217</td><td>AX-11194291</td><td>1</td><td>754105</td><td>C</td><td>T</td><td>+</td><td>2</td><td>1</td><td>⋯</td><td>-0.00347144</td><td>0.00589896</td><td>-0.00373881</td><td>-0.00189571</td><td>-0.00286888</td><td>0.000792278</td><td>-0.00191024</td><td>0.00307453</td><td>-0.000410934</td><td>1</td></tr>\n",
       "\t<tr><td>rs3131962  </td><td>Affx-13995532</td><td>AX-32225497</td><td>1</td><td>756604</td><td>A</td><td>G</td><td>+</td><td>2</td><td>1</td><td>⋯</td><td>         NA</td><td>        NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>         NA</td><td>        NA</td><td>          NA</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 159\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " rs\\_id & affymetrix\\_snp\\_id & affymetrix\\_probeset\\_id & chromosome & position & allele1\\_ref & allele2\\_alt & strand & array & Batch\\_b001\\_qc & ⋯ & PC32\\_loading & PC33\\_loading & PC34\\_loading & PC35\\_loading & PC36\\_loading & PC37\\_loading & PC38\\_loading & PC9\\_loading & PC40\\_loading & in\\_Phasing\\_Input\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <int> & <int> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t rs28659788  & Affx-13546538 & AX-32115783 & 1 & 723307 & C & G & + & 0 & 1 & ⋯ &          NA &         NA &          NA &          NA &          NA &          NA &          NA &         NA &           NA & 0\\\\\n",
       "\t rs116587930 & Affx-35298040 & AX-37361813 & 1 & 727841 & G & A & + & 2 & 1 & ⋯ &          NA &         NA &          NA &          NA &          NA &          NA &          NA &         NA &           NA & 0\\\\\n",
       "\t rs116720794 & Affx-13637449 & AX-32137419 & 1 & 729632 & C & T & + & 2 & 1 & ⋯ &          NA &         NA &          NA &          NA &          NA &          NA &          NA &         NA &           NA & 0\\\\\n",
       "\t rs3131972   & Affx-13945728 & AX-13191280 & 1 & 752721 & A & G & + & 2 & 1 & ⋯ &          NA &         NA &          NA &          NA &          NA &          NA &          NA &         NA &           NA & 1\\\\\n",
       "\t rs12184325  & Affx-13963217 & AX-11194291 & 1 & 754105 & C & T & + & 2 & 1 & ⋯ & -0.00347144 & 0.00589896 & -0.00373881 & -0.00189571 & -0.00286888 & 0.000792278 & -0.00191024 & 0.00307453 & -0.000410934 & 1\\\\\n",
       "\t rs3131962   & Affx-13995532 & AX-32225497 & 1 & 756604 & A & G & + & 2 & 1 & ⋯ &          NA &         NA &          NA &          NA &          NA &          NA &          NA &         NA &           NA & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 159\n",
       "\n",
       "| rs_id &lt;chr&gt; | affymetrix_snp_id &lt;chr&gt; | affymetrix_probeset_id &lt;chr&gt; | chromosome &lt;chr&gt; | position &lt;int&gt; | allele1_ref &lt;chr&gt; | allele2_alt &lt;chr&gt; | strand &lt;chr&gt; | array &lt;int&gt; | Batch_b001_qc &lt;int&gt; | ⋯ ⋯ | PC32_loading &lt;dbl&gt; | PC33_loading &lt;dbl&gt; | PC34_loading &lt;dbl&gt; | PC35_loading &lt;dbl&gt; | PC36_loading &lt;dbl&gt; | PC37_loading &lt;dbl&gt; | PC38_loading &lt;dbl&gt; | PC9_loading &lt;dbl&gt; | PC40_loading &lt;dbl&gt; | in_Phasing_Input &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| rs28659788  | Affx-13546538 | AX-32115783 | 1 | 723307 | C | G | + | 0 | 1 | ⋯ |          NA |         NA |          NA |          NA |          NA |          NA |          NA |         NA |           NA | 0 |\n",
       "| rs116587930 | Affx-35298040 | AX-37361813 | 1 | 727841 | G | A | + | 2 | 1 | ⋯ |          NA |         NA |          NA |          NA |          NA |          NA |          NA |         NA |           NA | 0 |\n",
       "| rs116720794 | Affx-13637449 | AX-32137419 | 1 | 729632 | C | T | + | 2 | 1 | ⋯ |          NA |         NA |          NA |          NA |          NA |          NA |          NA |         NA |           NA | 0 |\n",
       "| rs3131972   | Affx-13945728 | AX-13191280 | 1 | 752721 | A | G | + | 2 | 1 | ⋯ |          NA |         NA |          NA |          NA |          NA |          NA |          NA |         NA |           NA | 1 |\n",
       "| rs12184325  | Affx-13963217 | AX-11194291 | 1 | 754105 | C | T | + | 2 | 1 | ⋯ | -0.00347144 | 0.00589896 | -0.00373881 | -0.00189571 | -0.00286888 | 0.000792278 | -0.00191024 | 0.00307453 | -0.000410934 | 1 |\n",
       "| rs3131962   | Affx-13995532 | AX-32225497 | 1 | 756604 | A | G | + | 2 | 1 | ⋯ |          NA |         NA |          NA |          NA |          NA |          NA |          NA |         NA |           NA | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  rs_id       affymetrix_snp_id affymetrix_probeset_id chromosome position\n",
       "1 rs28659788  Affx-13546538     AX-32115783            1          723307  \n",
       "2 rs116587930 Affx-35298040     AX-37361813            1          727841  \n",
       "3 rs116720794 Affx-13637449     AX-32137419            1          729632  \n",
       "4 rs3131972   Affx-13945728     AX-13191280            1          752721  \n",
       "5 rs12184325  Affx-13963217     AX-11194291            1          754105  \n",
       "6 rs3131962   Affx-13995532     AX-32225497            1          756604  \n",
       "  allele1_ref allele2_alt strand array Batch_b001_qc ⋯ PC32_loading\n",
       "1 C           G           +      0     1             ⋯          NA \n",
       "2 G           A           +      2     1             ⋯          NA \n",
       "3 C           T           +      2     1             ⋯          NA \n",
       "4 A           G           +      2     1             ⋯          NA \n",
       "5 C           T           +      2     1             ⋯ -0.00347144 \n",
       "6 A           G           +      2     1             ⋯          NA \n",
       "  PC33_loading PC34_loading PC35_loading PC36_loading PC37_loading PC38_loading\n",
       "1         NA            NA           NA           NA           NA           NA \n",
       "2         NA            NA           NA           NA           NA           NA \n",
       "3         NA            NA           NA           NA           NA           NA \n",
       "4         NA            NA           NA           NA           NA           NA \n",
       "5 0.00589896   -0.00373881  -0.00189571  -0.00286888  0.000792278  -0.00191024 \n",
       "6         NA            NA           NA           NA           NA           NA \n",
       "  PC9_loading PC40_loading in_Phasing_Input\n",
       "1         NA            NA 0               \n",
       "2         NA            NA 0               \n",
       "3         NA            NA 0               \n",
       "4         NA            NA 1               \n",
       "5 0.00307453  -0.000410934 1               \n",
       "6         NA            NA 1               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_qc_df %>%\n",
    "head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'rs_id'</li><li>'affymetrix_snp_id'</li><li>'affymetrix_probeset_id'</li><li>'chromosome'</li><li>'position'</li><li>'allele1_ref'</li><li>'allele2_alt'</li><li>'strand'</li><li>'array'</li><li>'Batch_b001_qc'</li><li>'Batch_b002_qc'</li><li>'Batch_b003_qc'</li><li>'Batch_b004_qc'</li><li>'Batch_b005_qc'</li><li>'Batch_b006_qc'</li><li>'Batch_b007_qc'</li><li>'Batch_b008_qc'</li><li>'Batch_b009_qc'</li><li>'Batch_b010_qc'</li><li>'Batch_b011_qc'</li><li>'Batch_b012_qc'</li><li>'Batch_b013_qc'</li><li>'Batch_b014_qc'</li><li>'Batch_b015_qc'</li><li>'Batch_b016_qc'</li><li>'Batch_b017_qc'</li><li>'Batch_b018_qc'</li><li>'Batch_b019_qc'</li><li>'Batch_b020_qc'</li><li>'Batch_b021_qc'</li><li>'Batch_b022_qc'</li><li>'Batch_b023_qc'</li><li>'Batch_b024_qc'</li><li>'Batch_b025_qc'</li><li>'Batch_b026_qc'</li><li>'Batch_b027_qc'</li><li>'Batch_b028_qc'</li><li>'Batch_b029_qc'</li><li>'Batch_b030_qc'</li><li>'Batch_b031_qc'</li><li>'Batch_b032_qc'</li><li>'Batch_b033_qc'</li><li>'Batch_b034_qc'</li><li>'Batch_b035_qc'</li><li>'Batch_b036_qc'</li><li>'Batch_b037_qc'</li><li>'Batch_b038_qc'</li><li>'Batch_b039_qc'</li><li>'Batch_b040_qc'</li><li>'Batch_b041_qc'</li><li>'Batch_b042_qc'</li><li>'Batch_b043_qc'</li><li>'Batch_b044_qc'</li><li>'Batch_b045_qc'</li><li>'Batch_b046_qc'</li><li>'Batch_b047_qc'</li><li>'Batch_b048_qc'</li><li>'Batch_b049_qc'</li><li>'Batch_b050_qc'</li><li>'Batch_b051_qc'</li><li>'Batch_b052_qc'</li><li>'Batch_b053_qc'</li><li>'Batch_b054_qc'</li><li>'Batch_b055_qc'</li><li>'Batch_b056_qc'</li><li>'Batch_b057_qc'</li><li>'Batch_b058_qc'</li><li>'Batch_b059_qc'</li><li>'Batch_b060_qc'</li><li>'Batch_b061_qc'</li><li>'Batch_b062_qc'</li><li>'Batch_b063_qc'</li><li>'Batch_b064_qc'</li><li>'Batch_b065_qc'</li><li>'Batch_b066_qc'</li><li>'Batch_b067_qc'</li><li>'Batch_b068_qc'</li><li>'Batch_b069_qc'</li><li>'Batch_b070_qc'</li><li>'Batch_b071_qc'</li><li>'Batch_b072_qc'</li><li>'Batch_b073_qc'</li><li>'Batch_b074_qc'</li><li>'Batch_b075_qc'</li><li>'Batch_b076_qc'</li><li>'Batch_b077_qc'</li><li>'Batch_b078_qc'</li><li>'Batch_b079_qc'</li><li>'Batch_b080_qc'</li><li>'Batch_b081_qc'</li><li>'Batch_b082_qc'</li><li>'Batch_b083_qc'</li><li>'Batch_b084_qc'</li><li>'Batch_b085_qc'</li><li>'Batch_b086_qc'</li><li>'Batch_b087_qc'</li><li>'Batch_b088_qc'</li><li>'Batch_b089_qc'</li><li>'Batch_b090_qc'</li><li>'Batch_b091_qc'</li><li>'Batch_b092_qc'</li><li>'Batch_b093_qc'</li><li>'Batch_b094_qc'</li><li>'Batch_b095_qc'</li><li>'UKBiLEVEAX_b1_qc'</li><li>'UKBiLEVEAX_b2_qc'</li><li>'UKBiLEVEAX_b3_qc'</li><li>'UKBiLEVEAX_b4_qc'</li><li>'UKBiLEVEAX_b5_qc'</li><li>'UKBiLEVEAX_b6_qc'</li><li>'UKBiLEVEAX_b7_qc'</li><li>'UKBiLEVEAX_b8_qc'</li><li>'UKBiLEVEAX_b9_qc'</li><li>'UKBiLEVEAX_b10_qc'</li><li>'UKBiLEVEAX_b11_qc'</li><li>'in_HetMiss'</li><li>'in_Relatedness'</li><li>'in_PCA'</li><li>'PC1_loading'</li><li>'PC2_loading'</li><li>'PC3_loading'</li><li>'PC4_loading'</li><li>'PC5_loading'</li><li>'PC6_loading'</li><li>'PC7_loading'</li><li>'PC8_loading'</li><li>'PC9_loading'</li><li>'PC10_loading'</li><li>'PC11_loading'</li><li>'PC12_loading'</li><li>'PC13_loading'</li><li>'PC14_loading'</li><li>'PC15_loading'</li><li>'PC16_loading'</li><li>'PC17_loading'</li><li>'PC18_loading'</li><li>'PC9_loading'</li><li>'PC20_loading'</li><li>'PC21_loading'</li><li>'PC22_loading'</li><li>'PC23_loading'</li><li>'PC24_loading'</li><li>'PC25_loading'</li><li>'PC26_loading'</li><li>'PC27_loading'</li><li>'PC28_loading'</li><li>'PC9_loading'</li><li>'PC30_loading'</li><li>'PC31_loading'</li><li>'PC32_loading'</li><li>'PC33_loading'</li><li>'PC34_loading'</li><li>'PC35_loading'</li><li>'PC36_loading'</li><li>'PC37_loading'</li><li>'PC38_loading'</li><li>'PC9_loading'</li><li>'PC40_loading'</li><li>'in_Phasing_Input'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'rs\\_id'\n",
       "\\item 'affymetrix\\_snp\\_id'\n",
       "\\item 'affymetrix\\_probeset\\_id'\n",
       "\\item 'chromosome'\n",
       "\\item 'position'\n",
       "\\item 'allele1\\_ref'\n",
       "\\item 'allele2\\_alt'\n",
       "\\item 'strand'\n",
       "\\item 'array'\n",
       "\\item 'Batch\\_b001\\_qc'\n",
       "\\item 'Batch\\_b002\\_qc'\n",
       "\\item 'Batch\\_b003\\_qc'\n",
       "\\item 'Batch\\_b004\\_qc'\n",
       "\\item 'Batch\\_b005\\_qc'\n",
       "\\item 'Batch\\_b006\\_qc'\n",
       "\\item 'Batch\\_b007\\_qc'\n",
       "\\item 'Batch\\_b008\\_qc'\n",
       "\\item 'Batch\\_b009\\_qc'\n",
       "\\item 'Batch\\_b010\\_qc'\n",
       "\\item 'Batch\\_b011\\_qc'\n",
       "\\item 'Batch\\_b012\\_qc'\n",
       "\\item 'Batch\\_b013\\_qc'\n",
       "\\item 'Batch\\_b014\\_qc'\n",
       "\\item 'Batch\\_b015\\_qc'\n",
       "\\item 'Batch\\_b016\\_qc'\n",
       "\\item 'Batch\\_b017\\_qc'\n",
       "\\item 'Batch\\_b018\\_qc'\n",
       "\\item 'Batch\\_b019\\_qc'\n",
       "\\item 'Batch\\_b020\\_qc'\n",
       "\\item 'Batch\\_b021\\_qc'\n",
       "\\item 'Batch\\_b022\\_qc'\n",
       "\\item 'Batch\\_b023\\_qc'\n",
       "\\item 'Batch\\_b024\\_qc'\n",
       "\\item 'Batch\\_b025\\_qc'\n",
       "\\item 'Batch\\_b026\\_qc'\n",
       "\\item 'Batch\\_b027\\_qc'\n",
       "\\item 'Batch\\_b028\\_qc'\n",
       "\\item 'Batch\\_b029\\_qc'\n",
       "\\item 'Batch\\_b030\\_qc'\n",
       "\\item 'Batch\\_b031\\_qc'\n",
       "\\item 'Batch\\_b032\\_qc'\n",
       "\\item 'Batch\\_b033\\_qc'\n",
       "\\item 'Batch\\_b034\\_qc'\n",
       "\\item 'Batch\\_b035\\_qc'\n",
       "\\item 'Batch\\_b036\\_qc'\n",
       "\\item 'Batch\\_b037\\_qc'\n",
       "\\item 'Batch\\_b038\\_qc'\n",
       "\\item 'Batch\\_b039\\_qc'\n",
       "\\item 'Batch\\_b040\\_qc'\n",
       "\\item 'Batch\\_b041\\_qc'\n",
       "\\item 'Batch\\_b042\\_qc'\n",
       "\\item 'Batch\\_b043\\_qc'\n",
       "\\item 'Batch\\_b044\\_qc'\n",
       "\\item 'Batch\\_b045\\_qc'\n",
       "\\item 'Batch\\_b046\\_qc'\n",
       "\\item 'Batch\\_b047\\_qc'\n",
       "\\item 'Batch\\_b048\\_qc'\n",
       "\\item 'Batch\\_b049\\_qc'\n",
       "\\item 'Batch\\_b050\\_qc'\n",
       "\\item 'Batch\\_b051\\_qc'\n",
       "\\item 'Batch\\_b052\\_qc'\n",
       "\\item 'Batch\\_b053\\_qc'\n",
       "\\item 'Batch\\_b054\\_qc'\n",
       "\\item 'Batch\\_b055\\_qc'\n",
       "\\item 'Batch\\_b056\\_qc'\n",
       "\\item 'Batch\\_b057\\_qc'\n",
       "\\item 'Batch\\_b058\\_qc'\n",
       "\\item 'Batch\\_b059\\_qc'\n",
       "\\item 'Batch\\_b060\\_qc'\n",
       "\\item 'Batch\\_b061\\_qc'\n",
       "\\item 'Batch\\_b062\\_qc'\n",
       "\\item 'Batch\\_b063\\_qc'\n",
       "\\item 'Batch\\_b064\\_qc'\n",
       "\\item 'Batch\\_b065\\_qc'\n",
       "\\item 'Batch\\_b066\\_qc'\n",
       "\\item 'Batch\\_b067\\_qc'\n",
       "\\item 'Batch\\_b068\\_qc'\n",
       "\\item 'Batch\\_b069\\_qc'\n",
       "\\item 'Batch\\_b070\\_qc'\n",
       "\\item 'Batch\\_b071\\_qc'\n",
       "\\item 'Batch\\_b072\\_qc'\n",
       "\\item 'Batch\\_b073\\_qc'\n",
       "\\item 'Batch\\_b074\\_qc'\n",
       "\\item 'Batch\\_b075\\_qc'\n",
       "\\item 'Batch\\_b076\\_qc'\n",
       "\\item 'Batch\\_b077\\_qc'\n",
       "\\item 'Batch\\_b078\\_qc'\n",
       "\\item 'Batch\\_b079\\_qc'\n",
       "\\item 'Batch\\_b080\\_qc'\n",
       "\\item 'Batch\\_b081\\_qc'\n",
       "\\item 'Batch\\_b082\\_qc'\n",
       "\\item 'Batch\\_b083\\_qc'\n",
       "\\item 'Batch\\_b084\\_qc'\n",
       "\\item 'Batch\\_b085\\_qc'\n",
       "\\item 'Batch\\_b086\\_qc'\n",
       "\\item 'Batch\\_b087\\_qc'\n",
       "\\item 'Batch\\_b088\\_qc'\n",
       "\\item 'Batch\\_b089\\_qc'\n",
       "\\item 'Batch\\_b090\\_qc'\n",
       "\\item 'Batch\\_b091\\_qc'\n",
       "\\item 'Batch\\_b092\\_qc'\n",
       "\\item 'Batch\\_b093\\_qc'\n",
       "\\item 'Batch\\_b094\\_qc'\n",
       "\\item 'Batch\\_b095\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b1\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b2\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b3\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b4\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b5\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b6\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b7\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b8\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b9\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b10\\_qc'\n",
       "\\item 'UKBiLEVEAX\\_b11\\_qc'\n",
       "\\item 'in\\_HetMiss'\n",
       "\\item 'in\\_Relatedness'\n",
       "\\item 'in\\_PCA'\n",
       "\\item 'PC1\\_loading'\n",
       "\\item 'PC2\\_loading'\n",
       "\\item 'PC3\\_loading'\n",
       "\\item 'PC4\\_loading'\n",
       "\\item 'PC5\\_loading'\n",
       "\\item 'PC6\\_loading'\n",
       "\\item 'PC7\\_loading'\n",
       "\\item 'PC8\\_loading'\n",
       "\\item 'PC9\\_loading'\n",
       "\\item 'PC10\\_loading'\n",
       "\\item 'PC11\\_loading'\n",
       "\\item 'PC12\\_loading'\n",
       "\\item 'PC13\\_loading'\n",
       "\\item 'PC14\\_loading'\n",
       "\\item 'PC15\\_loading'\n",
       "\\item 'PC16\\_loading'\n",
       "\\item 'PC17\\_loading'\n",
       "\\item 'PC18\\_loading'\n",
       "\\item 'PC9\\_loading'\n",
       "\\item 'PC20\\_loading'\n",
       "\\item 'PC21\\_loading'\n",
       "\\item 'PC22\\_loading'\n",
       "\\item 'PC23\\_loading'\n",
       "\\item 'PC24\\_loading'\n",
       "\\item 'PC25\\_loading'\n",
       "\\item 'PC26\\_loading'\n",
       "\\item 'PC27\\_loading'\n",
       "\\item 'PC28\\_loading'\n",
       "\\item 'PC9\\_loading'\n",
       "\\item 'PC30\\_loading'\n",
       "\\item 'PC31\\_loading'\n",
       "\\item 'PC32\\_loading'\n",
       "\\item 'PC33\\_loading'\n",
       "\\item 'PC34\\_loading'\n",
       "\\item 'PC35\\_loading'\n",
       "\\item 'PC36\\_loading'\n",
       "\\item 'PC37\\_loading'\n",
       "\\item 'PC38\\_loading'\n",
       "\\item 'PC9\\_loading'\n",
       "\\item 'PC40\\_loading'\n",
       "\\item 'in\\_Phasing\\_Input'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'rs_id'\n",
       "2. 'affymetrix_snp_id'\n",
       "3. 'affymetrix_probeset_id'\n",
       "4. 'chromosome'\n",
       "5. 'position'\n",
       "6. 'allele1_ref'\n",
       "7. 'allele2_alt'\n",
       "8. 'strand'\n",
       "9. 'array'\n",
       "10. 'Batch_b001_qc'\n",
       "11. 'Batch_b002_qc'\n",
       "12. 'Batch_b003_qc'\n",
       "13. 'Batch_b004_qc'\n",
       "14. 'Batch_b005_qc'\n",
       "15. 'Batch_b006_qc'\n",
       "16. 'Batch_b007_qc'\n",
       "17. 'Batch_b008_qc'\n",
       "18. 'Batch_b009_qc'\n",
       "19. 'Batch_b010_qc'\n",
       "20. 'Batch_b011_qc'\n",
       "21. 'Batch_b012_qc'\n",
       "22. 'Batch_b013_qc'\n",
       "23. 'Batch_b014_qc'\n",
       "24. 'Batch_b015_qc'\n",
       "25. 'Batch_b016_qc'\n",
       "26. 'Batch_b017_qc'\n",
       "27. 'Batch_b018_qc'\n",
       "28. 'Batch_b019_qc'\n",
       "29. 'Batch_b020_qc'\n",
       "30. 'Batch_b021_qc'\n",
       "31. 'Batch_b022_qc'\n",
       "32. 'Batch_b023_qc'\n",
       "33. 'Batch_b024_qc'\n",
       "34. 'Batch_b025_qc'\n",
       "35. 'Batch_b026_qc'\n",
       "36. 'Batch_b027_qc'\n",
       "37. 'Batch_b028_qc'\n",
       "38. 'Batch_b029_qc'\n",
       "39. 'Batch_b030_qc'\n",
       "40. 'Batch_b031_qc'\n",
       "41. 'Batch_b032_qc'\n",
       "42. 'Batch_b033_qc'\n",
       "43. 'Batch_b034_qc'\n",
       "44. 'Batch_b035_qc'\n",
       "45. 'Batch_b036_qc'\n",
       "46. 'Batch_b037_qc'\n",
       "47. 'Batch_b038_qc'\n",
       "48. 'Batch_b039_qc'\n",
       "49. 'Batch_b040_qc'\n",
       "50. 'Batch_b041_qc'\n",
       "51. 'Batch_b042_qc'\n",
       "52. 'Batch_b043_qc'\n",
       "53. 'Batch_b044_qc'\n",
       "54. 'Batch_b045_qc'\n",
       "55. 'Batch_b046_qc'\n",
       "56. 'Batch_b047_qc'\n",
       "57. 'Batch_b048_qc'\n",
       "58. 'Batch_b049_qc'\n",
       "59. 'Batch_b050_qc'\n",
       "60. 'Batch_b051_qc'\n",
       "61. 'Batch_b052_qc'\n",
       "62. 'Batch_b053_qc'\n",
       "63. 'Batch_b054_qc'\n",
       "64. 'Batch_b055_qc'\n",
       "65. 'Batch_b056_qc'\n",
       "66. 'Batch_b057_qc'\n",
       "67. 'Batch_b058_qc'\n",
       "68. 'Batch_b059_qc'\n",
       "69. 'Batch_b060_qc'\n",
       "70. 'Batch_b061_qc'\n",
       "71. 'Batch_b062_qc'\n",
       "72. 'Batch_b063_qc'\n",
       "73. 'Batch_b064_qc'\n",
       "74. 'Batch_b065_qc'\n",
       "75. 'Batch_b066_qc'\n",
       "76. 'Batch_b067_qc'\n",
       "77. 'Batch_b068_qc'\n",
       "78. 'Batch_b069_qc'\n",
       "79. 'Batch_b070_qc'\n",
       "80. 'Batch_b071_qc'\n",
       "81. 'Batch_b072_qc'\n",
       "82. 'Batch_b073_qc'\n",
       "83. 'Batch_b074_qc'\n",
       "84. 'Batch_b075_qc'\n",
       "85. 'Batch_b076_qc'\n",
       "86. 'Batch_b077_qc'\n",
       "87. 'Batch_b078_qc'\n",
       "88. 'Batch_b079_qc'\n",
       "89. 'Batch_b080_qc'\n",
       "90. 'Batch_b081_qc'\n",
       "91. 'Batch_b082_qc'\n",
       "92. 'Batch_b083_qc'\n",
       "93. 'Batch_b084_qc'\n",
       "94. 'Batch_b085_qc'\n",
       "95. 'Batch_b086_qc'\n",
       "96. 'Batch_b087_qc'\n",
       "97. 'Batch_b088_qc'\n",
       "98. 'Batch_b089_qc'\n",
       "99. 'Batch_b090_qc'\n",
       "100. 'Batch_b091_qc'\n",
       "101. 'Batch_b092_qc'\n",
       "102. 'Batch_b093_qc'\n",
       "103. 'Batch_b094_qc'\n",
       "104. 'Batch_b095_qc'\n",
       "105. 'UKBiLEVEAX_b1_qc'\n",
       "106. 'UKBiLEVEAX_b2_qc'\n",
       "107. 'UKBiLEVEAX_b3_qc'\n",
       "108. 'UKBiLEVEAX_b4_qc'\n",
       "109. 'UKBiLEVEAX_b5_qc'\n",
       "110. 'UKBiLEVEAX_b6_qc'\n",
       "111. 'UKBiLEVEAX_b7_qc'\n",
       "112. 'UKBiLEVEAX_b8_qc'\n",
       "113. 'UKBiLEVEAX_b9_qc'\n",
       "114. 'UKBiLEVEAX_b10_qc'\n",
       "115. 'UKBiLEVEAX_b11_qc'\n",
       "116. 'in_HetMiss'\n",
       "117. 'in_Relatedness'\n",
       "118. 'in_PCA'\n",
       "119. 'PC1_loading'\n",
       "120. 'PC2_loading'\n",
       "121. 'PC3_loading'\n",
       "122. 'PC4_loading'\n",
       "123. 'PC5_loading'\n",
       "124. 'PC6_loading'\n",
       "125. 'PC7_loading'\n",
       "126. 'PC8_loading'\n",
       "127. 'PC9_loading'\n",
       "128. 'PC10_loading'\n",
       "129. 'PC11_loading'\n",
       "130. 'PC12_loading'\n",
       "131. 'PC13_loading'\n",
       "132. 'PC14_loading'\n",
       "133. 'PC15_loading'\n",
       "134. 'PC16_loading'\n",
       "135. 'PC17_loading'\n",
       "136. 'PC18_loading'\n",
       "137. 'PC9_loading'\n",
       "138. 'PC20_loading'\n",
       "139. 'PC21_loading'\n",
       "140. 'PC22_loading'\n",
       "141. 'PC23_loading'\n",
       "142. 'PC24_loading'\n",
       "143. 'PC25_loading'\n",
       "144. 'PC26_loading'\n",
       "145. 'PC27_loading'\n",
       "146. 'PC28_loading'\n",
       "147. 'PC9_loading'\n",
       "148. 'PC30_loading'\n",
       "149. 'PC31_loading'\n",
       "150. 'PC32_loading'\n",
       "151. 'PC33_loading'\n",
       "152. 'PC34_loading'\n",
       "153. 'PC35_loading'\n",
       "154. 'PC36_loading'\n",
       "155. 'PC37_loading'\n",
       "156. 'PC38_loading'\n",
       "157. 'PC9_loading'\n",
       "158. 'PC40_loading'\n",
       "159. 'in_Phasing_Input'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"rs_id\"                  \"affymetrix_snp_id\"      \"affymetrix_probeset_id\"\n",
       "  [4] \"chromosome\"             \"position\"               \"allele1_ref\"           \n",
       "  [7] \"allele2_alt\"            \"strand\"                 \"array\"                 \n",
       " [10] \"Batch_b001_qc\"          \"Batch_b002_qc\"          \"Batch_b003_qc\"         \n",
       " [13] \"Batch_b004_qc\"          \"Batch_b005_qc\"          \"Batch_b006_qc\"         \n",
       " [16] \"Batch_b007_qc\"          \"Batch_b008_qc\"          \"Batch_b009_qc\"         \n",
       " [19] \"Batch_b010_qc\"          \"Batch_b011_qc\"          \"Batch_b012_qc\"         \n",
       " [22] \"Batch_b013_qc\"          \"Batch_b014_qc\"          \"Batch_b015_qc\"         \n",
       " [25] \"Batch_b016_qc\"          \"Batch_b017_qc\"          \"Batch_b018_qc\"         \n",
       " [28] \"Batch_b019_qc\"          \"Batch_b020_qc\"          \"Batch_b021_qc\"         \n",
       " [31] \"Batch_b022_qc\"          \"Batch_b023_qc\"          \"Batch_b024_qc\"         \n",
       " [34] \"Batch_b025_qc\"          \"Batch_b026_qc\"          \"Batch_b027_qc\"         \n",
       " [37] \"Batch_b028_qc\"          \"Batch_b029_qc\"          \"Batch_b030_qc\"         \n",
       " [40] \"Batch_b031_qc\"          \"Batch_b032_qc\"          \"Batch_b033_qc\"         \n",
       " [43] \"Batch_b034_qc\"          \"Batch_b035_qc\"          \"Batch_b036_qc\"         \n",
       " [46] \"Batch_b037_qc\"          \"Batch_b038_qc\"          \"Batch_b039_qc\"         \n",
       " [49] \"Batch_b040_qc\"          \"Batch_b041_qc\"          \"Batch_b042_qc\"         \n",
       " [52] \"Batch_b043_qc\"          \"Batch_b044_qc\"          \"Batch_b045_qc\"         \n",
       " [55] \"Batch_b046_qc\"          \"Batch_b047_qc\"          \"Batch_b048_qc\"         \n",
       " [58] \"Batch_b049_qc\"          \"Batch_b050_qc\"          \"Batch_b051_qc\"         \n",
       " [61] \"Batch_b052_qc\"          \"Batch_b053_qc\"          \"Batch_b054_qc\"         \n",
       " [64] \"Batch_b055_qc\"          \"Batch_b056_qc\"          \"Batch_b057_qc\"         \n",
       " [67] \"Batch_b058_qc\"          \"Batch_b059_qc\"          \"Batch_b060_qc\"         \n",
       " [70] \"Batch_b061_qc\"          \"Batch_b062_qc\"          \"Batch_b063_qc\"         \n",
       " [73] \"Batch_b064_qc\"          \"Batch_b065_qc\"          \"Batch_b066_qc\"         \n",
       " [76] \"Batch_b067_qc\"          \"Batch_b068_qc\"          \"Batch_b069_qc\"         \n",
       " [79] \"Batch_b070_qc\"          \"Batch_b071_qc\"          \"Batch_b072_qc\"         \n",
       " [82] \"Batch_b073_qc\"          \"Batch_b074_qc\"          \"Batch_b075_qc\"         \n",
       " [85] \"Batch_b076_qc\"          \"Batch_b077_qc\"          \"Batch_b078_qc\"         \n",
       " [88] \"Batch_b079_qc\"          \"Batch_b080_qc\"          \"Batch_b081_qc\"         \n",
       " [91] \"Batch_b082_qc\"          \"Batch_b083_qc\"          \"Batch_b084_qc\"         \n",
       " [94] \"Batch_b085_qc\"          \"Batch_b086_qc\"          \"Batch_b087_qc\"         \n",
       " [97] \"Batch_b088_qc\"          \"Batch_b089_qc\"          \"Batch_b090_qc\"         \n",
       "[100] \"Batch_b091_qc\"          \"Batch_b092_qc\"          \"Batch_b093_qc\"         \n",
       "[103] \"Batch_b094_qc\"          \"Batch_b095_qc\"          \"UKBiLEVEAX_b1_qc\"      \n",
       "[106] \"UKBiLEVEAX_b2_qc\"       \"UKBiLEVEAX_b3_qc\"       \"UKBiLEVEAX_b4_qc\"      \n",
       "[109] \"UKBiLEVEAX_b5_qc\"       \"UKBiLEVEAX_b6_qc\"       \"UKBiLEVEAX_b7_qc\"      \n",
       "[112] \"UKBiLEVEAX_b8_qc\"       \"UKBiLEVEAX_b9_qc\"       \"UKBiLEVEAX_b10_qc\"     \n",
       "[115] \"UKBiLEVEAX_b11_qc\"      \"in_HetMiss\"             \"in_Relatedness\"        \n",
       "[118] \"in_PCA\"                 \"PC1_loading\"            \"PC2_loading\"           \n",
       "[121] \"PC3_loading\"            \"PC4_loading\"            \"PC5_loading\"           \n",
       "[124] \"PC6_loading\"            \"PC7_loading\"            \"PC8_loading\"           \n",
       "[127] \"PC9_loading\"            \"PC10_loading\"           \"PC11_loading\"          \n",
       "[130] \"PC12_loading\"           \"PC13_loading\"           \"PC14_loading\"          \n",
       "[133] \"PC15_loading\"           \"PC16_loading\"           \"PC17_loading\"          \n",
       "[136] \"PC18_loading\"           \"PC9_loading\"            \"PC20_loading\"          \n",
       "[139] \"PC21_loading\"           \"PC22_loading\"           \"PC23_loading\"          \n",
       "[142] \"PC24_loading\"           \"PC25_loading\"           \"PC26_loading\"          \n",
       "[145] \"PC27_loading\"           \"PC28_loading\"           \"PC9_loading\"           \n",
       "[148] \"PC30_loading\"           \"PC31_loading\"           \"PC32_loading\"          \n",
       "[151] \"PC33_loading\"           \"PC34_loading\"           \"PC35_loading\"          \n",
       "[154] \"PC36_loading\"           \"PC37_loading\"           \"PC38_loading\"          \n",
       "[157] \"PC9_loading\"            \"PC40_loading\"           \"in_Phasing_Input\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_qc_df %>%\n",
    "colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
